
```{r}
library("ggExtra")
library("ggplot2")
library('corrplot')
library(dplyr)
library(boot)
library(plotrix) 
library(cowplot)
library(VGAM) 
library(pheatmap)
library(RColorBrewer)
library(gplots)
library("reshape2")
library(ggcorrplot)
library(ComplexHeatmap)
library(circlize)
library(ggridges)
library(Skillings.Mack) #For the skillings.mack test, non-parametric repeated measures ANOVA bt with missing data structure
library(parallel)

```


Estimating the starting stem cell population of tissues from GTEx


```{r}
#load("/home/werner/xchrom_snp_skew/data/GTEx/snp_skews/sample_meta.Rdata")
#Skews with esscape filtered out
load("/home/werner/xchrom_snp_skew/data/GTEx/snp_skews/GATK_mod/v8_final_filtering_skew_and_stats_1_20_21_escape_filt_df.Rdata")
```

```{r}
head(skew_and_stats_df)
dim(skew_and_stats_df)

```



```{r}
skew_and_stats_df
```






Get some summary info before filtering on the number of SNPs and such.
Number of donors, average number of tissues per donor, total tissue sample size with estimated XCI skews

```{r}
#Ignore the cell cultures
temp_stats_df = skew_and_stats_df[!skew_and_stats_df$tissue %in% c('Cells - EBV-transformed lymphocytes','Cells - Cultured fibroblasts','Cells - Transformed fibroblasts'), ]

#Total number of donors
length(unique(temp_stats_df$donor))

#Number of estimates per tissue
table(temp_stats_df$tissue)

#total number of estimates
length(temp_stats_df$skew)

#Number of tissue samples per donor, but includes tissue replicates
mean(table(temp_stats_df$donor))

```

```{r}

#Average number of tissues with duplicated removed, still about 15
remove_dups <- function(x) {
  x = unique(x)
  return(length(x))
}


temp_stats_df = temp_stats_df %>%
 group_by(donor) %>%
 mutate(num_tiss_without_dups = remove_dups(tissue))


index = which(!duplicated(temp_stats_df$donor))
mean(temp_stats_df$num_tiss_without_dups[index])

```


Look at UPIC skews before skew estimate filtering
We actually don't keep any tissues from this donor acter filtering, maybe worth going back and looking at all the vcf files from this individual?

```{r}
skew_and_stats_df[grepl( 'X15G', skew_and_stats_df$donor), ]
```

```{r}
dim(skew_and_stats_df)

```


```{r}
skew_stat_plot = ggplot(skew_and_stats_df, aes(skew, num_snps)) + geom_point(alpha=1/5,size=2) + geom_density2d() +
  ylab('Number of SNPs used in skew estimate') + xlab('Skew estimate') + 
  xlim(.5,1) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_text(size=12),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12))

skew_stat_plot
ggsave(filename = 'num_snps_vs_skew.pdf', plot = skew_stat_plot, device = 'pdf', 
       path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/', useDingbats = F, height = 4, width = 6)

skew_stat_plot = ggplot(skew_and_stats_df, aes(skew, CI_width)) + geom_point(alpha=1/5,size=2) + geom_density2d() +
  ylab('CI around skew estimate') + xlab('Skew estimate') + 
  xlim(.5,1) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_text(size=12),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12))

skew_stat_plot
ggsave(filename = 'CI_vs_skew.pdf', plot = skew_stat_plot, device = 'pdf', 
       path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/', useDingbats = F, height = 4, width = 6)

skew_stat_plot = ggplot(skew_and_stats_df, aes(num_snps, CI_width)) + geom_point(alpha=1/5,size=2) + geom_density2d() +
  ylab('CI around skew estimate') + xlab('Number of SNPs') + geom_vline(xintercept = 10, color = 'red', size = 1.5) + geom_hline(yintercept = .15, color = 'red', size = 1.5) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_text(size=12),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12))

skew_stat_plot
ggsave(filename = 'num_snps_vs_CI.pdf', plot = skew_stat_plot, device = 'pdf', 
       path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/', useDingbats = F, height = 4, width = 6)
```

Set up a filter for samples with < 10 SNPs and CI >= .15

```{r}

num_snp_index = skew_and_stats_df$num_snps >= 10 & skew_and_stats_df$CI_width < .15
skew_and_stats_df = skew_and_stats_df[num_snp_index, ]
dim(skew_and_stats_df)

skew_stat_plot = ggplot(skew_and_stats_df, aes(skew, num_snps, color=p_significance)) + geom_point(alpha=1/5,size=1) + ylab('Number of SNPs used in skew estimate') + xlab('Skew estimate') + 
  ggtitle('KS-test for skew estimates') + theme(plot.title=element_text(hjust=.5)) + labs(color='KS-test significance')+ geom_density_2d() + xlim(.5,1)
ggMarginal(skew_stat_plot, groupColour=TRUE, groupFill=TRUE)


skew_stat_plot = ggplot(skew_and_stats_df, aes(skew, CI_width, color=p_significance)) + geom_point(alpha=1/5,size=1) + ylab('CI around skew estimate') + xlab('Skew estimate') + 
  ggtitle('KS-test for skew estimates') + theme(plot.title=element_text(hjust=.5)) + labs(color='KS-test significance')+ geom_density_2d() + xlim(.5,1)
ggMarginal(skew_stat_plot, groupColour=TRUE, groupFill=TRUE)

skew_stat_plot = ggplot(skew_and_stats_df, aes(num_snps, CI_width, color=p_significance)) + geom_point(alpha=1/5,size=1) + ylab('CI around skew estimate') + xlab('Number of SNPs') + 
  ggtitle('KS-test for skew estimates') + theme(plot.title=element_text(hjust=.5)) + labs(color='KS-test significance')+ geom_density_2d() + geom_vline(xintercept = 10)
ggMarginal(skew_stat_plot, groupColour=TRUE, groupFill=TRUE)

```

```{r}
table(skew_and_stats_df$p_significance)

```


Get summary stats on the number of well powered SNPs per sample, exclude the cell lines

```{r}

num_good_snps = as.integer(skew_and_stats_df$num_snps[!skew_and_stats_df$tissue %in% c('Cells - EBV-transformed lymphocytes','Cells - Cultured fibroblasts','Cells - Transformed fibroblasts')])
mean(num_good_snps)
sd(num_good_snps)

length(num_good_snps)

```

Number of donors, 

```{r}
table(table(skew_and_stats_df$donor[!skew_and_stats_df$tissue %in% c('Cells - EBV-transformed lymphocytes','Cells - Cultured fibroblasts','Cells - Transformed fibroblasts')]) > 0)

```



```{r}

tissue_bin_label = rep('Tissue skew < .55', dim(skew_and_stats_df)[1])
tissue_bin_label[skew_and_stats_df$skew >= .55 & skew_and_stats_df$skew < .6 ] = '.55 <= Tissue skew < .6'
tissue_bin_label[skew_and_stats_df$skew >= .6 & skew_and_stats_df$skew < .65 ] = '.6 <= Tissue skew < .65'
tissue_bin_label[skew_and_stats_df$skew >= .65 & skew_and_stats_df$skew < .7 ] = '.65 <= Tissue skew < .7'
tissue_bin_label[skew_and_stats_df$skew >= .7 & skew_and_stats_df$skew < .75 ] = '.7 <= Tissue skew < .75'
tissue_bin_label[skew_and_stats_df$skew >= .75 & skew_and_stats_df$skew < .8 ] = '.75 <= Tissue skew < .8'
tissue_bin_label[skew_and_stats_df$skew >= .8 & skew_and_stats_df$skew < .85 ] = '.8 <= Tissue skew < .85'
tissue_bin_label[skew_and_stats_df$skew >= .85 & skew_and_stats_df$skew < .9 ] = '.85 <= Tissue skew < .9'
tissue_bin_label[skew_and_stats_df$skew >= .9 & skew_and_stats_df$skew < .95 ] = '.9 <= Tissue skew < .95'

skew_and_stats_df$tissue_bin = tissue_bin_label
skew_and_stats_df$tissue_bin = factor(skew_and_stats_df$tissue_bin, levels = c('Tissue skew < .55','.55 <= Tissue skew < .6', 
                                                                               '.6 <= Tissue skew < .65', '.65 <= Tissue skew < .7', 
                                                                               '.7 <= Tissue skew < .75', '.75 <= Tissue skew < .8', 
                                                                               '.8 <= Tissue skew < .85', '.85 <= Tissue skew < .9', 
                                                                               '.9 <= Tissue skew < .95'))
ggplot(skew_and_stats_df, aes(x = tissue_bin, y = skew_sigma)) + geom_boxplot(varwidth = T)


```





Getting the age of the donors

```{r}
donor_phens = read.table("/home/werner/xchrom_snp_skew/data/GTEx/samples/GTEx_v8_Annotations_SubjectPhenotypesDS.txt", sep="\t", header=TRUE)  
donor_phens_names = sapply(strsplit(as.character(donor_phens$SUBJID), split='-', fixed=TRUE),'[[', 2)

usable_donors = names(table(skew_and_stats_df$donor))
table(usable_donors %in% donor_phens_names)


skew_and_stats_df$age = rep(NA, dim(skew_and_stats_df)[1])
for( i in 1:length(usable_donors)){
  #Get the age of the donor
  age = as.character(donor_phens$AGE[ grepl(usable_donors[i], as.character(donor_phens$SUBJID) ) ])
  #Pop it in for all the tissues from that donor
  skew_and_stats_df$age[skew_and_stats_df$donor == usable_donors[i]] = age

}

table(skew_and_stats_df$age)


```




Get summary plots of the number of tissues per person, number of people per tissue
Donors and number of tissues
Tissues and number of donors
Ranked bar graphs for both

```{r}
donors = skew_and_stats_df$donor[unique(skew_and_stats_df$donor)]
length(donors)
```



```{r}
ggplot(skew_and_stats_df, aes(x=age)) + geom_bar(fill='black') + ggtitle('Donor ages') + 
  theme( panel.grid.major = element_blank(), panel.grid.minor = element_blank(),plot.title=element_text(hjust=.5),
         axis.text.x = element_text(size=12),axis.title.x = element_text(size=14),axis.text.y = element_text(size=12),axis.title.y = element_text(size=14) ) + 
  ylab('Number of donors')  
  
#ggsave(filename='donor_ages.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs', units='in', height=5, width=5, dpi=300)
```







Add the germlayer and the grouped tissue

```{r}

#Grouping tissues into larger groups
sample_tissue_meta = as.character(skew_and_stats_df$tissue)
sample_tissue_meta[sample_tissue_meta == 'Adipose - Subcutaneous' | sample_tissue_meta == 'Adipose - Visceral (Omentum)'] = 'Adipose'
sample_tissue_meta[sample_tissue_meta == 'Artery - Aorta' | sample_tissue_meta == 'Artery - Coronary' | sample_tissue_meta == 'Artery - Tibial' ] = 'Artery'
sample_tissue_meta[sample_tissue_meta == 'Cervix - Ectocervix' | sample_tissue_meta == 'Cervix - Endocervix'] = 'Cervix'
sample_tissue_meta[sample_tissue_meta == 'Colon - Sigmoid' | sample_tissue_meta == 'Colon - Transverse'] = 'Colon'
sample_tissue_meta[sample_tissue_meta == 'Esophagus - Gastroesophageal Junction' | sample_tissue_meta == 'Esophagus - Mucosa' | sample_tissue_meta == 'Esophagus - Muscularis' ] = 'Esophagus'
sample_tissue_meta[sample_tissue_meta == 'Heart - Atrial Appendage' | sample_tissue_meta == 'Heart - Left Ventricle'] = 'Heart'
sample_tissue_meta[sample_tissue_meta == 'Skin - Not Sun Exposed (Suprapubic)' | sample_tissue_meta == 'Skin - Sun Exposed (Lower leg)'] = 'Skin'
sample_tissue_meta[sample_tissue_meta == 'Brain - Amygdala' | sample_tissue_meta == 'Brain - Anterior cingulate cortex (BA24)' | sample_tissue_meta == 'Brain - Caudate (basal ganglia)' | sample_tissue_meta == 'Brain - Cerebellar Hemisphere' | sample_tissue_meta == 'Brain - Cerebellum' | sample_tissue_meta == 'Brain - Cortex' | sample_tissue_meta == 'Brain - Frontal Cortex (BA9)' | sample_tissue_meta == 'Brain - Hippocampus' | sample_tissue_meta == 'Brain - Hypothalamus' | sample_tissue_meta == 'Brain - Nucleus accumbens (basal ganglia)' | sample_tissue_meta == 'Brain - Putamen (basal ganglia)' | sample_tissue_meta == 'Brain - Spinal cord (cervical c-1)' | sample_tissue_meta == 'Brain - Substantia nigra'] = 'Brain'

table(sample_tissue_meta)


skew_and_stats_df$grouped.tissue = sample_tissue_meta

```

```{r}
#Grouping by germ layer
temp = as.character(skew_and_stats_df$grouped.tissue)

temp[temp == 'Adipose' | temp == 'Heart' | temp == 'Whole Blood' | temp == 'Cells - EBV-transformed lymphocytes' | temp == 'Cells - Transformed fibroblasts' | temp == 'Cells - Cultured fibroblasts'| temp == 'Muscle - Skeletal' | temp == 'Adrenal Gland' | temp == 'Artery' | temp == 'Fallopian Tube' | temp == 'Kidney - Cortex' | temp == 'Kidney - Medulla' | temp == 'Spleen' | temp == 'Uterus' | temp == 'Cervix' ] = 'Mesoderm'


temp[temp == 'Brain' | temp == 'Skin' | temp == 'Breast - Mammary Tissue' | temp == 'Nerve - Tibial' | temp == 'Minor Salivary Gland' | temp == 'Pituitary' ] = 'Ectoderm'


temp[temp == 'Lung' | temp == 'Thyroid' | temp == 'Pancreas' | temp == 'Liver' | temp == 'Colon' | temp == 'Small Intestine - Terminal Ileum' | temp == 'Vagina' | temp == 'Ovary' | temp == 'Stomach' | temp == 'Esophagus' | temp == 'Bladder' ] = 'Endoderm'

skew_and_stats_df$germ_layer = temp
head(skew_and_stats_df)

```

Make a small dataframe with just the tissue names and germlayer info

```{r}

tissues = names(table(skew_and_stats_df$tissue))
germ_layer = c()

for( i in 1:length(tissues)){
  germ_layer = c(germ_layer, as.character(skew_and_stats_df$germ_layer[skew_and_stats_df$tissue == tissues[i]][1]))
}

germ_layer_df = data.frame(tissues = tissues, germ_layers = germ_layer)
germ_layer_df

```



```{r}
dim(skew_and_stats_df)

```


Set up some matrices we'll be reusing and the germ layer annotations/colors.

```{r}
donors = names(table(skew_and_stats_df$donor))
length(donors)
#Start with all the tissues, exclude the male tissues
tissues =  names(table(skew_and_stats_df$tissue))
tissues = tissues[! tissues %in% c('Prostate', 'Testis','body_site_s')]
length(tissues)

grouped_tissues = names(table(skew_and_stats_df$grouped.tissue))
grouped_tissues = grouped_tissues[! grouped_tissues %in% c('Prostate', 'Testis','body_site_s')]
length(grouped_tissues)
```


```{r}

names(table(skew_and_stats_df$germ_layer))

```




```{r}
#Get germ layers for the filt_tissues

germ_layers = vector(mode='character', length=length(tissues))

for( i in 1:length(tissues)){
  index = which(skew_and_stats_df$tissue == tissues[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
  
}

germ_layers

```


Save the germ_layer color specifications
```{r}
germ_layer_colors = brewer.pal(n = 4, name = "Spectral")
ecto_color = germ_layer_colors[1]
endo_color = germ_layer_colors[2]
meso_color = germ_layer_colors[4]

```



Start off looking at a table of pairwise donations, number of donors contributing tissues X,Y
Just getting a sense of how much pairwise donations there are to work with
And make a matrix with the donors names that do donate both of those tissues


```{r}
tiss_tiss_donors_df = as.data.frame(matrix(nrow = length(tissues), ncol = length(tissues)))
rownames(tiss_tiss_donors_df) = tissues
colnames(tiss_tiss_donors_df) = tissues

tiss_tiss_donors_names_matrix = matrix(rep(list(), length(tissues)*length(tissues)), nrow=length(tissues), ncol=length(tissues))
rownames(tiss_tiss_donors_names_matrix) = tissues
colnames(tiss_tiss_donors_names_matrix) = tissues

for(i in 1:length(tissues)){
  
  current_tiss = tissues[i]
  
  for(j in 1: length(tissues)){
     
    next_tiss = tissues[j]
    
    donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
    #Just going to ignore the donors that have multiple donations for the same tiss
    duplicates = donors_for_tiss[duplicated(donors_for_tiss)] 
    donors_for_tiss = donors_for_tiss[!donors_for_tiss %in% duplicates]  
    
    donors_for_next_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == next_tiss])
    duplicates = donors_for_next_tiss[duplicated(donors_for_next_tiss)] 
    donors_for_next_tiss = donors_for_next_tiss[!donors_for_next_tiss %in% duplicates]
    
    tiss_tiss_donors_df[i,j] = sum(unique(donors_for_tiss) %in% unique(donors_for_next_tiss))
    tiss_tiss_donors_names_matrix[i,j] = list(unique(donors_for_tiss[donors_for_tiss %in% donors_for_next_tiss]))
  }
}

tiss_tiss_donors_df
tiss_tiss_donors_names_matrix

```


Get a binary matrix too for which donors donate which tissues

```{r}
diagonal = diag(tiss_tiss_donors_names_matrix)


binary_tissue_donor_matrix = matrix(nrow = length(names(diagonal)), ncol = length(donors))
rownames(binary_tissue_donor_matrix) = names(diagonal)
colnames(binary_tissue_donor_matrix) = donors

for(i in 1:length(names(diagonal))){
  
  donors_for_tiss = diagonal[[i]]
  index = colnames(binary_tissue_donor_matrix) %in% donors_for_tiss
  binary_tissue_donor_matrix[i, index] = 1
  binary_tissue_donor_matrix[i, !index] = 0
}


binary_tissue_donor_matrix = binary_tissue_donor_matrix[!rownames(binary_tissue_donor_matrix) %in% c('Testis','Prostate','body_site_s'), ]


```



Heatmap showing the donor tissue intersections
Add a germ layer annotation
Add marginal boxplot annotations to get the number of donors per tissue and number of tissues per donor




```{r}
row_index = !rownames(binary_tissue_donor_matrix) %in% c('Cells - EBV-transformed lymphocytes','Cells - Cultured fibroblasts','Cells - Transformed fibroblasts')
binary_tissue_donor_matrix = binary_tissue_donor_matrix[row_index, ]

```

```{r}
germ_layers = germ_layers[row_index]

```


```{r}

#Get the number of tissues per donor
num_tiss_per_donor = colSums(binary_tissue_donor_matrix)
max_tiss_per_donor = max(num_tiss_per_donor)
#Get the number of donors per tissue
num_donors_per_tissue = rowSums(binary_tissue_donor_matrix)
max_donor_per_tiss = max(num_donors_per_tissue)

#Germ layer annotation and number of donors per tissue
row_annot = HeatmapAnnotation(germ_layer = germ_layers, donors_per_tiss = anno_barplot(num_donors_per_tissue, gp = gpar(fill = rep(1,nrow(binary_tissue_donor_matrix)))),
                                     which = 'row',
                                     col = list(germ_layer=c('Ectoderm'=ecto_color, 'Mesoderm'=meso_color, 'Endoderm' = endo_color)), 
                                     gp = gpar(col = "black"), show_annotation_name = c(FALSE,TRUE),
                                     annotation_legend_param = list(title = 'Germ layer'))
names(row_annot) = c('Germ layers','Donors per tissue')

#number of tissues per donor column point annotation
tissues_per_donor_annot = HeatmapAnnotation(tissues_per_donor = anno_barplot(num_tiss_per_donor),which = 'column')
names(tissues_per_donor_annot) = c('Tissues per donor')

#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/binary_donors_per_tissue_heatmap.pdf', height = 8, width =8, useDingbats = FALSE)

#Colors for the heatmap
colors = structure(c('white','black'), names = c("0", "1"))
Heatmap(binary_tissue_donor_matrix, col=colors, column_title = 'Donor tissue contributions', use_raster = TRUE,
        clustering_method_rows = "ward.D2",clustering_method_columns = "ward.D2", 
        show_row_dend=FALSE, show_column_dend = FALSE,
        row_names_gp = gpar(fontsize = 8), show_column_names = FALSE,
        right_annotation = row_annot,
        top_annotation = tissues_per_donor_annot,
        heatmap_legend_param = list( title = "Donated", at = c(0, 1), labels = c("FALSE", 'TRUE')))
#dev.off()



Heatmap(binary_tissue_donor_matrix, col=colors, column_title = 'Donor tissue contributions', use_raster = TRUE,
        clustering_method_rows = "ward.D2",clustering_method_columns = "ward.D2", 
        show_row_dend=FALSE, show_column_dend = FALSE, 
        row_names_gp = gpar(fontsize = 8), show_column_names = FALSE,
        right_annotation = row_annot,
        top_annotation = tissues_per_donor_annot,
        heatmap_legend_param = list( title = "Donated", at = c(0, 1), labels = c("FALSE", 'TRUE')))
```


```{r}

table(colSums(binary_tissue_donor_matrix) > 0)



```



Get boxplots per donor of tissue skews, ordered by mean donor skew

```{r}
culture_index = !skew_and_stats_df$tissue %in% c('Cells - EBV-transformed lymphocytes','Cells - Cultured fibroblasts','Cells - Transformed fibroblasts')

subset_skew_and_stats = skew_and_stats_df[ culture_index , c('skew','donor')]
donor_mean_skews = aggregate(subset_skew_and_stats[ ,1], list(subset_skew_and_stats$donor), median)
donor_mean_skews = donor_mean_skews[order(donor_mean_skews$x), ]
ordered_donor_names = as.character(donor_mean_skews$Group.1)

subset_skew_and_stats$donor = factor(subset_skew_and_stats$donor, levels = ordered_donor_names)
subset_skew_and_stats
```



```{r}
first_100 = levels(subset_skew_and_stats$donor)[1:100]
first_index = subset_skew_and_stats$donor %in% first_100

ggplot(subset_skew_and_stats[first_index ,], aes(x = donor, y = skew)) + geom_boxplot(outlier.size = .5) +
  ylim(.5,1) + ylab('Tissue skews') + xlab('Donors') +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_blank(),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12), 
        axis.ticks.x = element_blank())
ggsave(filename = 'first_100_order_donors_skew_boxplots.pdf', device = 'pdf',
       path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/', height = 6, width = 10, useDingbats = F)


second_100 = levels(subset_skew_and_stats$donor)[101:200]
second_index = subset_skew_and_stats$donor %in% second_100

ggplot(subset_skew_and_stats[second_index ,], aes(x = donor, y = skew)) + geom_boxplot(outlier.size = .5) +
  ylim(.5,1) + ylab('Tissue skews') + xlab('Donors') +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_blank(),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12), 
        axis.ticks.x = element_blank())

ggsave(filename = 'second_100_order_donors_skew_boxplots.pdf', device = 'pdf',
       path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/', height = 6, width = 10, useDingbats = F)

third_100 = levels(subset_skew_and_stats$donor)[201:length(subset_skew_and_stats$donor)]
third_index = subset_skew_and_stats$donor %in% third_100

ggplot(subset_skew_and_stats[third_index ,], aes(x = donor, y = skew)) + geom_boxplot(outlier.size = .5) +
  ylim(.5,1) + ylab('Tissue skews') + xlab('Donors') +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_blank(),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12), 
        axis.ticks.x = element_blank())
ggsave(filename = 'third_100_order_donors_skew_boxplots.pdf', device = 'pdf',
       path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/', height = 6, width = 10, useDingbats = F)

```


```{r}
mean(colSums(binary_tissue_donor_matrix))

rowSums(binary_tissue_donor_matrix)


sum(colSums(binary_tissue_donor_matrix))
```






###########################################
Heat map showing donors by tissues, to actually show the skews of all the tissues. Main point of the paper and need to show it
Also calculate how many double samples from the same donor and tissue there are, plot a distribution of differences from the same tissue
##########################################



```{r}
dup_tissues = c()
diff_skew_dup_tiss = c()
tissue_label = c()
for(i in 1:length(donors)){
  
  #Get the skews for that donor
  donor_data = skew_and_stats_df[skew_and_stats_df$donor == donors[i], ]
  #See of there are any duplicate tissue samples
  dup = duplicated(as.character(donor_data$tissue))
  donor_dup = as.character(donor_data$tissue[dup])
  dup_tissues = c(dup_tissues, donor_dup)
  
  #Get the absolute differences in skew between the duplicated tissues
  if(length(donor_dup) == 0){next}
  for(j in 1:length(donor_dup)){
    dup_skews = donor_data$skew[donor_data$tissue == donor_dup[j]]
    #Get the pairwise differences between those skews, dist() will calculate the euclidean distance
    differences = c(dist(dup_skews))
    tiss_label = rep(donor_dup[j], length(differences))
    diff_skew_dup_tiss = c(diff_skew_dup_tiss, differences)
    tissue_label = c(tissue_label,tiss_label ) 
  }

}

```

```{r}
#pdf('/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/num_tiss_replicates_barplot.pdf', height =8, width = 8, useDingbats = FALSE )
par(mar = c(18,4,4,1))
barplot(table(tissue_label), names.arg = names(table(tissue_label)), srt = 90, ylab = 'Number of tissue replicates', 
        horiz = F, las = 2, cex.names = 1)
#dev.off()
```



Most tissue replicates have very similar skews, just average together the skew estimates for those with replicate samples

```{r}
#pdf('/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/skew_diff_btw_tiss_reps_hist.pdf', height =5, width = 5, useDingbats = FALSE )
hist(diff_skew_dup_tiss, breaks = 32, main = 'Replicate tissue samples have comparable skews', xlab = 'Skew difference between replicates')
#dev.off()
```

```{r}
mean(diff_skew_dup_tiss)

sd(diff_skew_dup_tiss)
```

box plots of replicate deviation ordered by tissue
```{r}
rep_skew_dev_df = data.frame(tissue = tissue_label, skew_dev = diff_skew_dup_tiss)

#Order by median deviance
median_rep_devs = aggregate(rep_skew_dev_df[ ,'skew_dev'], list(rep_skew_dev_df$tissue), mean)
median_rep_devs = median_rep_devs[order(median_rep_devs$x), ]
ordered_tissues = as.character(median_rep_devs$Group.1)

rep_skew_dev_df$tissue = factor(rep_skew_dev_df$tissue, levels = ordered_tissues)
```


```{r}
ggplot(rep_skew_dev_df, aes(x = tissue, y = skew_dev)) + geom_boxplot() +
  ylab('Deviation in skew between tissue replicates') +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_text(angle = 90, hjust = .95, vjust = .2),
        axis.title.y = element_text(size=12), axis.title.x = element_blank(), 
        axis.ticks.x = element_blank())

ggsave(filename = 'ordered_tissue_replicate_skew_dev_boxplots.pdf', device = 'pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/', 
       width = 8, height = 8, useDingbats = F)
```






Build a donor X tissue matrix with skews, averaging the replicates

```{r}
donor_tiss_skew_matrix = matrix(nrow = length(donors), ncol = length(tissues))
rownames(donor_tiss_skew_matrix) = donors                                
colnames(donor_tiss_skew_matrix) = tissues


for(i in 1:length(donors)){
  
  row_index = rownames(donor_tiss_skew_matrix) == donors[i]
  
  skews = skew_and_stats_df$skew[skew_and_stats_df$donor == donors[i]]
  donor_tiss = as.character(skew_and_stats_df$tissue[skew_and_stats_df$donor == donors[i]])
  dup_tissues = unique(donor_tiss[duplicated(donor_tiss)])
  if(length(dup_tissues) == 0){
    #Reorder to match the order of the matrix
    col_index = colnames(donor_tiss_skew_matrix) %in% donor_tiss
    present = colnames(donor_tiss_skew_matrix)[col_index]
    index = match(present, donor_tiss)
    donor_tiss_skew_matrix[row_index, col_index] = skews[index]
  }else{

    #Replace all skew values with the mean skew
    for(j in 1:length(dup_tissues)){
      mean_skew = mean(skews[donor_tiss == dup_tissues[j]])
      skews[donor_tiss == dup_tissues[j]] = mean_skew
    }
    
    #Get rid of the duplicates
    donor_tiss = donor_tiss[!duplicated(donor_tiss)]
    skews = skews[!duplicated(donor_tiss)]
    
    #Reorder to match the order of the matrix
    col_index = colnames(donor_tiss_skew_matrix) %in% donor_tiss
    present = colnames(donor_tiss_skew_matrix)[col_index]
    index = match(present, donor_tiss)
    donor_tiss_skew_matrix[row_index, col_index] = skews[index]
  }
  
}

donor_tiss_skew_matrix = donor_tiss_skew_matrix[ ,!colnames(donor_tiss_skew_matrix) %in% 
                                                   c('Prostate','Testis', 'Kidney - Medulla', "Cells - EBV-transformed lymphocytes", "Cells - Cultured fibroblasts", "Cells - Transformed fibroblasts")]

```



Reorder donors by their mean skew
```{r}
mean_skews = rowMeans(donor_tiss_skew_matrix, na.rm = TRUE)
mean_skews = mean_skews[order(mean_skews, decreasing = FALSE)]
index = match(names(mean_skews), rownames(donor_tiss_skew_matrix))
donor_tiss_skew_matrix = donor_tiss_skew_matrix[index, ]
table(rownames(donor_tiss_skew_matrix) == names(mean_skews))
```

Reorder tissues by germ layer
```{r}
mat_cols = colnames(donor_tiss_skew_matrix)
#Get germ layers for the filt_tissues

germ_layers = vector(mode='character', length=length(mat_cols))

for( i in 1:length(mat_cols)){
  index = which(skew_and_stats_df$tissue == mat_cols[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
  
}

reorder_index = order(germ_layers)
germ_layers = germ_layers[reorder_index]
donor_tiss_skew_matrix = donor_tiss_skew_matrix[ ,reorder_index ]

```

Exclude donors with only 1 tissue
```{r}
keep_index = rowSums(!is.na(donor_tiss_skew_matrix)) > 1
donor_tiss_skew_matrix = donor_tiss_skew_matrix[keep_index, ]
mean_skews = mean_skews[keep_index]
```


```{r}
#Add an annotation looking at the variation in skew per donor, if higher skewed individuals have really lower variance than others, maybe suggests inactivation bias?

donor_skew_vars = matrixStats::rowVars(donor_tiss_skew_matrix, na.rm = T)

```

Exclude low sampled tissues just for visualization
```{r}
keep_index = colSums(!is.na(donor_tiss_skew_matrix)) > 10
donor_tiss_skew_matrix = donor_tiss_skew_matrix[, keep_index]
germ_layers = germ_layers[keep_index]
```


Turn Nas to 0 for clustering
```{r}
donor_tiss_skew_matrix[is.na(donor_tiss_skew_matrix)] = .499
```



First look at all tissues

```{r}
library(viridis)

```


```{r}
#Germ layer annotation and number of donors per tissue
col_annot = HeatmapAnnotation(germ_layer = germ_layers,
                                     which = 'column',
                                     col = list(germ_layer=c('Ectoderm'=ecto_color, 'Mesoderm'=meso_color, 'Endoderm' = endo_color)), 
                                     gp = gpar(col = "black"), show_annotation_name = TRUE,
                                     annotation_legend_param = list(title = 'Germ layer'))

pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/donors_by_tissues_skew_heatmap.pdf', height = 7, width = 6, useDingbats = FALSE)
max_skew = max(skew_and_stats_df$skew)
skew_scale = 500:(max_skew*1000)/1000

col_fun = colorRamp2(c(.499, skew_scale), 
                    c('black',viridis(length(skew_scale), option = 'plasma')))


#col_fun_var = colorRamp2(c(min(donor_skew_vars, na.rm = T), max(donor_skew_vars, na.rm = T)), c('white','purple4'))
#ha_1 = rowAnnotation(mean_donor_skew = mean_skews,
#                   donor_vars = donor_skew_vars,
#                   col = list(mean_donor_skew = col_fun, donor_vars = col_fun_var), 
#                   show_legend = c(FALSE, TRUE))
ha_2 = rowAnnotation(mean_donor_skew = mean_skews,
                   col = list(mean_donor_skew = col_fun), 
                   show_legend = FALSE)

Heatmap(donor_tiss_skew_matrix, show_row_names = FALSE, show_column_names = TRUE, name = 'All tissues', show_column_dend = FALSE,
        col = col_fun, use_raster = TRUE, row_title = 'Donors', 
        cluster_rows = FALSE,cluster_columns = FALSE, 
        right_annotation = ha_2,
        top_annotation = col_annot,
        column_names_gp = grid::gpar(fontsize = 8),
        heatmap_legend_param = list(title = 'Skew'))
dev.off()

Heatmap(donor_tiss_skew_matrix, show_row_names = FALSE, show_column_names = TRUE, name = 'All tissues', show_column_dend = FALSE,
        col = col_fun, use_raster = TRUE, row_title = 'Donors', 
        cluster_rows = FALSE,cluster_columns = FALSE, 
        right_annotation = ha_2,
        top_annotation = col_annot,
        column_names_gp = grid::gpar(fontsize = 8),
        heatmap_legend_param = list(title = 'Skew'))

```
save the donors that have a mean skew >= .75, for later

```{r}

skewed_donors = names(mean_skews)[(mean_skews>=.75)]
skewed_donors 
```


Filter out tissues with only a few samples and donors with only a few samples

```{r}
pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/donors_by_tissues_skew_heatmap_filtered_tiss.pdf', height = 7, width = 6, useDingbats = FALSE)
temp_mat = donor_tiss_skew_matrix 
row_index = rowSums(temp_mat > .499) >= (.2* dim(temp_mat)[2])
col_index = colSums(temp_mat > .499) >= (.2* dim(temp_mat)[1])
temp_mat = temp_mat[row_index , col_index]


#col_fun = colorRamp2(c(.499, .5,.75, max_skew, 1), c('white','purple','orange', "red3", 'red3'))
ha = rowAnnotation(mean_donor_skew = mean_skews[row_index], col = list(mean_donor_skew = col_fun), show_legend = FALSE)
col_annot = HeatmapAnnotation(germ_layer = germ_layers[col_index],
                                     which = 'column',
                                     col = list(germ_layer=c('Ectoderm'=ecto_color, 'Mesoderm'=meso_color, 'Endoderm' = endo_color)), 
                                     gp = gpar(col = "black"), show_annotation_name = TRUE,
                                     annotation_legend_param = list(title = 'Germ layer'))

Heatmap(temp_mat, show_row_names = FALSE, show_column_names = TRUE, name = 'Filtered tissues', show_column_dend = FALSE,
        col = col_fun, use_raster = TRUE, row_title = 'Donors', 
        cluster_rows = FALSE, cluster_columns = FALSE, 
        right_annotation = ha,
        top_annotation = col_annot,
        column_names_gp = grid::gpar(fontsize = 8),
        heatmap_legend_param = list(title = 'Skew'))
dev.off()

Heatmap(temp_mat, show_row_names = FALSE, show_column_names = TRUE, name = 'Filtered tissues', show_column_dend = FALSE,
        col = col_fun, use_raster = TRUE, row_title = 'Donors', 
        cluster_rows = FALSE, cluster_columns = FALSE, 
        right_annotation = ha,
        top_annotation = col_annot,
        column_names_gp = grid::gpar(fontsize = 8),
        heatmap_legend_param = list(title = 'Skew'))


```


Only plot germ layers

```{r}
temp_mat = donor_tiss_skew_matrix 
filt_tiss = as.character(germ_layer_df$tissues[germ_layer_df$germ_layers == 'Ectoderm'])
temp_mat = temp_mat[, colnames(temp_mat) %in% filt_tiss]

max_skew = max(temp_mat)
skew_scale = 500:(max_skew*1000)/1000
col_fun = colorRamp2(c(.499, skew_scale), 
                    c('black',viridis(length(skew_scale), option = 'plasma')))

temp_temp = temp_mat
row_index = rowSums(temp_temp > .499) > 0
temp_temp = temp_temp[row_index, ]

ha = rowAnnotation(mean_donor_skew = mean_skews[row_index], col = list(mean_donor_skew = col_fun), show_legend = FALSE)
pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/donors_by_tissues_skew_heatmap_ectoderm.pdf', height = 7, width = 6, useDingbats = FALSE)
Heatmap(temp_temp, show_row_names = FALSE, show_column_names = FALSE, name = 'Ectoderm', 
        col = col_fun, use_raster = TRUE, 
        cluster_rows = FALSE,clustering_method_columns = "ward.D", 
        right_annotation = ha)
dev.off()
Heatmap(temp_temp, show_row_names = FALSE, show_column_names = FALSE, name = 'Ectoderm', 
        col = col_fun, use_raster = TRUE, 
        cluster_rows = FALSE,clustering_method_columns = "ward.D", 
        right_annotation = ha)
```





```{r}
temp_mat = donor_tiss_skew_matrix 
filt_tiss = as.character(germ_layer_df$tissues[germ_layer_df$germ_layers == 'Endoderm'])
temp_mat = temp_mat[, colnames(temp_mat) %in% filt_tiss]

max_skew = max(temp_mat)
skew_scale = 500:(max_skew*1000)/1000
col_fun = colorRamp2(c(.499, skew_scale), 
                    c('black',viridis(length(skew_scale), option = 'plasma')))

temp_temp = temp_mat
row_index = rowSums(temp_temp > .499) > 0
temp_temp = temp_temp[row_index, ]

ha = rowAnnotation(mean_donor_skew = mean_skews[row_index], col = list(mean_donor_skew = col_fun), show_legend = FALSE)
pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/donors_by_tissues_skew_heatmap_endoderm.pdf', height = 7, width = 6, useDingbats = FALSE)
Heatmap(temp_temp, show_row_names = FALSE, show_column_names = FALSE, name = 'Ectoderm', 
        col = col_fun, use_raster = TRUE, 
        cluster_rows = FALSE,clustering_method_columns = "ward.D", 
        right_annotation = ha)
dev.off()
Heatmap(temp_temp, show_row_names = FALSE, show_column_names = FALSE, name = 'Ectoderm', 
        col = col_fun, use_raster = TRUE, 
        cluster_rows = FALSE,clustering_method_columns = "ward.D", 
        right_annotation = ha)

```
```{r}
temp_mat = donor_tiss_skew_matrix 
filt_tiss = as.character(germ_layer_df$tissues[germ_layer_df$germ_layers == 'Mesoderm'])
temp_mat = temp_mat[, colnames(temp_mat) %in% filt_tiss]

max_skew = max(temp_mat)
skew_scale = 500:(max_skew*1000)/1000
col_fun = colorRamp2(c(.499, skew_scale), 
                    c('black',viridis(length(skew_scale), option = 'plasma')))

temp_temp = temp_mat
row_index = rowSums(temp_temp > .499) > 0
temp_temp = temp_temp[row_index, ]

ha = rowAnnotation(mean_donor_skew = mean_skews[row_index], col = list(mean_donor_skew = col_fun), show_legend = FALSE)
pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/summary_graphs/donors_by_tissues_skew_heatmap_mesoderm.pdf', height = 7, width = 6, useDingbats = FALSE)
Heatmap(temp_temp, show_row_names = FALSE, show_column_names = FALSE, name = 'Ectoderm', 
        col = col_fun, use_raster = TRUE, 
        cluster_rows = FALSE,clustering_method_columns = "ward.D", 
        right_annotation = ha)
dev.off()
Heatmap(temp_temp, show_row_names = FALSE, show_column_names = FALSE, name = 'Ectoderm', 
        col = col_fun, use_raster = TRUE, 
        cluster_rows = FALSE,clustering_method_columns = "ward.D", 
        right_annotation = ha)

```


################################################
Tissue cell pop estimations
################################################


Trying out fitting the tails of the tissue distributions to get the tissue stem pool size estimate
Going with sampling from a binomial distribution to generate theoretical cdfs to compare to the empirical cdfs. Finding the N that minimizes the residuals between the cdfs for values
< .4 and > .6, from the ENTEx phased analysis these are the skews we are confident in

```{r}

# Functions to fold and unfold ratios 
folded <- function(x) { apply( cbind(x,1-x), 1, max)} 
unfold <- function(x) { c(x,1-x) } 
```



```{r}

#Tissues with at leasst X donors
filt_tissues = names(table(skew_and_stats_df$tissue)[table(skew_and_stats_df$tissue) >= 10])


for(tissue in filt_tissues){
  
  skews = unfold(skew_and_stats_df$skew[skew_and_stats_df$tissue == tissue])
  hist(skews, main = sprintf('XCI skews: %s', tissue), xlim = c(0,1), breaks = 16)
  #hist(skews, main = sprintf('XCI skews: %s', tissue), xlim = c(0,1), breaks = 16, prob=TRUE)
  #lines(density(skews))
}


```




The below is fine, but maybe try playing around with the density function. Does it make sense? Binomial is discrete, doesn't have a desntiy, it has a mass??



Checking to get the pure theoretical normal percentiles
```{r}
tissue = 'Adipose - Subcutaneous'

n = 16
p = .5
norm_sd = sqrt( (p*(1-p)) / n )

percentiles = seq(.01,.99,.01)

#Get the empirical skews for that tissue
empir_skews = unfold(skew_and_stats_df$skew[skew_and_stats_df$tissue == tissue])
num_skews = length(empir_skews)
#Get the quantiles for these skews
empir_quants = quantile(empir_skews, probs = percentiles, type=1)
#Get the theoretical percentiles from the normal distribution
norm_percentiles = qnorm(percentiles, mean=p, sd = norm_sd)
bin_percentiles = qbinom(percentiles, size=n, prob=p) / n

plot(empir_quants, percentiles, xlim = c(0,max(percentiles)), main = 'Normal percentiles')
points(norm_percentiles, percentiles, type = 'o', col='red')


plot(empir_quants, percentiles, xlim = c(0,max(percentiles)), main = 'Binomial percentiles')
points(bin_percentiles, percentiles, type = 'o', col='red')


quant_filt = empir_quants <= .4 | empir_quants >= .6 
#Calculating a general coefficient of determination

error = empir_quants[quant_filt] - norm_percentiles[quant_filt] 
total_sum_squares = sum( (empir_quants[quant_filt]  - mean(empir_quants[quant_filt] ))^2 )
sum_squares_error = sum(error^2)
coef_determ = 1 - (sum_squares_error / total_sum_squares)
coef_determ


error = empir_quants[quant_filt] - bin_percentiles[quant_filt] 
total_sum_squares = sum( (empir_quants[quant_filt]  - mean(empir_quants[quant_filt] ))^2 )
sum_squares_error = sum(error^2)
coef_determ = 1 - (sum_squares_error / total_sum_squares)
coef_determ

```



```{r}

for(tissue in c('Adipose - Subcutaneous', 'Whole Blood')){

  p = .5
  n = 2:50
  percentiles = seq(.01, .99, 0.01)

  bin_error_vector = vector(mode = 'numeric', length = length(n))
  norm_error_vector = vector(mode = 'numeric', length = length(n))
  
  #Get the empirical skews for that tissue
  empir_skews = unfold(skew_and_stats_df$skew[skew_and_stats_df$tissue == tissue])
  num_skews = length(empir_skews)
  #Get the quantiles for these skews
  empir_quants = quantile(empir_skews, probs = percentiles, type=1)
  #Only going to be fitting over the confident skews
  quant_filt = empir_quants <= .4 | empir_quants >= .6
  
  
  
  for(i in 1:length(n)){
    
    #Get the normal and binomial percentiles for the n parameter
    norm_sd = sqrt( (p*(1-p)) / n[i] )
    norm_percentiles = qnorm(percentiles, mean=p, sd = norm_sd)
    bin_percentiles = qbinom(percentiles, size=n[i], prob=p) / n[i]
    
    if(n[i] %in% c(4,8,16,32)){

      plot(empir_quants, percentiles, xlim = c(0,1), xlab='XCI skews', ylab = 'Percentiles', 
           main=sprintf('Empirical vs theoretical percentiles, Binomial N = %g', n[i]))
      points(bin_percentiles, percentiles, type = 'o', col='red')
      legend(x = 'bottomright', legend=c('Empirical','Theoretical'), col = c('black','red'), pch=15)
      
      plot(empir_quants, percentiles, xlim = c(0,1), xlab='XCI skews', ylab = 'Percentiles', 
           main=sprintf('Empirical vs theoretical percentiles, Normal N = %g', n[i]))
      points(norm_percentiles, percentiles, type = 'o', col='red')
      legend(x = 'bottomright', legend=c('Empirical','Theoretical'), col = c('black','red'), pch=15)       
        
      }    
    
    bin_error_vector[i] = sum((empir_quants[quant_filt] - bin_percentiles[quant_filt])^2)
    norm_error_vector[i] = sum((empir_quants[quant_filt] - norm_percentiles[quant_filt])^2)
    

  }

  plot(n, bin_error_vector, ylab = 'Sum sq_error empirical - binomial percentiles', xlab = 'N', 
       main = sprintf('%s Num donors= %g', tissue, num_skews/2), xlim = c(0,n[length(n)]))
  
  plot(n, norm_error_vector, ylab = 'Sum sq_error empirical - normal percentiles', xlab = 'N', 
     main = sprintf('%s Num donors= %g', tissue, num_skews/2), xlim = c(0,n[length(n)]))

}

```




```{r}

binomial_fitted_est_n = vector(mode='numeric', length = length(filt_tissues))          
binomial_fitted_error = vector(mode='numeric', length = length(filt_tissues))
binomial_variance_explained = vector(mode='numeric', length = length(filt_tissues))

normal_fitted_est_n = vector(mode='numeric', length = length(filt_tissues))
normal_fitted_error = vector(mode='numeric', length = length(filt_tissues))
normal_variance_explained = vector(mode='numeric', length = length(filt_tissues))
num_samples_per_tiss= vector(mode='numeric', length = length(filt_tissues))

for(k in 1:length(filt_tissues)){

  p = .5
  n = 2:50
  percentiles = seq(.01, .99, 0.01)

  bin_error_vector = vector(mode = 'numeric', length = length(n))
  norm_error_vector = vector(mode = 'numeric', length = length(n))
  
  #Get the empirical skews for that tissue
  empir_skews = unfold(skew_and_stats_df$skew[skew_and_stats_df$tissue == filt_tissues[k]])
  num_skews = length(empir_skews)
  num_samples_per_tiss[k] = num_skews / 2 #Number of unfolded skews is twice that of the sample size for that tissue
  #Get the quantiles for these skews
  empir_quants = quantile(empir_skews, probs = percentiles, type=1)
  #Only going to be fitting over the confident skews
  quant_filt = empir_quants <= .4 | empir_quants >= .6
    
  for(i in 1:length(n)){
    #Get the normal and binomial percentiles for the n parameter
    norm_sd = sqrt( (p*(1-p)) / n[i] )
    norm_percentiles = qnorm(percentiles, mean=p, sd = norm_sd)
    bin_percentiles = qbinom(percentiles, size=n[i], prob=p) / n[i]
    #And the error between the empirical and theoretical percentiles
    bin_error_vector[i] = sum((empir_quants[quant_filt] - bin_percentiles[quant_filt])^2)
    norm_error_vector[i] = sum((empir_quants[quant_filt] - norm_percentiles[quant_filt])^2)
  }

  #Visualize the error over the different n parameters
  plot(n, bin_error_vector, ylab = 'Sum sq_error empirical - binomial percentiles', xlab = 'N', 
       main = sprintf('%s Num donors= %g', filt_tissues[k], num_skews/2), xlim = c(0,n[length(n)]))
  plot(n, norm_error_vector, ylab = 'Sum sq_error empirical - normal percentiles', xlab = 'N', 
     main = sprintf('%s Num donors= %g', filt_tissues[k], num_skews/2), xlim = c(0,n[length(n)]))

  #Save the best fitting n
  binomial_fitted_est_n[k] = n[which.min(bin_error_vector)]
  normal_fitted_est_n[k] = n[which.min(norm_error_vector)]
  
  #Save the variance explained by the best model
  norm_sd = sqrt( (p*(1-p)) / normal_fitted_est_n[k] )
  norm_percentiles = qnorm(percentiles, mean=p, sd = norm_sd)
  #Assess variance explained across whole distribution, rather than the confident skews we fitted to
  error = empir_quants - norm_percentiles 
  total_sum_squares = sum( (empir_quants  - mean(empir_quants))^2 )
  sum_squares_error = sum(error^2)
  normal_variance_explained[k] = 1 - (sum_squares_error / total_sum_squares)
  
  #And for the binomial fits
  bin_percentiles = qbinom(percentiles, size=binomial_fitted_est_n[k], prob=p) / binomial_fitted_est_n[k]
  error = empir_quants - bin_percentiles 
  total_sum_squares = sum( (empir_quants  - mean(empir_quants))^2 )
  sum_squares_error = sum(error^2)
  binomial_variance_explained[k] = 1 - (sum_squares_error / total_sum_squares)  
}

```

```{r}
par(mar = c(12,4,2,2))
index = order(binomial_variance_explained)
barplot(binomial_variance_explained[index], names = filt_tissues[index], las = 2, cex.names = .65, ylab = '% variance explained by best fitted binomial', 
        main = 'XCI skew variance explained by Binomial model', ylim = c(0,1), cex.lab = .75)

index = order(normal_variance_explained)
barplot(normal_variance_explained[index], names = filt_tissues[index], las = 2, cex.names = .65, ylab = '% variance explained by best fitted normal', 
        main = 'XCI skew variance explained by Normal model', ylim = c(0,1), cex.lab = .75)


plot(NULL, xlim = c(0, length(filt_tissues)), ylim = c(.80,1), xaxt = 'n', ylab = '% variance explained', xlab = '',
      main = 'Binomial model can explain most of the variance in XCI skews across tissues', cex.main = 1)
points(binomial_variance_explained[index], pch = 19)
axis(1, at = 1:length(filt_tissues), labels = filt_tissues[index], las=3, cex.axis= .70)

plot(NULL, xlim = c(0, length(filt_tissues)), ylim = c(.80,1), xaxt = 'n', ylab = '% variance explained', xlab = '',
      main = 'Normal model can explain most of the variance in XCI skews across tissues', cex.main = 1)
points(normal_variance_explained[index], pch = 19)
axis(1, at = 1:length(filt_tissues), labels = filt_tissues[index], las=3, cex.axis= .70)


plot(num_samples_per_tiss,normal_variance_explained, xlab = 'Number of samples per tissue', ylab = '% skew variance explained by best fitted normal', 
     main = 'Normal model can explain most of the variance in XCI skews across tissues')




```




```{r}
#Get germ layers for the filt_tissues

germ_layers = vector(mode='character', length=length(filt_tissues))

for( i in 1:length(filt_tissues)){
  index = which(skew_and_stats_df$tissue == filt_tissues[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
  
}

germ_layers

```




Also run with Sara's original mmethod for estimating, using the var(unfolded(skews)) into N = p* q / var
```{r}

cell_pop_est_vector = vector(mode = 'numeric', length = length(filt_tissues))

for( i in 1: length(filt_tissues)){
  
  skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == filt_tissues[i]]
  skew_var = var(unfold(skews))
  estimated_n = .25/skew_var
  cell_pop_est_vector[i] = estimated_n 

}


est_stem_cell_pop_df = data.frame(binomial_est_n = binomial_fitted_est_n, normal_est_n = normal_fitted_est_n, var_est_n = cell_pop_est_vector, 
                                  tissues = filt_tissues, num_samples = num_samples_per_tiss, germ_layers = germ_layers)




rownames(est_stem_cell_pop_df) = filt_tissues
est_stem_cell_pop_df = arrange(est_stem_cell_pop_df, germ_layers)

est_stem_cell_pop_df$tissues = factor(est_stem_cell_pop_df$tissues, levels = est_stem_cell_pop_df$tissues)
est_stem_cell_pop_df

```



```{r}
par(pty='s')


p_1= ggplot(est_stem_cell_pop_df, aes(x=tissues, y=binomial_est_n, color=germ_layers, size=num_samples)) + geom_point() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.y = element_text(face="bold", size=12)) + 
  ggtitle('Tissue Stem cell population size estimates - fitting binomial percentiles') + ylab('Estimated Num of cells') +
  xlab('Tissues') + ylim(0,25) + 
  scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color)) + 
  geom_hline(yintercept=16, linetype='dashed', color='blue') +
  geom_hline(yintercept=8, linetype='dashed', color='blue')

ggsave(filename='max_fitted_binomial_stem_cell_pop_ests.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/', units='in', height=5, width=7, dpi=300, useDingbats=FALSE)



p_1= ggplot(est_stem_cell_pop_df, aes(x=tissues, y=normal_est_n, color=germ_layers, size=num_samples)) + geom_point() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.y = element_text(face="bold", size=12)) + 
  ggtitle('Tissue Stem cell population size estimates - fitting normal percentiles') + ylab('Estimated Num of cells') +
  xlab('Tissues')+ ylim(0,25)
p_1+ scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color)) + 
  geom_hline(yintercept=16, linetype='dashed', color='blue') +
  geom_hline(yintercept=8, linetype='dashed', color='blue')

ggsave(filename='max_fitted_normal_stem_cell_pop_ests.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/', units='in', height=5, width=7, dpi=300, useDingbats=FALSE)


p_1= ggplot(est_stem_cell_pop_df, aes(x=tissues, y=var_est_n, color=germ_layers, size=num_samples)) + geom_point() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.y = element_text(face="bold", size=12)) + 
  ggtitle('Tissue Stem cell population size estimates - using variance estimate') + ylab('Estimated Num of cells') +
  xlab('Tissues')+ ylim(0,25)
p_1+ scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color)) + 
  geom_hline(yintercept=16, linetype='dashed', color='blue') +
  geom_hline(yintercept=8, linetype='dashed', color='blue')

ggsave(filename='fitted_from_var_stem_cell_pop_ests.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/', units='in', height=5, width=7, dpi=300, useDingbats=FALSE)

pdf('/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/comparing_the_two_methods.pdf',height = 4, width = 5)

pear_cor = cor(est_stem_cell_pop_df$binomial_est_n, est_stem_cell_pop_df$normal_est_n, method='pearson')
title = sprintf('Comparing stem cell size estimates R: %f', pear_cor)
plot(est_stem_cell_pop_df$binomial_est_n, est_stem_cell_pop_df$normal_est_n, pch=19, xlab='Estimated cell num from fitting a binomial', 
     ylab='Estimated cell num from fitting a normal', main=title)
abline(a=0, b=1, col='red')

pear_cor = cor(est_stem_cell_pop_df$binomial_est_n, est_stem_cell_pop_df$var_est_n, method='pearson')
title = sprintf('Comparing stem cell size estimates R: %f', pear_cor)
plot(est_stem_cell_pop_df$binomial_est_n, est_stem_cell_pop_df$var_est_n, pch=19, xlab='Estimated cell num from fitting a binomial', ylab='Using variance estimates', main=title)
abline(a=0, b=1, col='red')

pear_cor = cor(est_stem_cell_pop_df$normal_est_n, est_stem_cell_pop_df$var_est_n, method='pearson')
title = sprintf('Comparing stem cell size estimates R: %f', pear_cor)
plot(est_stem_cell_pop_df$normal_est_n, est_stem_cell_pop_df$var_est_n, pch=19, xlab='Fitting Normal percentiles', ylab='Using variance estimates', main=title)
abline(a=0, b=1, col='red')

dev.off()


```



Bootstrap the normal fitting method to get CIs. Trying the basic bootstrap with percentile CIs first

```{r}
#Function for getting a single N estimate fitting a normal distribution
fit_normal_to_skew_dist = function(skews){
  
  p = .5
  n = 2:75
  percentiles = seq(.01, .99, 0.01)
  norm_error_vector = vector(mode = 'numeric', length = length(n))
  #Get the quantiles for these skews
  empir_quants = quantile(skews, probs = percentiles, type=1)
  #Only going to be fitting over the confident skews
  quant_filt = empir_quants <= .4 | empir_quants >= .6
    
  for(i in 1:length(n)){
    #Get the normal and binomial percentiles for the n parameter
    norm_sd = sqrt( (p*(1-p)) / n[i] )
    norm_percentiles = qnorm(percentiles, mean=p, sd = norm_sd)
    #And the error between the empirical and theoretical percentiles
    norm_error_vector[i] = sum((empir_quants[quant_filt] - norm_percentiles[quant_filt])^2)
  }

  fitted_n = n[which.min(norm_error_vector)]  
  return(fitted_n)
}

#Function for getting a tissue's skews and random sampling them
bootstrap_n_est_norm = function(tissue){
  empir_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tissue]   #Get the folded skews
  num_samps = length(empir_skews)  #Sample number for that tissue
  sampled_skews = sample(empir_skews, size = num_samps, replace = TRUE) #Resample the folded skews
  fitted_n = fit_normal_to_skew_dist(unfold(sampled_skews)) #Fit the unfolded skews
  return(fitted_n)
}

```



```{r}

num_bootstraps = 2000

ci = 95
ci_end = ((100-ci) / 2) / 100

lower_ci_vec = vector(mode = 'numeric', length = length(filt_tissues))
names(lower_ci_vec) = filt_tissues
upper_ci_vec = vector(mode = 'numeric', length = length(filt_tissues))
names(upper_ci_vec) = filt_tissues
for(i in 1:length(filt_tissues)){

  tissue = filt_tissues[i]
  n_boots = unlist(mclapply(1:num_bootstraps, function(i) bootstrap_n_est_norm(tissue), mc.cores = 5))
  n_boots = n_boots[order(n_boots)]
  hist(n_boots, breaks = 16, main = sprintf('Bootstrap distribution of N: %i sims', num_bootstraps))
  
  #Basic bootstrap CIs
  #org_fit_n = est_stem_cell_pop_df$normal_est_n[est_stem_cell_pop_df$tissues == tissue]
  #upper_ci_vec[i] = 2*org_fit_n - n_boots[num_bootstraps * (1 - (1-ci_end))]
  #lower_ci_vec[i] =  2*org_fit_n - n_boots[num_bootstraps * (1- ci_end)]
  #Percentile CIs
  lower_ci_vec[i] = n_boots[num_bootstraps * (1 - (1-ci_end))]
  upper_ci_vec[i] = n_boots[num_bootstraps * (1- ci_end)]
  abline(v = lower_ci_vec[i],col = 'red')
  abline(v = upper_ci_vec[i],col = 'red')
}

#Add the CI too the dataframe
index = match(as.character(est_stem_cell_pop_df$tissues), filt_tissues)
lower_ci_vec = lower_ci_vec[index]
upper_ci_vec = upper_ci_vec[index]

est_stem_cell_pop_df$lower_norm_ci = lower_ci_vec
est_stem_cell_pop_df$upper_norm_ci = upper_ci_vec


```



```{r}
#Rearrange by germ layer and then cell number estimate
est_stem_cell_pop_df = est_stem_cell_pop_df[order(est_stem_cell_pop_df$germ_layers, est_stem_cell_pop_df$normal_est_n, decreasing = T) , ]

est_stem_cell_pop_df$tissues = factor(est_stem_cell_pop_df$tissues, levels = est_stem_cell_pop_df$tissues)

```

```{r}
cell_line_index = !est_stem_cell_pop_df$tissues %in% c('Cells - Cultured fibroblasts', 'Cells - EBV-transformed lymphocytes', 'Cells - Transformed fibroblasts')

mean(est_stem_cell_pop_df$normal_est_n[cell_line_index])

```


```{r}

ggplot(est_stem_cell_pop_df[cell_line_index, ], aes(x = tissues, y =normal_est_n, color = germ_layers)) + geom_point(size = 2) +
  geom_errorbar(aes(x = tissues, ymin = lower_norm_ci, ymax = upper_norm_ci),size=.5, alpha = .5) + 
  scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color)) + 
  geom_hline(yintercept=16, linetype='dashed', color='blue') +
  geom_hline(yintercept=6, linetype='dashed', color='blue') +
  ylab('Estimated cell number during XCI') + xlab('Tissues') +
  scale_y_continuous(breaks=c(6, 16, 20, 25)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(face="bold", size=12))

ggsave(filename='max_fitted_normal_stem_cell_pop_ests_with_ci.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/', units='in', height=5, width=7, dpi=300, useDingbats=FALSE)

```
save cell number estimates

```{r}

save(est_stem_cell_pop_df, file = '/home/werner/xchrom_snp_skew/data/GTEx/snp_skews/GATK_mod/cell_num_estimates/all_tissue_XCI_cell_num_ests_df.Rdata')

```





```{r}
which(filt_tissues == 'Whole Blood')

```


```{r}
current_tissue = filt_tissues[49]
fitted_normal_n = est_stem_cell_pop_df$normal_est_n[est_stem_cell_pop_df$tissues == current_tissue]
sigma = sqrt(p*(1-p)/fitted_normal_n)
dnorm_x = 0:1000/1000

fitted_normal = dnorm(dnorm_x, mean=p, sd = sigma)

theoretical_norms = list()
n = c(4,8,16,32)
for(i in 1:length(n)){ 
  sigma_t = sqrt(p*(1-p)/n[i])
  #theoretical_norms[[i]] = dnorm(dnorm_x, mean=p, sd = sigma)
  theoretical_norms[[i]] = sigma_t
}

skews = unfold(skew_and_stats_df$skew[skew_and_stats_df$tissue == current_tissue])

num_bins = 50
fill_vec = rep('dodgerblue', num_bins+1)
fill_vec[21:31] = 'grey'
color_vec = rep('black', num_bins+1)
color_vec[21:31] = 'grey'
df = data.frame(skews = skews)

ggplot(df, aes(x = skews)) + geom_histogram(aes(y = ..density..), fill = fill_vec, color = color_vec, binwidth = 1/num_bins) +
  xlim(0,1)+ xlab(sprintf('%s XCI skews', current_tissue)) +
  stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[1]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'dashed')) +
  stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[2]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'dotted')) + 
  stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[3]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'dotdash')) +
  stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[4]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'twodash')) +
  stat_function(fun = dnorm, args = list(mean = .5, sd = sigma), n = num_bins, size = 2, aes(linetype = 'solid')) +
  scale_linetype_manual(name="Estimated cell number",values=c('dashed','dotted','dotdash','twodash','solid'),
                        labels = c('4 cells','8 cells','16 cells','32 cells',sprintf('Estimated %i cells', fitted_normal_n)),
                        breaks=c('dashed','dotted','dotdash','twodash','solid'), 
                        guide = guide_legend(override.aes = list(size = c(.5, .5, .5, .5, 2)))) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_text(size=12),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12), 
        legend.position = c(.85,.75))

```


Pared down version with only the best fitting line
```{r}

ggplot(df, aes(x = skews)) + geom_histogram(aes(y = ..density..), fill = fill_vec, color = color_vec, binwidth = 1/num_bins) +
  xlim(0,1)+ xlab(sprintf('%s XCI skews', current_tissue)) +
  stat_function(fun = dnorm, args = list(mean = .5, sd = sigma), n = num_bins, size = 3, aes(linetype = 'solid')) +
  scale_linetype_manual(name="Estimated cell number",values=c('solid'),
                        labels = c(sprintf('Estimated %i cells', fitted_normal_n)),
                        breaks=c('solid')) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_text(size=12),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12), 
        legend.position = c(.85,.75))

```


Plot the binomial distributions overlain with estimated normal distributions,using the variance of the fitted binomials

```{r}
p = .5
num_bins = 50
#Greying out the .4-.6 skews
fill_vec = rep('dodgerblue', num_bins+1)
fill_vec[21:31] = 'grey'
color_vec = rep('black', num_bins+1)
color_vec[21:31] = 'grey'

for(j in 1:length(filt_tissues)){

  current_tissue = filt_tissues[j]
  fitted_normal_n = est_stem_cell_pop_df$normal_est_n[est_stem_cell_pop_df$tissues == current_tissue]
  sigma = sqrt(p*(1-p)/fitted_normal_n)
  dnorm_x = 0:1000/1000
  
  fitted_normal = dnorm(dnorm_x, mean=p, sd = sigma)
  
  theoretical_norms = list()
  n = c(4,8,16,32)
  for(i in 1:length(n)){ 
    sigma_t = sqrt(p*(1-p)/n[i])
    #theoretical_norms[[i]] = dnorm(dnorm_x, mean=p, sd = sigma)
    theoretical_norms[[i]] = sigma_t
  }
  
  skews = unfold(skew_and_stats_df$skew[skew_and_stats_df$tissue == current_tissue])
  df = data.frame(skews = skews)
  
  g = ggplot(df, aes(x = skews)) + geom_histogram(aes(y = ..density..), fill = fill_vec, color = color_vec, binwidth = 1/num_bins) +
    xlim(0,1)+ xlab(sprintf('%s XCI skews', current_tissue)) +
    stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[1]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'dashed')) +
    stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[2]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'dotted')) + 
    stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[3]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'dotdash')) +
    stat_function(fun = dnorm, args = list(mean = .5, sd = theoretical_norms[[4]]), n = num_bins, size = .75, alpha = .5, aes(linetype = 'twodash')) +
    stat_function(fun = dnorm, args = list(mean = .5, sd = sigma), n = num_bins, size = 2, aes(linetype = 'solid')) +
    scale_linetype_manual(name="Estimated cell number",values=c('dashed','dotted','dotdash','twodash','solid'),
                          labels = c('4 cells','8 cells','16 cells','32 cells',sprintf('Estimated %i cells', fitted_normal_n)),
                          breaks=c('dashed','dotted','dotdash','twodash','solid'), 
                          guide = guide_legend(override.aes = list(size = c(.5, .5, .5, .5, 2)))) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
          panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
          axis.text.y = element_text(size=12), axis.text.x = element_text(size=12),
          axis.title.y = element_text(size=12), axis.title.x = element_text(size=12), 
          legend.position = c(.8,.75))  
  print(g)
  
  #Fixing tissue name for file name
  tiss_name = current_tissue
  tiss_name = gsub(' ', '_', tiss_name, fixed = TRUE)
  tiss_name = gsub('-', '_', tiss_name, fixed = TRUE)

  ggsave(plot = g, filename = sprintf('%s_normal_fits.pdf', tiss_name) ,
         path ='/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/', 
         width = 6, height = 4, useDingbats = FALSE)
  
  
  #Pared down version
  g1 = ggplot(df, aes(x = skews)) + geom_histogram(aes(y = ..density..), fill = fill_vec, color = color_vec, binwidth = 1/num_bins) +
  xlim(0,1)+ xlab(sprintf('%s XCI skews', current_tissue)) +
  stat_function(fun = dnorm, args = list(mean = .5, sd = sigma), n = num_bins, size = 3, aes(linetype = 'solid')) +
  scale_linetype_manual(name="Estimated cell number",values=c('solid'),
                        labels = c(sprintf('Estimated %i cells', fitted_normal_n)),
                        breaks=c('solid')) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(size=12), axis.text.x = element_text(size=12),
        axis.title.y = element_text(size=12), axis.title.x = element_text(size=12), 
        legend.position = c(.85,.75))
    ggsave(plot = g1, filename = sprintf('%s_normal_fits_minimal_vers.pdf', tiss_name) ,
         path ='/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/', 
         width = 6, height = 4, useDingbats = FALSE)
  
}

```


#################################################
Variance of skews within individuals
#################################################



1.) Distribution of skew standard deviations for donors across tissues
2.) Distribution of deviance from mean donor folded skew per tissue, unfolded mean is goinng to be .5


```{r}
#Only sticking with the filtered tissues
tissues =  filt_tissues
length(tissues)
```




```{r}
#Standard deviations for the skews per donor

donor_sd_skews = vector(mode='numeric', length = length(donors))
donor_num_skews = vector(mode='numeric', length = length(donors))
donor_mean_skews = vector(mode='numeric', length = length(donors))
donor_median_skews = vector(mode='numeric', length = length(donors))
for( i in 1:length(donors)){

  if(dim(skew_and_stats_df[skew_and_stats_df$donor == donors[i], ])[1] <=1 ){next} #only one tissue for that donor
  
  #Grab all the skews for that donor
  donor_skews_folded = skew_and_stats_df$skew[skew_and_stats_df$donor == donors[i]]
  donor_sd_skews[i]= sd(donor_skews_folded)
  donor_num_skews[i]= length(donor_skews_folded)
  donor_mean_skews[i]= mean(donor_skews_folded)
  donor_median_skews[i] = median(donor_skews_folded)
}

#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/all_donor_skews_SD_distribution.pdf', width=5, height=4)
total_sd_mean = mean(donor_sd_skews)
title = "Standard Deviations for each donor's skew distribution"
hist(donor_sd_skews, main=title, freq=FALSE,cex.lab=1.5, cex.main=1.5,xlab='Standard Deviation' )
lines(density(donor_sd_skews), lwd=2)
abline(v=total_sd_mean, col='red', lwd=2, lty=2)
#dev.off()

hist(donor_sd_skews/donor_mean_skews, main='Donor folded skew CoefVar', breaks=16, xlim=c(0,.2), prob=TRUE)
lines(density(donor_sd_skews/donor_mean_skews, cut=0, na.rm=TRUE), lwd=2)
#SD shouldn't scale with sample size, but still good to check
plot(donor_sd_skews,donor_num_skews, main='Donor folded skew SDs vs number of tissue skews for donor', ylim=c(0,40), xlim=c(0,.2) )

pear_mean_sd = cor(donor_mean_skews, donor_sd_skews, method='pearson')
plot(donor_mean_skews, donor_sd_skews, main=sprintf('Donor mean skew vs SD of skews Rp= %f', pear_mean_sd), xlim=c(.5, 1))
pear_median_sd = cor(donor_median_skews, donor_sd_skews, method='pearson')
plot(donor_median_skews, donor_sd_skews, main=sprintf('Donor median skew vs SD of skews Rp= %f', pear_median_sd), xlim=c(.5, 1))


plot(donor_mean_skews, donor_num_skews, main ='Donor mean skew versus number of tissues')

```


Getting the tissue specific deviances from the donor mean
Also plot the distribution of donor mean skews for all the donors for that tissue. Checking if a tissue is sampled only from donors with typically high skews, might be meaningful


```{r}

for(i in 1:length(tissues)){

  current_tiss = tissues[i]
  
  
  #Get the donors for that tissue
  donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
  if(length(donors_for_tiss) == 0){next}
  #To store the tissue mean skew deviances from each donor
  tiss_skew_deviance = vector(mode='numeric', length = length(donors_for_tiss))
  tiss_donor_mean_skews  = vector(mode='numeric', length = length(donors_for_tiss))
  tiss_donor_skews  = vector(mode='numeric', length = length(donors_for_tiss))
  #Go through the donors
  for(j in 1:length(donors_for_tiss)){
  
    temp_donor = donors_for_tiss[j]
    
    if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] <=1 ){next} #only one tissue
    
    #get the mean folded skew for the donor
    donor_skews_mean = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor])     ########Not excluding the tissue of interest in the mean estimate, do it properly later on
    
    #Get the skew of that tissue from that donor
    
    #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
    tiss_donor_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
    
    tiss_skew_deviance[j] = abs(donor_skews_mean - tiss_donor_skew)
    tiss_donor_mean_skews[j] = donor_skews_mean
    tiss_donor_skews [j] = tiss_donor_skew
  }
  
  
  title = sprintf('Skew deviance for %s', current_tiss)
  hist(tiss_skew_deviance, main = title, xlab='Skew deviance from the donor folded mean skew ', xlim=c(0,.5), breaks=16, prob=TRUE)
  lines(density(tiss_skew_deviance, cut=0), lwd=2)
  abline(v=total_sd_mean, col='blue', lty='dashed')
  abline(v=total_sd_mean*2, col='red', lty='dashed')
  legend("topright", c("1 average donor SD", "2 average donor SD"), col=c('blue', 'red'), lty='dashed')
  
  title = sprintf('All donor mean skews for %s', current_tiss)
  hist(tiss_donor_mean_skews, main = title, xlab = 'Donor mean skews', xlim = c(.5,1), breaks = 16, prob=TRUE)
  lines(density(tiss_donor_mean_skews, cut=0), lwd=2)
  
  title = sprintf('Tissue skews vs SD for %s', current_tiss)
  plot(tiss_skew_deviance, tiss_donor_skews, main=title, ylab = 'Tissue skews', xlab='Skew deviance from donor mean skew', ylim = c(.5, 1), xlim = c(0, .4))
}

```




######################################
Playing around with the wilcoxen rank sum test as a classifier
######################################


```{r}
#Testing getting different AUCs from the wilcoxen rank sum test, trying to get an intuition for things

auc_wcrst <- function(labels, scores){
  labels <- as.logical(labels)
  n1 <- sum(labels)   #Number of positives
  n2 <- sum(!labels)  #Number of negatives
  R1 <- sum(rank(scores)[labels])   #Rank sum of positives
  U1 <- R1 - ((n1 * (n1 + 1))/2 )       #U-statistic
  return(U1/(n1 * n2))                      #AUC calculation
  
  
}

```



```{r}
group_1 = 45
group_2 = 55

category = rep(1, group_1)
category = c(category, rep(0, group_2))
#category = sample(category, size=length(category))



auc_vector = vector(mode='numeric', length = 250)
means = c(0,1,2,3,4,5,6,7)

for(j in means){

  for(i in 1:length(auc_vector)){
    
    group_1_vals = rnorm(group_1, mean = j, sd = 3)
    group_2_vals = rnorm(group_2, mean = 4, sd = 1)
    prediction = c(group_1_vals, group_2_vals)
    
    auc_vector[i] = auc_wcrst(category, prediction)
  
  }
  title = sprintf('positive normal dist mean: %i  , negative normal dist mean: %i ', j, 4)
  hist(auc_vector, xlim = c(0,1), xlab = 'AUC', main = title)

}
```


Make a function to draw an ROC curve

```{r}

#Input mean donor skews and tissue skews and a threshold
get_roc = function(mean_donor_skews, tissue_skews, threshold){
  #Classify by the mean donor skews
  classifier = mean_donor_skews >= threshold
  #Rank by the tissue skews
  rank_index = order(tissue_skews, decreasing = TRUE)
  tissue_skews = tissue_skews[rank_index]
  classifier = classifier[rank_index]
  
  #If there's no positives, just return NA
  if(sum(classifier) == 0){return(NA)}
  
  #Also get the AUC and pvalue
  auc = auc_wcrst(classifier, tissue_skews)
  MWU_test = suppressWarnings(wilcox.test(tissue_skews[classifier], tissue_skews[!classifier], alternative='two.sided'))
  auc_pvalue = MWU_test$p.value
  
  #Set up FPR (x-axis) and TPR (y-axis) vectors
  x_vec = vector(mode = 'numeric', length = length(classifier) + 1)
  y_vec = vector(mode = 'numeric', length = length(classifier) + 1)
  x_vec[1] = 0
  y_vec[1] = 0
  
  #Go through the ranked list, when a negative comes up, increment x_axis by one. When positive, increment y-axis by 1
  for( i in 1:length(classifier)){
    
    if(classifier[i] == 0){
      x_vec[i+1] = x_vec[i] + 1
      y_vec[i+1] = y_vec[i]
      }else{
      x_vec[i+1] = x_vec[i]
      y_vec[i+1] = y_vec[i] + 1}  
  }
  #Normalize by the total number of negatives and positives
  x_vec = x_vec / sum(!classifier)
  y_vec = y_vec / sum(classifier)

  #Just make an equal length vector for the AUC
  auc_vec = rep(auc, length(classifier)+1)
  auc_pvalue_vec = rep(auc_pvalue, length(classifier)+1)
  
  roc_coords = data.frame(FPR = x_vec, TPR = y_vec, AUC = auc_vec, pvalue = auc_pvalue_vec)
  return(roc_coords)
}

mean_donor_skews = sample(seq(.5,1, .001), 50, replace = T)
tissue_skews = sample(seq(.5,1, .001), 50, replace = T)
threshold = .7

test = get_roc(mean_donor_skews, tissue_skews, threshold)
test
plot(test$FPR, test$TPR, type = 'l',lty = 1)



```


#######################################
Trying out some classifier approaches to interrogating skew
1.) Can the skew of a tissue predict the more generally skewed donors? Ranking the skew of a tissue and see if it can predict the mean skew of the donor. Withold the skew of the tissue from the mean calculation
  
#######################################

function to go through tissues and get the donor mean skew and tissue skews
```{r}

get_mean_donor_and_tissue_skews = function(current_tiss){

  #Get the donors for that tissue
  donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
  num_donors = length(donors_for_tiss) 
  if(num_donors == 0){return(NA)}
  #To store the tissue mean skew deviances from each donor, and the skew of the tissue
  tiss_donor_skew = vector(mode='numeric', length = num_donors)
  donor_mean_skews = vector(mode='numeric', length = num_donors)
  
  #Go through the donors
  for(j in 1:num_donors){
  
    temp_donor = donors_for_tiss[j]
    if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] <=1 ){next} #Only one tissue for that donor
    
    #get the mean folded skew for the donor, withold the current tissue skew
    donor_mean_skews[j] = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue != current_tiss])
    
    #Get the skew of that tissue from that donor
    #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
    tiss_donor_skew[j] = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
  }
  
  skew_df = data.frame(donor_mean_skews = donor_mean_skews, tissue_skews = tiss_donor_skew)
  return(skew_df)
}

```

BrBG, PiYG, PRGn, PuOr, RdBu, RdGy, RdYlBu, RdYlGn, Spectral


```{r}

all_roc_funcs = function(threshold, file_label, dens_adjust = 1){

  
  auc_pvalue_vec = c()
  
  for(i in 1:length(tissues)){
    current_tiss = tissues[i]
  
    #Get the mean donor skews and tissue skews
    skew_df = get_mean_donor_and_tissue_skews(current_tiss)
    if(is.null(dim(skew_df))){next}
    donor_mean_skews = skew_df$donor_mean_skews
    tiss_donor_skew = skew_df$tissue_skews
    
    #Get pvalues for the AUCs, only going to plot the ROC curves for the sig AUCs
    classifier = donor_mean_skews >= threshold
    
    #If there's no positives, just move on
    if(sum(classifier) == 0){next}
    MWU_test = suppressWarnings(wilcox.test(tiss_donor_skew[classifier], tiss_donor_skew[!classifier], alternative='two.sided'))
    auc_pvalue_vec[i] = MWU_test$p.value
  }
  
  auc_pvalue_vec = p.adjust(auc_pvalue_vec, method = 'BH')
  sig_auc_index = which(auc_pvalue_vec <= .05)
  
  
  
  file = sprintf('/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/sig_roc_curves_%s.pdf', file_label)
  pdf(file, useDingbats = F, height = 4, width = 4)
  par(pty = 's')
  title = sprintf('Mean donor skew threshold: %0.2f', threshold)
  plot(1, type="l", xlab="FPR", ylab="TPR", xlim=c(0, 1), ylim=c(0, 1), main = title)
  
  #Now go through and only plot the ROC curves for the significant AUCs
  auc = c()
  all_roc_df = data.frame(FPR = double(), TPR = double(), AUC = double(), pvalue = double())
  for(i in sig_auc_index){
    current_tiss = tissues[i]
  
    #Get the mean donor skews and tissue skews
    skew_df = get_mean_donor_and_tissue_skews(current_tiss)
    #if(is.null(dim(skew_df))){next}
    donor_mean_skews = skew_df$donor_mean_skews
    tiss_donor_skew = skew_df$tissue_skews
    
    roc_dataframe = get_roc(donor_mean_skews, tiss_donor_skew, threshold)
    #Grab the AUC for later
    auc = c(auc, roc_dataframe$AUC[1])
    #Plot that tissue's ROC
    lines(roc_dataframe$FPR, roc_dataframe$TPR, type = 'l', lty = 1)
    #Save it
    all_roc_df = rbind(all_roc_df, roc_dataframe)
  }
  
  abline(a = 0, b = 1, col = 'red')
  dev.off()
  
  #Plot the ROCs as a 2d density plot
  p = ggplot(all_roc_df, aes(x=FPR, y=TPR) ) +
  stat_density_2d(aes(fill = ..density..), geom = "raster", contour = FALSE, n = 200, adjust = dens_adjust) +
  scale_fill_distiller(palette= "Blues", direction=1) +
  ylim(0,1) + xlim(0,1) + coord_fixed() +
  ylab('TPR') + xlab('FPR') + ggtitle(sprintf('Threshold: %1.2f',threshold)) +
  geom_abline(slope = 1, intercept = 0, color = 'red') +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
       panel.border = element_rect(colour = "black", fill=NA, size=1),
       plot.title = element_text(hjust = .5), 
       axis.title.y = element_text(size =12), axis.text.y = element_text(size = 12), 
       axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 12),
       panel.background = element_rect(fill = 'white'))
  
  print(p)
  file_name_2d_roc = sprintf('combined_2d_density_roc_%i.pdf', threshold*100)
  ggsave(filename = file_name_2d_roc, device = 'pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/', plot = p, 
       height = 6, width = 6, dpi = 300, useDingbats = F)
  
  #Get a dataframe of the significant AUCs for plotting later
  auc_label = rep(sprintf('Mean donor skew threshold: %0.2f', threshold), length(sig_auc_index))
  auc_df = data.frame(auc = auc, label = auc_label, tissue = tissues[sig_auc_index])
  
  return(auc_df)

}

sig_auc_60_df = all_roc_funcs(threshold = .60, file_label = '60', dens_adjust = 2)
sig_auc_65_df = all_roc_funcs(threshold = .65, file_label = '65', dens_adjust = 2)
sig_auc_70_df = all_roc_funcs(threshold = .70, file_label = '70', dens_adjust = 2)
sig_auc_75_df = all_roc_funcs(threshold = .75, file_label = '75', dens_adjust = 2)




```


Get the average AUC

```{r}
mean(sig_auc_60_df$auc)
mean(sig_auc_65_df$auc)
mean(sig_auc_70_df$auc)
mean(sig_auc_75_df$auc)
```





```{r}
all_threshold_auc_df = rbind(sig_auc_60_df, sig_auc_65_df, sig_auc_70_df, sig_auc_75_df)
all_threshold_auc_df
all_threshold_auc_df$label = factor(all_threshold_auc_df$label, levels = c("Mean donor skew threshold: 0.75","Mean donor skew threshold: 0.70",
                                                                           "Mean donor skew threshold: 0.65","Mean donor skew threshold: 0.60"))
```



```{r}

ggplot(all_threshold_auc_df, aes(x= auc, y = label)) + geom_violin() + geom_jitter(height = 0.1, width = 0, alpha = .5, size = 2) +
  xlim(.5,1) + xlab('AUC') +
  scale_y_discrete(position = "right") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 12),
        axis.title.y = element_blank(), axis.text.y = element_text(size = 12), 
        plot.title = element_text(size = 12, hjust = .5), 
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        panel.background = element_rect(fill = 'white'))
ggsave(filename = 'auc_distributions_all_thresh_violins.pdf', device = 'pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/', 
       height = 6, width = 6, dpi = 300, useDingbats = F)

```





```{r}

cross_tissue_auc_65 = rep(NA, length(tissues))
cross_tissue_auc_7 = rep(NA, length(tissues))
cross_tissue_auc_75 = rep(NA, length(tissues))
cross_tissue_auc_8 = rep(NA, length(tissues))

cross_tissue_auc_65_pval = rep(NA, length(tissues))
cross_tissue_auc_7_pval = rep(NA, length(tissues))
cross_tissue_auc_75_pval = rep(NA, length(tissues))
cross_tissue_auc_8_pval = rep(NA, length(tissues))

donor_skew_thresh = seq( from=.5, to=1, by=.005)


for(i in 1:length(tissues)){
  current_tiss = tissues[i]
  
  
  #Get the donors for that tissue
  donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
  num_donors = length(donors_for_tiss) 
  if(num_donors == 0){next}
  #To store the tissue mean skew deviances from each donor, and the skew of the tissue
  tiss_donor_skew = vector(mode='numeric', length = num_donors)
  donor_mean_skews = vector(mode='numeric', length = num_donors)
  
  #Go through the donors
  for(j in 1:num_donors){
  
    temp_donor = donors_for_tiss[j]
    
    if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] <=1 ){next} #Only one tissue for that donor
    
    #get the mean folded skew for the donor, withold the current tissue skew
    donor_mean_skews[j] = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue != current_tiss])
    
    #Get the skew of that tissue from that donor
    #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
    tiss_donor_skew[j] = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
  }
  
  
  auc = vector(mode='numeric', length=length(donor_skew_thresh))
  
  for(k in 1:length(donor_skew_thresh)){
    labels = donor_mean_skews >= donor_skew_thresh[k]
    auc[k] = auc_wcrst(labels, tiss_donor_skew)
  }
  
  title = sprintf('%s : %g donors', current_tiss, num_donors)
  plot(donor_skew_thresh, auc, main = title, xlab = 'Positive threshold: Donor mean skew', ylab = 'AUC for skew', ylim = c(0,1))
  lines(donor_skew_thresh, auc)
  
  
  
  #Save the AUC at skew threshold of .8 for all the tissues
  labels = donor_mean_skews  >= .65
  if(sum(labels) == 0){next}
  cross_tissue_auc_65[i] = auc_wcrst(labels, tiss_donor_skew)
  MWU_test = wilcox.test(tiss_donor_skew[labels], tiss_donor_skew[!labels], alternative='two.sided')
  cross_tissue_auc_65_pval[i] = MWU_test$p.value
  
  
  labels = donor_mean_skews  >= .7
  if(sum(labels) == 0){next}
  cross_tissue_auc_7[i] = auc_wcrst(labels, tiss_donor_skew)
  MWU_test = wilcox.test(tiss_donor_skew[labels], tiss_donor_skew[!labels], alternative='two.sided')
  cross_tissue_auc_7_pval[i] = MWU_test$p.value
  
  
  labels = donor_mean_skews  >= .75
  if(sum(labels) == 0){next}
  cross_tissue_auc_75[i] = auc_wcrst(labels, tiss_donor_skew)
  MWU_test = wilcox.test(tiss_donor_skew[labels], tiss_donor_skew[!labels], alternative='two.sided')
  cross_tissue_auc_75_pval[i] = MWU_test$p.value
  
  labels = donor_mean_skews  >= .8
  if(sum(labels) == 0){next}
  cross_tissue_auc_8[i] = auc_wcrst(labels, tiss_donor_skew)
  MWU_test = wilcox.test(tiss_donor_skew[labels], tiss_donor_skew[!labels], alternative='two.sided')
  cross_tissue_auc_8_pval[i] = MWU_test$p.value

}

hist(cross_tissue_auc_65, main = 'All tissues, donor mean skew thresh = .65', xlab = 'AUC for tissue skew predicting donor mean skew', prob=FALSE, xlim=c(0,1), breaks=32)
#lines(density(cross_tissue_auc_7, na.rm=TRUE, cut=0), lwd=2)
abline(v=.5, col='red')
par(pty='s')
plot(cross_tissue_auc_65,cross_tissue_auc_65_pval, main = 'AUC versus p_value, donor mean skew thresh = .65', ylim = c(0,1), xlim = c(.4,1),
     ylab = 'two.sided p_value', xlab = 'AUC for tissue skew predicting donor mean skew')
abline(h=.05, col='red')


hist(cross_tissue_auc_7, main = 'All tissues, donor mean skew thresh = .70', xlab = 'AUC for tissue skew predicting donor mean skew', prob=FALSE, xlim=c(0,1), breaks=32)
#lines(density(cross_tissue_auc_7, na.rm=TRUE, cut=0), lwd=2)
abline(v=.5, col='red')
par(pty='s')
plot(cross_tissue_auc_7,cross_tissue_auc_7_pval, main = 'AUC versus p_value, donor mean skew thresh = .70', ylim = c(0,1), xlim = c(.4,1),
     ylab = 'two.sided p_value', xlab = 'AUC for tissue skew predicting donor mean skew')
abline(h=.05, col='red')

hist(cross_tissue_auc_75, main = 'All tissues, donor mean skew thresh = .75', xlab = 'AUC for tissue skew predicting donor mean skew', prob=FALSE, xlim=c(0,1), breaks=32)
#lines(density(cross_tissue_auc_75, na.rm=TRUE, cut=0), lwd=2)
abline(v=.5, col='red')
par(pty='s')
plot(cross_tissue_auc_75,cross_tissue_auc_75_pval, main = 'AUC versus p_value, donor mean skew thresh = .75', ylim = c(0,1), xlim = c(.4,1),
     ylab = 'two.sided p_value', xlab = 'AUC for tissue skew predicting donor mean skew')
abline(h=.05, col='red')

hist(cross_tissue_auc_8, main = 'All tissues, donor mean skew thresh = .80', xlab = 'AUC for tissue skew predicting donor mean skew', prob=FALSE, xlim=c(0,1), breaks=32)
#lines(density(cross_tissue_auc_8, na.rm=TRUE, cut=0), lwd=2)
abline(v=.5, col='red')
par(pty='s')
plot(cross_tissue_auc_8,cross_tissue_auc_8_pval, main = 'AUC versus p_value, donor mean skew thresh = .80', ylim = c(0,1), xlim = c(.4,1),
     ylab = 'two.sided p_value', xlab = 'AUC for tissue skew predicting donor mean skew')
abline(h=.05, col='red')
```


```{r}
tissue_auc_donor_mean_skew_df = data.frame(tissue = tissues, auc_65 = cross_tissue_auc_65, auc_65_pval = cross_tissue_auc_65_pval, 
                                           auc_70 = cross_tissue_auc_7, auc_70_pval = cross_tissue_auc_7_pval, 
                                           auc_75 = cross_tissue_auc_75, auc_75_pval = cross_tissue_auc_75_pval, 
                                           auc_80 = cross_tissue_auc_8, auc_80_pval = cross_tissue_auc_8_pval)

index = order(tissue_auc_donor_mean_skew_df$auc_65)
par(mar = c(12,4,2,2))
barp = barplot(tissue_auc_donor_mean_skew_df$auc_65[index] , names = tissue_auc_donor_mean_skew_df $tissue[index], main = 'Tissue AUC for predicting mean donor skew (positives >= .65)', ylab = 'AUC', las=2, col = 'grey', cex.names=.65, ylim = c(0,1))
abline(h=.5, col = 'red')

index = order(tissue_auc_donor_mean_skew_df$auc_70)
par(mar = c(12,4,2,2))
barp = barplot(tissue_auc_donor_mean_skew_df$auc_70[index] , names = tissue_auc_donor_mean_skew_df $tissue[index], main = 'Tissue AUC for predicting mean donor skew (positives >= .70)', ylab = 'AUC', las=2, col = 'grey', cex.names=.65, ylim = c(0,1))
abline(h=.5, col = 'red')

index = order(tissue_auc_donor_mean_skew_df$auc_75)
par(mar = c(12,4,2,2))
barp = barplot(tissue_auc_donor_mean_skew_df$auc_75[index] , names = tissue_auc_donor_mean_skew_df $tissue[index], main = 'Tissue AUC for predicting mean donor skew (positives >= .75)', ylab = 'AUC', las=2, col = 'grey', cex.names=.65, ylim = c(0,1))
abline(h=.5, col = 'red')

index = order(tissue_auc_donor_mean_skew_df$auc_80)
par(mar = c(12,4,2,2))
barp = barplot(tissue_auc_donor_mean_skew_df$auc_80[index] , names = tissue_auc_donor_mean_skew_df $tissue[index], main = 'Tissue AUC for predicting mean donor skew (positives >= .80)', ylab = 'AUC', las=2, col = 'grey', cex.names=.65, ylim = c(0,1))
abline(h=.5, col = 'red')

tissue_auc_donor_mean_skew_df

```

```{r}
tissue_auc_donor_mean_skew_df

```


just looking at the significant AUCs for a skew threshold of .75

```{r}
opar = par()
#Filter out the NA tissues
temp_df = tissue_auc_donor_mean_skew_df[!is.na(tissue_auc_donor_mean_skew_df$auc_75), ]

pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/sig_auc_distribution_tiss_skew_predict_donor_skew.pdf', height = 4.5, useDingbats = FALSE)
corrected_pvals = p.adjust(temp_df$auc_75_pval, method = 'BH')
sig_aucs = temp_df$auc_75[corrected_pvals <= .05]
par(pty='s')
hist(sig_aucs, main = 'Tissues have high performance predicting mean donor skews', 
     xlab = 'AUC for tissue skew predicting donor mean skew', xlim=c(0,1), breaks = seq(0,1,1/40))
abline(v=.5, col='black', lty = 2)
dev.off()

corrected_pvals = p.adjust(temp_df$auc_75_pval, method = 'BH')
sig_aucs = temp_df$auc_75[corrected_pvals <= .05]
par(pty='s')
hist(sig_aucs, main = 'Tissues have high performance predicting mean donor skews', 
     xlab = 'AUC for tissue skew predicting donor mean skew', xlim=c(0,1), breaks = seq(0,1,1/40))
abline(v=.5, col='black', lty = 2)

#Filter out insignificant AUCs and reorder based off of AUC
temp_df = temp_df[corrected_pvals <= .05, ]
temp_df = temp_df[order(temp_df$auc_75), ]
tissue_names = as.character(temp_df$tissue)


pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/sig_auc_barplot_tiss_skew_predict_donor_skew.pdf', height = 4.5, useDingbats = FALSE)
par(opar)
par(mar = c(12,4,2,2))
barp = barplot(temp_df$auc_75 , names = tissue_names, main = 'Tissue AUC for predicting mean donor skew (positives >= .75)', ylab = 'AUC', las=2, col = 'grey', cex.names=.7, ylim = c(0,1))
abline(h=.5, col = 'black', lty = 2)
dev.off()

barp = barplot(temp_df$auc_75 , names = tissue_names, main = 'Tissue AUC for predicting mean donor skew (positives >= .75)', ylab = 'AUC', las=2, col = 'grey', cex.names=.7, ylim = c(0,1))
abline(h=.5, col = 'black', lty = 2)

#Now color by germ layer
temp_germ_layer = vector(mode = 'character', length = length(tissue_names))
for(i in 1:length(temp_germ_layer)){
  tissue = tissue_names[i]
  temp_germ_layer[i] = as.character(skew_and_stats_df$germ_layer[skew_and_stats_df$tissue == tissue][1])
}

temp_germ_layer_color = vector(mode = 'character', length = length(tissue_names))
temp_germ_layer_color[temp_germ_layer == 'Ectoderm'] = ecto_color
temp_germ_layer_color[temp_germ_layer == 'Endoderm'] = endo_color
temp_germ_layer_color[temp_germ_layer == 'Mesoderm'] = meso_color

pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/sig_auc_barplot_germlayer_color_tiss_skew_predict_donor_skew.pdf', height = 4.5, useDingbats = FALSE)

#Order by germ layer
index = order(temp_germ_layer_color)
temp_df = temp_df[index, ]
temp_germ_layer = temp_germ_layer[index]
temp_germ_layer_color = temp_germ_layer_color[index]

#Order each germ layer by AUC
ecto_index = which(temp_germ_layer == 'Ectoderm')
ecto_index = ecto_index[order(temp_df$auc_75[ecto_index])]
endo_index = which(temp_germ_layer == 'Endoderm')
endo_index = endo_index[order(temp_df$auc_75[endo_index])]
meso_index = which(temp_germ_layer == 'Mesoderm')
meso_index = meso_index[order(temp_df$auc_75[meso_index])]
final_index = c(ecto_index, endo_index, meso_index)


par(opar)
par(mar = c(12,4,2,2))
barp = barplot(temp_df$auc_75[final_index] , names = as.character(temp_df$tissue[final_index]),
               main = 'Tissue AUC for predicting mean donor skew (positives >= .75)', ylab = 'AUC', 
               las=2, col = temp_germ_layer_color[final_index], cex.names=.7, ylim = c(0,1))
abline(h=.5, col = 'black', lty = 2)
legend("bottomright", c("Ectoderm", "Endoderm", "Mesoderm"), col=c(ecto_color, endo_color, meso_color), lwd=8, cex=.75)
dev.off()

par(opar)
par(mar = c(12,4,2,2))
barp = barplot(temp_df$auc_75[final_index] , names = as.character(temp_df$tissue[final_index]),
               main = 'Tissue AUC for predicting mean donor skew (positives >= .75)', ylab = 'AUC', 
               las=2, col = temp_germ_layer_color[final_index], cex.names=.7, ylim = c(0,1))
abline(h=.5, col = 'black', lty = 2)
legend("bottomright", c("Ectoderm", "Endoderm", "Mesoderm"), col=c(ecto_color, endo_color, meso_color), lwd=8, cex = .75)

```


Do the same for the insignificant AUCs

```{r}
#Filter out the NA tissues
temp_df = tissue_auc_donor_mean_skew_df[!is.na(tissue_auc_donor_mean_skew_df$auc_75), ]
corrected_pvals = p.adjust(temp_df$auc_75_pval, method = 'BH')

#Filter out significant AUCs and reorder based off of AUC
temp_df = temp_df[corrected_pvals > .05, ]
temp_df = temp_df[order(temp_df$auc_75), ]
tissue_names = as.character(temp_df$tissue)
#Now color by germ layer

temp_germ_layer = vector(mode = 'character', length = length(tissue_names))
for(i in 1:length(temp_germ_layer)){
  tissue = tissue_names[i]
  temp_germ_layer[i] = as.character(skew_and_stats_df$germ_layer[skew_and_stats_df$tissue == tissue][1])
}

temp_germ_layer_color = vector(mode = 'character', length = length(tissue_names))
temp_germ_layer_color[temp_germ_layer == 'Ectoderm'] = ecto_color
temp_germ_layer_color[temp_germ_layer == 'Endoderm'] = endo_color
temp_germ_layer_color[temp_germ_layer == 'Mesoderm'] = meso_color

#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/nonsig_auc_barplot_germlayer_color_tiss_skew_predict_donor_skew.pdf', height = 4.5, useDingbats = FALSE)

#Order by germ layer
index = order(temp_germ_layer_color)
temp_df = temp_df[index, ]
temp_germ_layer = temp_germ_layer[index]
temp_germ_layer_color = temp_germ_layer_color[index]

#Order each germ layer by AUC
ecto_index = which(temp_germ_layer == 'Ectoderm')
ecto_index = ecto_index[order(temp_df$auc_75[ecto_index])]
endo_index = which(temp_germ_layer == 'Endoderm')
endo_index = endo_index[order(temp_df$auc_75[endo_index])]
meso_index = which(temp_germ_layer == 'Mesoderm')
meso_index = meso_index[order(temp_df$auc_75[meso_index])]
final_index = c(ecto_index, endo_index, meso_index)


par(opar)
par(mar = c(12,4,2,2))
barp = barplot(temp_df$auc_75[final_index] , names = as.character(temp_df$tissue[final_index]),
               main = 'Tissues with non-significant AUC p-values after FDR correction', ylab = 'AUC', 
               las=2, col = temp_germ_layer_color[final_index], cex.names=.7, ylim = c(0,1))
abline(h=.5, col = 'black', lty = 2)
legend("bottomright", c("Ectoderm", "Endoderm", "Mesoderm"), col=c(ecto_color, endo_color, meso_color), lwd=8, cex=.75)
#dev.off()

```


Or just show violin plots for the AUC distributions


```{r}
temp_df = tissue_auc_donor_mean_skew_df[ , c('tissue', 'auc_75','auc_75_pval')]

tissue_names = as.character(temp_df$tissue)

temp_germ_layer = vector(mode = 'character', length = length(tissue_names))
for(i in 1:length(temp_germ_layer)){
  tissue = tissue_names[i]
  temp_germ_layer[i] = as.character(skew_and_stats_df$germ_layer[skew_and_stats_df$tissue == tissue][1])
}

temp_df$germ_layer = temp_germ_layer


corrected_pvals = p.adjust(temp_df$auc_75_pval, method = 'BH')



temp_df = temp_df[!is.na(temp_df$auc_75) & corrected_pvals <= .05, ]

ggplot(temp_df, aes(x = germ_layer, y = auc_75, fill = germ_layer)) + geom_violin() +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=.75, aes( x = germ_layer, y = auc_75), inherit.aes = FALSE) +
  scale_fill_manual(values = c(ecto_color, endo_color, meso_color)) +
  ylim(.5, 1) + ylab('Tissue skews predict mean donor skew (AUC)') + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        axis.title.x = element_blank(), axis.text.x = element_text(size = 12),
        axis.title.y = element_text(size = 12), axis.text.y = element_text(size = 12), 
        plot.title = element_text(size = 12, hjust = .5), 
        panel.border = element_rect(colour = "black", fill=NA, size=1), 
        panel.background = element_rect(fill = 'white'))

ggsave(filename = 'sig_auc_tiss_skew_predict_mean_donor_skew_violin_plots.pdf', device = 'pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/tiss_auc_donor_skew/', 
       dpi = 300, width = 4, height = 4, useDingbats = FALSE)

```




######################################
Now do it for age, see if the ranking of skews can predict the donors age
```{r}

age_aucs_50 = rep(NA, length(tissues))
age_aucs_60 = rep(NA, length(tissues))
tissue_num_donors = vector(mode='numeric', length=length(tissues))
tissue_num_old_donors = vector(mode='numeric', length=length(tissues))

age_aucs_50_pval = rep(NA, length(tissues))
age_aucs_60_pval = rep(NA, length(tissues))


age_tissue_auc_df = data.frame(tissue = tissues)


for(i in 1:length(tissues)){
  current_tiss = tissues[i]
  
  
  #Get the donors for that tissue
  donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
  num_donors = length(donors_for_tiss) 
  if(num_donors == 0){next}
  #To store the tissue mean skew deviances from each donor, and the skew of the tissue
  tiss_donor_skew = vector(mode='numeric', length = num_donors)
  donor_ages = vector(mode='numeric', length = num_donors)
  
  #Go through the donors
  for(j in 1:num_donors){
  
    temp_donor = donors_for_tiss[j]
    if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] ==1 ){
      #Only one tissue for that donor
      tiss_donor_skew[j] = skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss] 
      } 
    #grab the age of the donor
    donor_ages[j] = skew_and_stats_df$age[skew_and_stats_df$donor == temp_donor][1]
    
    #Get the skew of that tissue from that donor
    #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
    tiss_donor_skew[j] = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
  }
  

  #Positive cases with be the 60-69 and 70-79 donors
  labels = !donor_ages %in% c('20-29', '30-39', '40-49')
  if(sum(labels) == 0){next}
  age_aucs_50[i] = auc_wcrst(labels, tiss_donor_skew)
  MWU_test = wilcox.test(tiss_donor_skew[labels], tiss_donor_skew[!labels], alternative='two.sided')
  age_aucs_50_pval[i] = MWU_test$p.value
  
  labels = !donor_ages %in% c('20-29', '30-39', '40-49', '50-59')
  if(sum(labels) == 0){next}
  age_aucs_60[i] = auc_wcrst(labels, tiss_donor_skew)
  MWU_test = wilcox.test(tiss_donor_skew[labels], tiss_donor_skew[!labels], alternative='two.sided')
  age_aucs_60_pval[i] = MWU_test$p.value
  
  
  tissue_num_donors[i] = num_donors
  tissue_num_old_donors[i] = sum(labels)
  
}


hist(age_aucs_50, main = 'All tissues, positive ages = 50-80', xlab = 'AUC for tissue skew predicting donor age (positives >= 50 years)', prob=TRUE, xlim=c(0,1), breaks=16)
lines(density(age_aucs_50, na.rm=TRUE, cut=0), lwd=2)
abline(v=.5, col='red')


#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/age_auc/auc_distribution_tissue_skew_predicting_age_60_and_greater.pdf')
hist(age_aucs_60, main = 'All tissues, positive ages = 60-80', xlab = 'AUC for tissue skew predicting donor age (positives >= 60 years)', prob=TRUE, xlim=c(0,1), breaks=16)
lines(density(age_aucs_60, na.rm=TRUE, cut=0), lwd=2)
abline(v=.5, col='red')
#dev.off()

hist(age_aucs_60, main = 'All tissues, positive ages = 60-80', xlab = 'AUC for tissue skew predicting donor age (positives >= 60 years)', prob=TRUE, xlim=c(0,1), breaks=16)
lines(density(age_aucs_60, na.rm=TRUE, cut=0), lwd=2)
abline(v=.5, col='red')

age_tissue_auc_df$aucs_50 = age_aucs_50
age_tissue_auc_df$aucs_50_pval = p.adjust(age_aucs_50_pval, method='BH')
age_tissue_auc_df$aucs_60 = age_aucs_60
age_tissue_auc_df$aucs_60_pval = p.adjust(age_aucs_60_pval, method='BH')
age_tissue_auc_df$num_donors =  tissue_num_donors
age_tissue_auc_df$num_old_donors =  tissue_num_old_donors


par(pty='s')
plot(age_tissue_auc_df$aucs_50,age_tissue_auc_df$aucs_50_pval,ylim=c(0,1),xlim=c(0,1), main = 'AUC versus p_value, donor age thresh = 50', ylab = 'two.sided p_value', 
     xlab = 'AUC for tissue skew predicting donor age (positives >= 50 years)')
abline(h=.05, col='red')


#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/age_auc/auc_pvals_tissue_skew_predicting_age_60_and_greater.pdf')
plot(age_tissue_auc_df$aucs_60,age_tissue_auc_df$aucs_60_pval,ylim=c(0,1),xlim=c(-.15,1), main = 'AUC versus p_value, donor age thresh = 60', ylab = 'two.sided p_value', 
     xlab = 'AUC for tissue skew predicting donor age (positives >= 60 years)')
abline(h=.05, col='red')
index = age_tissue_auc_df$aucs_60_pval<= .2
text(age_tissue_auc_df$aucs_60[index], age_tissue_auc_df$aucs_60_pval[index], labels = tissues[index], cex=.7, pos=3)
#dev.off()

par(pty='s')
plot(age_tissue_auc_df$aucs_60,age_tissue_auc_df$aucs_60_pval,ylim=c(0,1),xlim=c(-.15,1), main = 'AUC versus p_value, donor age thresh = 60', ylab = 'two.sided p_value', 
     xlab = 'AUC for tissue skew predicting donor age (positives >= 60 years)')
abline(h=.05, col='red')
index = age_tissue_auc_df$aucs_60_pval<= .2
text(age_tissue_auc_df$aucs_60[index], age_tissue_auc_df$aucs_60_pval[index], labels = tissues[index], cex=.7, pos=3)


age_tissue_auc_df



```

Show the sskew distributions for the Brain Frontal Cortex and blood, separated by the age threshold
```{r}

brain_old_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == 'Brain - Frontal Cortex (BA9)' & skew_and_stats_df$age %in% c('60-69', '70-79')]
brain_new_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == 'Brain - Frontal Cortex (BA9)' & skew_and_stats_df$age %in% c('20-29', '30-39', '40-49' , '50-59')]

blood_old_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == 'Whole Blood' & skew_and_stats_df$age %in% c('60-69', '70-79')]
blood_new_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == 'Whole Blood' & skew_and_stats_df$age %in% c('20-29', '30-39', '40-49' , '50-59')]

muscle_old_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == 'Muscle - Skeletal' & skew_and_stats_df$age %in% c('60-69', '70-79')]
muscle_new_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == 'Muscle - Skeletal' & skew_and_stats_df$age %in% c('20-29', '30-39', '40-49' , '50-59')]


#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/age_auc/brain_skews_by_age.pdf', width = 5, height=4)
hist(brain_old_skews, main='Brain - Frontal Cortex (BA9) skews by age',col=rgb(0.8,0.8,0.8,0.5), xlim = c(.5,1), xlab = ('Folded skew estimate'), freq=FALSE, cex.lab=1.5)
hist(brain_new_skews, col=rgb(0.1,0.1,0.1,0.5), add=T, freq=FALSE)
legend("topright", c("Young", "Old"), col=c(rgb(0.1,0.1,0.1,0.5), rgb(0.8,0.8,0.8,0.5)), lwd=10)
lines(density(brain_old_skews),col='grey', lwd=5)
lines(density(brain_new_skews),col='black', lwd=5)
#dev.off()

#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/age_auc/blood_skews_by_age.pdf', width = 5, height=4)
hist(blood_new_skews, main='Whole Blood skews by age', col = rgb(0.1,0.1,0.1,0.5), xlim = c(.5,1), xlab = ('Folded skew estimate'), freq=FALSE, cex.lab=1.5)
hist(blood_old_skews, col=rgb(0.8,0.8,0.8,0.5), add=T,freq=FALSE)
legend("topright", c("Young", "Old"), col=c(rgb(0.1,0.1,0.1,0.5), rgb(0.8,0.8,0.8,0.5)), lwd=10)
lines(density(blood_old_skews),col='grey', lwd=5)
lines(density(blood_new_skews),col='black', lwd=5)
#dev.off()

#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/age_auc/muscle_skews_by_age.pdf', width = 5, height=4)
hist(muscle_new_skews, main='Muscle - Skeletal skews by age', col = rgb(0.1,0.1,0.1,0.5), xlim = c(.5,1), xlab = ('Folded skew estimate'), freq=FALSE, cex.lab=1.5)
hist(muscle_old_skews, col=rgb(0.8,0.8,0.8,0.5), add=T,freq=FALSE)
legend("topright", c("Young", "Old"), col=c(rgb(0.1,0.1,0.1,0.5), rgb(0.8,0.8,0.8,0.5)), lwd=10)
lines(density(muscle_old_skews),col='grey', lwd=5)
lines(density(muscle_new_skews),col='black', lwd=5)
#dev.off()
```

```{r}

index = order(age_tissue_auc_df$aucs_50)

par(mar = c(12,4,2,2))
barp = barplot(age_tissue_auc_df$aucs_50[index] , names = age_tissue_auc_df$tissue[index], main = 'Tissue AUC for predicting donor age (positives >= 50 years)', ylab = 'AUC', las=2, col = 'grey', cex.names=.65, ylim = c(0,1))
abline(h=.5, col = 'red')

par(mar = c(12,4,2,2))
barp = barplot(age_tissue_auc_df$num_donors[index] , names = age_tissue_auc_df$tissue[index], main = 'Number of donors per tissue', ylab = 'Num donors', las=2, col = 'grey', cex.names=.65, 
               ylim = c(0,160))

par(mar = c(12,4,2,2))
barp = barplot(age_tissue_auc_df$num_old_donors[index] , names = age_tissue_auc_df$tissue[index], main = 'Number of old donors per tissue', ylab = 'Num donors', las=2, col = 'grey', cex.names=.65, 
               ylim = c(0,160))


index = order(age_tissue_auc_df$aucs_60)

par(mar = c(12,4,2,2))
barp = barplot(age_tissue_auc_df$aucs_60[index] , names = age_tissue_auc_df$tissue[index], main = 'Tissue AUC for predicting donor age (positives >= 60 years)', ylab = 'AUC', las=2, col = 'grey', cex.names=.65, ylim = c(0,1))
abline(h=.5, col = 'red')

par(mar = c(12,4,2,2))
barp = barplot(age_tissue_auc_df$num_donors[index] , names = age_tissue_auc_df$tissue[index], main = 'Number of donors per tissue', ylab = 'Num donors', las=2, col = 'grey', cex.names=.65, 
               ylim = c(0,160))

par(mar = c(12,4,2,2))
barp = barplot(age_tissue_auc_df$num_old_donors[index] , names = age_tissue_auc_df$tissue[index], main = 'Number of old donors per tissue', ylab = 'Num donors', las=2, col = 'grey', cex.names=.65, 
               ylim = c(0,160))

```



```{r}
ranks_50_auc = rank(age_tissue_auc_df$aucs_50)
ranks_60_auc = rank(age_tissue_auc_df$aucs_60)

par(pty='s')
plot(ranks_50_auc, ranks_60_auc, main='Rank of tissue AUC for predicting donor age', xlab='Donor positives >= 50 years', ylab = 'Donor positives >= 60 years')
abline(a=0, b=1, col='red')

```




Look at the variance in skew per tissue
Did this previously in different ways, but just do similar bar plots as above to show the ranking of tissue skew variance

```{r}
tiss_skew_variance = vector(mode='numeric', length=length(tissues))

for(i in 1:length(tissues)){
  current_tiss = tissues[i]
  
  tiss_skew_variance[i] = var(skew_and_stats_df$skew[skew_and_stats_df$tissue == current_tiss])
  
}

index = order(tiss_skew_variance)
par(mar = c(12,4,2,2))
barp = barplot(tiss_skew_variance[index] , names = tissues[index], main = 'Tissue skew variance', ylab = 'skew variance', las=2, col = 'grey', cex.names=.65)

```




################################
Get a heatmap of tissues versus donor
Showing the tissues deviance from the donor's mean skew
################################

Empty matrix
```{r}
#Get the list of donors
all_donors = as.character(unique(skew_and_stats_df$donor))

tiss_deviance_df = as.data.frame(matrix(nrow = length(tissues), ncol = length(all_donors)))
rownames(tiss_deviance_df) = tissues
colnames(tiss_deviance_df) = all_donors
tiss_deviance_df

```


```{r}


for(i in 1:length(tissues)){
  current_tiss = tissues[i]
  
  
  #Get the donors for that tissue
  donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
  num_donors = length(donors_for_tiss) 
  if(num_donors == 0){next}

  #Go through the donors
  for(j in 1:num_donors){
  
    temp_donor = donors_for_tiss[j]
    if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] <=1 ){next} #Only one tissue for that donor
    #get the mean folded skew for the donor, withold the current tissue skew
    donor_mean_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue != current_tiss])
    
    #Get the skew of that tissue from that donor
    #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
    tiss_donor_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
    
    deviance = donor_mean_skew - tiss_donor_skew
    
    tiss_deviance_df[rownames(tiss_deviance_df) == current_tiss, colnames(tiss_deviance_df) == temp_donor] = deviance
    
    
  }
  
}

tiss_deviance_df

```





The direction of the deviance doesnt mean anything in the folded space, so just look at the absolute value



```{r}

pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/avg_tiss_deviance_from_donor_mean_skew_barplot.pdf', width=7, height=5)

sum_tiss_dev_vec = abs(rowMeans(tiss_deviance_df, na.rm = TRUE))

index = order(sum_tiss_dev_vec)
par(mar = c(12,4,2,2))
barp = barplot(sum_tiss_dev_vec[index] , names = tissues[index], main = 'Average tissue skew deviance from donor', ylab = 'Average skew deviance from donor skew', las=2, col = 'grey', cex.names=.65, cex.lab=.75)

dev.off()

sum_tiss_dev_vec = abs(rowMeans(tiss_deviance_df, na.rm = TRUE))
names(sum_tiss_dev_vec) = rownames(tiss_deviance_df)
index = order(sum_tiss_dev_vec)
par(mar = c(12,4,2,2))
barp = barplot(sum_tiss_dev_vec[index] , names = tissues[index], main = 'Average tissue skew deviance from donor', ylab = 'Average skew deviance from donor skew', las=2, col = 'grey', cex.names=.65, cex.lab=.75)
```




 Also look at the whole distributions of deviance



```{r}
deviance_plotting_df = melt(t(tiss_deviance_df))

colnames(deviance_plotting_df) = c('donor','tissue','deviance')

deviance_plotting_df

```

Add the germ layer for coloring

```{r}
sum(deviance_plotting_df$tissue == 'Adipose - Subcutaneous')

```

```{r}
germ_layer_column = c()
for( i in 1:length(tissues)){
  #Get the germ layer
  germ_layer = as.character(skew_and_stats_df$germ_layer[skew_and_stats_df$tissue == tissues[i]][1])
  rep(germ_layer, 317) 
  germ_layer_column = c(germ_layer_column,rep(germ_layer, 317) )
}

deviance_plotting_df$germ_layer = germ_layer_column 
deviance_plotting_df$deviance = abs(deviance_plotting_df$deviance)
deviance_plotting_df
```




```{r}


deviance_plotting_df$tissue = with(deviance_plotting_df, reorder(tissue, deviance, var, na.rm =TRUE))

ggplot(deviance_plotting_df,  aes(x=tissue, y=deviance, fill = germ_layer)) + 
  geom_boxplot( outlier.size = .1,) +
  geom_point(position = 'jitter', size = .1) +
  scale_fill_manual(values = c(ecto_color, endo_color, meso_color)) +
  ylab("Deviance from donor mean skew") + ggtitle('Tissue distributions for skew deviance from donor skew') + xlab('')+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,face="bold", size=8), 
        plot.title=element_text(hjust=.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        axis.text.y = element_text(face="bold", size=12), 
        plot.margin = margin(.1, .1, .1, .1, 'in'))
  
ggplot(deviance_plotting_df,  aes(x=tissue, y=deviance, fill = germ_layer)) + 
  geom_violin(scale = 'width') +
  #geom_point(position = 'jitter', size = .05) +
  scale_fill_manual(values = c(ecto_color, endo_color, meso_color)) +
  ylab("Deviance from donor mean skew") + ggtitle('Tissue distributions for skew deviance from donor skew') + xlab('')+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,face="bold", size=8), 
        plot.title=element_text(hjust=.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        axis.text.y = element_text(face="bold", size=12), 
        plot.margin = margin(.1, .1, .1, .1, 'in'), 
        panel.border = element_rect(colour = "black", fill=NA, size=1), panel.background = element_rect(fill = 'white')) +
  geom_hline(yintercept = .1, color = 'red',linetype = 'dashed', size = .75)+
  facet_wrap(~germ_layer , scales = 'free_x')
ggsave(filename = 'tissue_dev_from_donor_mean_skew_violin_plots_v5.pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/', device = 'pdf',
       dpi = 300, width = 7, height = 5, useDingbats = FALSE )



```



############################################
Trying out the tissue lineage specific stemcell population estimates

Skew of tissue will follow a binomial distribution where N = num_cells_tiss and p = skew of the donor (estimating it as the mean skew)
But, each instance of a tissue will becoming from a different donor who will have a different P. 
What we want is the deviance from the mean, the variance there is indicative of the num_cells_tiss, 
So, turn everything into a Z-score and solve for the standard normal distribution

Remember SD of the skews is sqrt( p*q / N )
p  = mean donor skew
N = num_cells_tiss

Zi = skew_i - p / sqrt( p*q / N )  

So the donor mean is going to be different, but N is a constant. And we know the SD of zscores is equal to 1. So just solve for N setting the SD(z-scores) to 1

Rewrite zscores:  zi = ti*sqrt(N) where ti = skew_i - p / sqrt(pq)

SD(ti*sqrt(N)) = 1

so
sqrt( 1/m-1 * sum( (zi - mean(z)) ^2) ) = 1
taking out the sqrt(N) term in the zi and z_bar terms gets

sqrt( 1/m-1 * N * sum( (ti - mean(t)) ^2) ) = 1

Solving for N gets  N = m-1 / sum( (ti - mean(t)) ^2) where m is the number of tissue samples

Adding an error term SD(zscores) = 1 + e
simplifies to 
N = ( (1+e)^2 * (m-1) )  / sum( (ti - mean(t)) ^2)

Can simulate/bootstrap error to get confidence intervals around the N estimate


###############################################


Trying to model the error of sampling z-scores from a population
Over sample in the simulation to get a true (close to) disribution of error
The choice of N and p doesnt matter when simulating the error. 
The sample size is what's going to cause variation in the error
Can check below

```{r}
n = 25
p1 = .25
p2 = .5
p3 = .75
mod_skews_1 = rbinom(10000, n, p1) / n
z_skews_1 = ((mod_skews_1 - p1) * sqrt(n)) / sqrt(p1*(1-p1))
mod_skews_2 = rbinom(1000000, n, p2) / n
z_skews_2 = ((mod_skews_2 - p2) * sqrt(n)) / sqrt(p2*(1-p2))
mod_skews_3 = rbinom(1000000, n, p3) / n
z_skews_3 = ((mod_skews_3 - p3) * sqrt(n)) / sqrt(p3*(1-p3))

error = c()
for( i in 1:10000){
  z_skew_sample = sample(z_skews_3, size = 300, replace = TRUE)
  #z_skew_sample = c(z_skew_sample, sample(z_skews_2, size = 10, replace = TRUE) )
  #z_skew_sample = c(z_skew_sample, sample(z_skews_3, size = 10, replace = TRUE) )
  error = c(error, 1 - sd(z_skew_sample))
}

hist(error)
max(error)
sample(error, size = 30, replace = TRUE)



```


Function to derive a z-score error distribution for a set sample size

```{r}

zscore_error_dist = function(sample_size){
  n=16
  p = .75
  #Get a simulated distribution of skews
  model_skews = rbinom(10000, n, p) / n
  #Convert to zscores
  z_skews = ((model_skews - p) * sqrt(n)) / sqrt(p*(1-p))
  
  error = c()
  #Oversample when building the error distribution
  #But use the tissue sample size when sampling the zskews
  for( i in 1:10000){
    z_skew_sample = sample(z_skews, size = sample_size, replace = TRUE)
    #expectation is that the sd of the z scores will be 1
    error = c(error, 1 - sd(z_skew_sample))
  }
  #Return the error distribution
  return(error)
}


```

Function to bootstrap the actual t values

```{r}

bootstrap_t_values = function(t_vec){
  m = length(t_vec)
  simulation = sample(t_vec, size = m, replace = TRUE)
  t_bar_sim = mean(simulation)
  N = (m-1) / sum((simulation - t_bar_sim)^2)
  return(N)
}


```


Looking at whether folding/unfolding makes a difference. Wouldn't folding cause a bias for positive deviations, a zscore distribution not centered at 0???
Answer is yes, unfolding makes more sense

```{r}

current_tiss = tissues[1]
#Get the donors for that tissue
donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
num_donors = length(donors_for_tiss)
t_vec = vector(mode = 'numeric', length = num_donors)

#Go through the donors
for(j in 1:num_donors){

  temp_donor = donors_for_tiss[j]
  if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] <=1 ){next} #Only one tissue for that donor
  #get the mean folded skew for the donor, withold the current tissue skew
  donor_mean_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue != current_tiss])
  
  #Get the skew of that tissue from that donor
  #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
  tiss_donor_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
  
  #Get the t value
  t = (tiss_donor_skew - donor_mean_skew) / (sqrt(donor_mean_skew * (1-donor_mean_skew)))
  t_vec[j] = t

}

#Get the N estimate without error simulation first
m = length(t_vec)
t_bar = mean(t_vec)
N_folded = (m-1) / sum((t_vec - t_bar)^2)
title = sprintf('Estimated N from folded: %f', N_folded)
hist(t_vec, main = title)
abline(v = t_bar, col = 'red')

z_vec = t_vec * sqrt(N_folded)
sd(z_vec)
hist(z_vec, main = 'Zscored values using folded data')


#Get the donors for that tissue
donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
num_donors = length(donors_for_tiss)
t_vec = vector(mode = 'numeric', length = num_donors)
t_vec_unfolded = vector(mode = 'numeric', length = num_donors)
#Go through the donors
for(j in 1:num_donors){

  temp_donor = donors_for_tiss[j]
  if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] <=1 ){next} #Only one tissue for that donor
  #get the mean folded skew for the donor, withold the current tissue skew
  donor_mean_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue != current_tiss])
  
  #Get the skew of that tissue from that donor
  #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
  tiss_donor_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
  #Get the t value
  t = (tiss_donor_skew - donor_mean_skew) / (sqrt(donor_mean_skew * (1-donor_mean_skew)))
  t_vec[j] = t

}

#Unfolding the results is the same as multiplying by -1
t_vec = c(t_vec, -1*t_vec)

#Get the N estimate without error simulation first
m = length(t_vec)
t_bar = mean(t_vec)
N_unfolded = (m-1) / sum((t_vec - t_bar)^2)
title = sprintf('Estimated N from unfolded: %f', N_unfolded)
hist(t_vec, main = title)
abline(v = t_bar, col = 'red')

z_vec = t_vec * sqrt(N_unfolded)
sd(z_vec)
hist(z_vec, main = 'Zscored values using unfolded data')

```





```{r}
num_bootstraps = 2000

ci = 95
ci_end = ((100-ci) / 2) / 100

n_vec = vector(mode = 'numeric', length = length(tissues))
num_samples_per_tiss = vector(mode = 'numeric', length = length(tissues))
upper_ci_vec_error = vector(mode = 'numeric', length = length(tissues))
lower_ci_vec_error = vector(mode = 'numeric', length = length(tissues))
upper_ci_vec_skew_boot = vector(mode = 'numeric', length = length(tissues))
lower_ci_vec_skew_boot = vector(mode = 'numeric', length = length(tissues))

for(i in 1:length(tissues)){

  
  current_tiss = tissues[i]
  #Get the donors for that tissue
  donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
  num_donors = length(donors_for_tiss)
  t_vec = vector(mode = 'numeric', length = num_donors)
  if(num_donors == 0){next}
  
  #Go through the donors
  for(j in 1:num_donors){
  
    temp_donor = donors_for_tiss[j]
    if(dim(skew_and_stats_df[skew_and_stats_df$donor == temp_donor, ])[1] <=1 ){next} #Only one tissue for that donor
    #get the mean folded skew for the donor, withold the current tissue skew
    donor_mean_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue != current_tiss])
    
    #Get the skew of that tissue from that donor
    #So there are a few tissues that have multiple samples from the same donor. Just averaging the skew estimates at the moment.
    tiss_donor_skew = mean(skew_and_stats_df$skew[skew_and_stats_df$donor == temp_donor & skew_and_stats_df$tissue == current_tiss])
    
    #Get the t value
    t = (tiss_donor_skew - donor_mean_skew) / (sqrt(donor_mean_skew * (1-donor_mean_skew)))
    t_vec[j] = t
  
  }
   
  #Unfold the results
  t_vec = c(t_vec, -1*t_vec)
  
  #Get the N estimate without error simulation first
  m = length(t_vec)
  t_bar = mean(t_vec)
  N = (m-1) / sum((t_vec - t_bar)^2)
  n_vec[i] =  N
  
  t_vec_bootstrap_n = unlist(mclapply(1:num_bootstraps, function(i) bootstrap_t_values(t_vec) ))
  t_vec_bootstrap_n = t_vec_bootstrap_n[order(t_vec_bootstrap_n)]
  n_lower_ci = t_vec_bootstrap_n[num_bootstraps * (1 - (1 - ci_end))]
  n_upper_ci = t_vec_bootstrap_n[num_bootstraps *  (1 - ci_end)]
  lower_ci_vec_skew_boot[i] = n_lower_ci
  upper_ci_vec_skew_boot[i] = n_upper_ci 
 
  #Save the number of samples per tissue
  num_samples_per_tiss[i] =  m
  
  #And look at the plot of the supposed Z-scores, ti*sqrt(N)
  z_scores = t_vec * sqrt(N)
  title = sprintf('%s z_scores, N: %.3f , SD: %.3f', current_tiss, N, sd(z_scores))
  hist(z_scores, xlim = c(-4, 4), main = title, breaks = 16)
  
  title = sprintf('%s t values', current_tiss)
  hist(t_vec, xlim = c(min(t_vec)-.1, max(t_vec)+.1), main = title, breaks = 16)
  
  #Now get an error distribution and bootstrap from there to get CI for the N estimate
  #error_dist = zscore_error_dist(m)
  
  #Now bootstrap 2000 times error sampling to get a distribution of N estimates to get CI from
  #Random sample error from that distribution
  #error_sample = sample(error_dist, size = num_bootstraps, replace = TRUE)
  
  #n_est_dist_error = unlist(mclapply(1:length(error_sample), function(i) (((1+error_sample[i])^2) * (m-1)) / sum((t_vec - t_bar)^2), mc.cores = 5))
  #n_est_dist_error = n_est_dist_error[order(n_est_dist_error)]
  #n_lower_ci = n_est_dist_error[num_bootstraps * (1 - (1 - ci_end))]
  #n_upper_ci = n_est_dist_error[num_bootstraps *  (1 - ci_end)]
  #lower_ci_vec_error[i] = n_lower_ci
  #upper_ci_vec_error[i] = n_upper_ci
}

  
#Get the germlayers for the filtered tissues
germ_layers = vector(mode='character', length=length(tissues))

for( i in 1:length(tissues)){
  index = which(skew_and_stats_df$tissue == tissues[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
}


est_tissue_lineage_cell_num_df = data.frame(est_n = n_vec, upper_ci_skew = upper_ci_vec_skew_boot, lower_ci_skew = lower_ci_vec_skew_boot,
                                            #upper_ci_error = upper_ci_vec_error, lower_ci_error = lower_ci_vec_error, 
                                  tissues = tissues, num_samples = num_samples_per_tiss, germ_layers = germ_layers)

rownames(est_tissue_lineage_cell_num_df) = tissues


#Group by estimated number
est_tissue_lineage_cell_num_df = est_tissue_lineage_cell_num_df[order(est_tissue_lineage_cell_num_df$est_n), ]


#And reset the factor levels for the tissues
est_tissue_lineage_cell_num_df$tissues = factor(est_tissue_lineage_cell_num_df$tissues, levels = est_tissue_lineage_cell_num_df$tissues)
est_tissue_lineage_cell_num_df 
```


```{r}
est_tissue_lineage_cell_num_df

```

```{r}
deviance_plotting_df

```


Also replot the average deviance from donor skew per tissue, matching the order of the cell num estimates

```{r}
cell_line_index = !est_tissue_lineage_cell_num_df$tissues %in% c('Cells - Cultured fibroblasts','Cells - EBV-transformed lymphocytes','Cells - Transformed fibroblasts')
tiss_to_match = as.character(est_tissue_lineage_cell_num_df[cell_line_index,'tissues' ])


tiss_ordered_by_dev_var = levels(deviance_plotting_df$tissue)

var_dev_df = deviance_plotting_df %>%
  group_by(tissue) %>%
  summarise_at(vars(deviance), funs(var(., na.rm=TRUE)))


tiss_ordered_by_dev_var = levels(var_dev_df$tissue)
index = match(tiss_to_match,tiss_ordered_by_dev_var )

var_dev_df = var_dev_df[index, ]
var_dev_df
```

```{r}
table(as.character(est_tissue_lineage_cell_num_df[cell_line_index,'tissues' ]) == as.character(var_dev_df$tissue))


```



```{r}
#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/variance_in_tiss_deviance_from_donor_skew.pdf', height=5, width=7, useDingbats = FALSE )
barplot(var_dev_df$deviance, names = as.character(var_dev_df$tissue), ylab = 'Variance in skew deviation from mean donor skew', las=2, col = 'grey', cex.names=.65, cex.lab=.75)
#dev.off()
```




```{r}
cell_line_index = !est_tissue_lineage_cell_num_df$tissues %in% c('Cells - Cultured fibroblasts','Cells - EBV-transformed lymphocytes','Cells - Transformed fibroblasts')


ggplot(est_tissue_lineage_cell_num_df[cell_line_index, ], aes(x=tissues, y=est_n)) + geom_point( size = 2) + 
  geom_errorbar(aes(x = tissues, ymin = lower_ci_skew, ymax = upper_ci_skew),size=.5, alpha = .5) +
  ggtitle('Tissue specific cell estimates') + ylab('# cells at time of tissue specification') + xlab('Tissues') +
  scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
        axis.text.y = element_text(face="bold", size=12))

ggsave(filename='tissue_specific_stem_cell_pop_ests_with_ci.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/stem_cell_pop_estimates/', units='in', height=5, width=7, dpi=300, useDingbats=FALSE)

#ggplot(est_tissue_lineage_cell_num_df, aes(x=tissues, y=est_n, color=germ_layers)) + geom_point() + 
#  geom_errorbar(aes(x = tissues, ymin = lower_ci_error, ymax = upper_ci_error)) +
#  ggtitle('CI: bootstrapping with a simulated error distribution') + ylab('# cells at time of tissue specification') + xlab('Tissues') +
#  scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color)) +
#  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
#        plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
#        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
#        panel.background = element_rect(fill = 'white'), panel.border = element_rect(color = "black", fill=NA, size=1),  
#        axis.text.y = element_text(face="bold", size=12))
```

Save cell number estimates


```{r}

save(est_tissue_lineage_cell_num_df, 
     file = '/home/werner/xchrom_snp_skew/data/GTEx/snp_skews/GATK_mod/cell_num_estimates/tissue_specific_lineage_cell_num_ests_df.Rdata')

```




Simulated data examples for a schematic explaining the above

```{r}
N_1 = 30
p_25 = .25
p_5 = .5
p_75 = .75
p_9 = .9
skews = (0:N_1)/N_1

pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tissue_specific_cell_num_ests_model/small_n_binomials.pdf', height = 4, width = 4, useDingbats = F)
probs_25 = dbinom(x = 0:N_1, size = N_1, prob = p_25)
probs_5 = dbinom(x = 0:N_1, size = N_1, prob = p_5)
probs_75 = dbinom(x = 0:N_1, size = N_1, prob = p_75)
probs_9 = dbinom(x = 0:N_1, size = N_1, prob = p_9)
plot(skews, probs_25, type ='l', lwd=2, pch=16, ylab = 'probability', xlab = 'Tissue skews', main = 'Small N')
points(skews, probs_5, type ='l', lwd=2, pch=16, col = 'red')
points(skews, probs_75, type ='l', lwd=2, pch=16, col = 'blue')
dev.off()


pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tissue_specific_cell_num_ests_model/large_n_binomials.pdf', height = 4, width = 4, useDingbats = F)
N_1 = 100
skews = (0:N_1)/N_1
probs_25 = dbinom(x = 0:N_1, size = N_1, prob = p_25)
probs_5 = dbinom(x = 0:N_1, size = N_1, prob = p_5)
probs_75 = dbinom(x = 0:N_1, size = N_1, prob = p_75)
plot(skews, probs_25, type ='l', lwd=2, pch=16, ylab = 'probability', xlab = 'Tissue skews', main = 'Large N')
points(skews, probs_5, type ='l', lwd=2, pch=16, col = 'red')
points(skews, probs_75, type ='l', lwd=2, pch=16, col = 'blue')
dev.off()

pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tissue_specific_cell_num_ests_model/standard_normal.pdf', height = 4, width = 4, useDingbats = F)
par(pty = 's')
x=seq(-4,4,length=200)
y=dnorm(x,mean=0,sd=1)
plot(x,y,type="l",lwd=2,col="black", xlab = 'z score', ylab = 'density', main = 'Standard normal distribution')
dev.off()

```








################################################
Testing tissue Xs ability to predict tissue Ys skew
If skew is generally the same within a donor, than most tissues will be able to accurately predict other tissues skew
###############################################



One important thing will be to settle on a reasonable threshold for determining meaningful tiss_x predicts tiss_y AUCs

-Several things to consider. First look at the CI distribution as it relates to skew, could pick a skew threshold where the CIs are strong?
-The other is to do the AUC analysis across all thresholds, and pool together all the AUCs from all the pairwise comparisons, see if there is a clear cutoff empirically



Starting with the CI distribution
```{r}

plot(skew_and_stats_df$CI_width ,skew_and_stats_df$skew)


plot(skew_and_stats_df$skew ,skew_and_stats_df$CI_width)


#Okay, slightly concerning. In general CI ranges from 0 to 1.5, but there's a separate distribution of skews from .65 to .8 that are the unconfident ones
#These CIs were derived from bootstrap resampling of the SNP skew distributions, so really a proxy for how variable the SNP skews were for a sample.
#Could filter out the skews with CI >= .2 regardless, probs a good idea


```
 




Redo the tiss_tiss_donors_df and the tiss_tiss_donors_names_matrices for the filtered tissues
```{r}
filt_tissues
tissues = filt_tissues

```


```{r}
tiss_tiss_donors_df = as.data.frame(matrix(nrow = length(tissues), ncol = length(tissues)))
rownames(tiss_tiss_donors_df) = tissues
colnames(tiss_tiss_donors_df) = tissues

tiss_tiss_donors_names_matrix = matrix(rep(list(), length(tissues)*length(tissues)), nrow=length(tissues), ncol=length(tissues))
rownames(tiss_tiss_donors_names_matrix) = tissues
colnames(tiss_tiss_donors_names_matrix) = tissues

for(i in 1:length(tissues)){
  
  current_tiss = tissues[i]
  
  for(j in 1: length(tissues)){
    
    next_tiss = tissues[j]
    
    donors_for_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == current_tiss])
    #Just going to ignore the donors that have multiple donations for the same tiss
    duplicates = donors_for_tiss[duplicated(donors_for_tiss)] 
    donors_for_tiss = donors_for_tiss[!donors_for_tiss %in% duplicates]  
    
    donors_for_next_tiss = as.character(skew_and_stats_df$donor[skew_and_stats_df$tissue == next_tiss])
    duplicates = donors_for_next_tiss[duplicated(donors_for_next_tiss)] 
    donors_for_next_tiss = donors_for_next_tiss[!donors_for_next_tiss %in% duplicates]
    
    tiss_tiss_donors_df[i,j] = sum(unique(donors_for_tiss) %in% unique(donors_for_next_tiss))
    tiss_tiss_donors_names_matrix[i,j] = list(unique(donors_for_tiss[donors_for_tiss %in% donors_for_next_tiss]))
  }
}

tiss_tiss_donors_df
tiss_tiss_donors_names_matrix

```



Get a binary matrix too for which donors donate which tissues

```{r}
diagonal = diag(tiss_tiss_donors_names_matrix)


binary_tissue_donor_matrix = matrix(nrow = length(names(diagonal)), ncol = length(donors))
rownames(binary_tissue_donor_matrix) = names(diagonal)
colnames(binary_tissue_donor_matrix) = donors

for(i in 1:length(names(diagonal))){
  
  donors_for_tiss = diagonal[[i]]
  index = colnames(binary_tissue_donor_matrix) %in% donors_for_tiss
  binary_tissue_donor_matrix[i, index] = 1
  binary_tissue_donor_matrix[i, !index] = 0
}


binary_tissue_donor_matrix = binary_tissue_donor_matrix[!rownames(binary_tissue_donor_matrix) %in% c('Testis','Prostate','body_site_s'), ]


```



For permutation test and bootstrap CI of pearson correlations

```{r}


bootstrap_corrs_func = function(x,y){
  index = 1:length(x)
  bootstrap = sample(index, size = length(x), replace=TRUE)
  corr = suppressWarnings(cor(x[bootstrap], y[bootstrap], method='pearson'))
  return(corr)
}


cor_permute = function(x,y){
  permute_y = sample(y,size=length(y), replace=FALSE)
  corr = suppressWarnings(cor(x,permute_y, method='pearson'))
  return(corr)
}




```


Getting a correlation threshold from the null distribution of correlations using a sample size of 20

```{r}
poss_skews = seq(.5, 1, .001)
x = sample(poss_skews, 20, replace = TRUE)
x
```


```{r}
sig_correlations = c()
num_bootstraps = 1000

for(i in 1:num_bootstraps){
  num_permutes = 10000
  #x =c(.5,.55,.6,.65,.7,.75,.8,.85,.9,.95,.5,.55,.6,.65,.7,.75,.8,.85,.9,.95)
  poss_skews = seq(.5, 1, .001)
  x = sample(poss_skews, 20, replace = TRUE)
  y = x
  permute_cors = unlist(mclapply(1:num_permutes, function(i) cor_permute(x,y), mc.cores=5))
  
  order_permute_cors = abs(permute_cors)
  order_permute_cors = order_permute_cors[order(order_permute_cors)]
  #hist(order_permute_cors)
  sig_correlation = order_permute_cors[num_permutes*.95]

  sig_correlations = c(sig_correlations, sig_correlation)

}


hist(sig_correlations)
mean(sig_correlations)
correlation_threshold = mean(sig_correlations)
```



```{r}

skewed_donors = names(mean_skews)[(mean_skews>=.75)]
skewed_donors 
```



Make a matrix to hold all the AUCs for all thresholds

Threshold on the ranks of the skews for tissue_2 rather than a sequence of skews


```{r}

tiss_tiss_paired_wilcox_pvals_matrix = matrix(nrow = length(tissues), ncol = length(tissues))
rownames(tiss_tiss_paired_wilcox_pvals_matrix) = tissues
colnames(tiss_tiss_paired_wilcox_pvals_matrix) = tissues

tiss_tiss_7_aucs_matrix = matrix(nrow=length(tissues), ncol=length(tissues))
rownames(tiss_tiss_7_aucs_matrix) = tissues
colnames(tiss_tiss_7_aucs_matrix) = tissues

binned_tiss_tiss_7_aucs_matrix = matrix(nrow=length(tissues), ncol=length(tissues))
rownames(binned_tiss_tiss_7_aucs_matrix) = tissues
colnames(binned_tiss_tiss_7_aucs_matrix) = tissues

tiss_tiss_correlation_matrix = matrix(nrow=length(tissues), ncol=length(tissues))
rownames(tiss_tiss_correlation_matrix) = tissues
colnames(tiss_tiss_correlation_matrix) = tissues

tiss_tiss_corr_pvalues = matrix(nrow=length(tissues), ncol=length(tissues))
rownames(tiss_tiss_corr_pvalues) = tissues
colnames(tiss_tiss_corr_pvalues) = tissues

tiss_tiss_corr_lowerCI = matrix(nrow=length(tissues), ncol=length(tissues))
rownames(tiss_tiss_corr_lowerCI) = tissues
colnames(tiss_tiss_corr_lowerCI) = tissues

tiss_tiss_corr_upperCI = matrix(nrow=length(tissues), ncol=length(tissues))
rownames(tiss_tiss_corr_upperCI) = tissues
colnames(tiss_tiss_corr_upperCI) = tissues

num_permutes = 10000  #For correlation permutation test
num_bootstraps = 10000 #For correlation bootstrapping CI


#Keep all the AUCs with their corresponding skew threshold from the tiss_2, for plotting
all_aucs = c()
all_skews = c()


for(i in 1:length(tissues)){
  
  tiss_1 = rownames(tiss_tiss_donors_names_matrix)[i]

  for(j in 1:length(tissues)){

    tiss_2 = colnames(tiss_tiss_donors_names_matrix)[j]
    test_donors = tiss_tiss_donors_names_matrix[[i,j]]
    
    #Exclude the donors that have high mean skew, see how that affects the correlations
    #test_donors = test_donors[!test_donors %in% skewed_donors]
    
    if(length(test_donors) <= 1){next}
    #Make sure to reorder the skews so the donor order matches the test_donor order. Ensures that the order of skew for both tissues is the same by donor
    tiss_1_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
    tiss_1_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
    index_1 = match(test_donors, tiss_1_donors)
    tiss_1_skews = tiss_1_skews[index_1]
    
    
    tiss_2_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
    tiss_2_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
    index_2 = match(test_donors, tiss_2_donors)
    tiss_2_skews = tiss_2_skews[index_2]
    
    #Paired wilcoxen test
    paired_wilcox_test = suppressWarnings(wilcox.test(tiss_1_skews, tiss_2_skews, paired=TRUE, alternative = 'two.sided'))
    tiss_tiss_paired_wilcox_pvals_matrix[i,j] = paired_wilcox_test$p.value
    
  
    #Testing all the possible thresholds for AUC analysis, just really for looking at the results of one tissue
    label_thresh = rank(tiss_2_skews)
    auc_vector = vector(mode='numeric', length=length(label_thresh))
    
    for(k in 1:length(label_thresh)){
      labels = label_thresh > k    #Thresholding on the rank of the skews for tissue_2, positives are the upper ranks  
      auc_vector[k] = auc_wcrst(labels, tiss_1_skews)
    }
    
    #Keep all the AUCs and their skew thresholds, for plotting
    
    #Order the tiss_skews, will match the order of the ranks, to plot the corresponding AUC correctly
    ordered_tiss_2_skews = tiss_2_skews[order(tiss_2_skews)]
    
    all_aucs = c(all_aucs, auc_vector)
    all_skews = c(all_skews,ordered_tiss_2_skews)
    
    if(i == 1){
    
      #Visualize for the first row of comparisons
      x_title = sprintf("%s skew threshold for +'s",tiss_2)
      y_title = sprintf("%s AUC for predicting +'s", tiss_1)
      #print(plot(ordered_tiss_2_skews, auc_vector, ylim=c(0,1), xlim = c(.5,1), xlab = x_title, ylab = y_title))
    }
    
    #Also get the pearson correlation
    original_cor = cor(tiss_1_skews, tiss_2_skews, method = 'pearson')
    if(original_cor <= .7 & original_cor >= .5){
      y_title = sprintf("%s skews",tiss_2)
      x_title = sprintf("%s skews", tiss_1)
      #par(pty = 's')
      #print(plot(tiss_1_skews, tiss_2_skews, xlim = c(.5,1), ylim = c(.5,1), xlab = x_title, ylab = y_title))
      
    }
    tiss_tiss_correlation_matrix[i,j] = original_cor
    
    #Do permutation and bootstrapping to get a correlation pvalue and CI
    permute_cors = unlist(mclapply(1:num_permutes, function(i) cor_permute(tiss_1_skews,tiss_2_skews), mc.cores=5))
    tiss_tiss_corr_pvalues[i,j] = sum(abs(permute_cors) > abs(original_cor) ) / num_permutes
    
    #bootstrap_corrs = unlist(mclapply(1:num_bootstraps, function(i) bootstrap_corrs_func(tiss_1_skews,tiss_2_skews), mc.cores=5))
    #tiss_tiss_corr_lowerCI[i,j] = quantile(bootstrap_corrs, .025, names=FALSE, na.rm=TRUE)
    #tiss_tiss_corr_upperCI[i,j] = quantile(bootstrap_corrs, .975, names=FALSE, na.rm=TRUE)   
  
    
    #Save the AUC from a .7 skew threshold for all tissues
    labels = tiss_2_skews >= .7
    auc_7 = auc_wcrst(labels, tiss_1_skews)
    tiss_tiss_7_aucs_matrix[i,j] = auc_7
    
    #Now do AUC for a binned skew, where skews <= .6 for either tissue go to .5
    tiss_1_skews[tiss_1_skews <=.6] = .55
    tiss_2_skews[tiss_2_skews <=.6] = .55
    
    labels = tiss_2_skews >= .7
    auc_7 = auc_wcrst(labels, tiss_1_skews)
    binned_tiss_tiss_7_aucs_matrix[i,j] = auc_7    
    
      
  }
  print(sprintf('Done with %s at %s', tiss_1, Sys.time()))
  
}



```





Save the correlations for later

```{r}
save(tiss_tiss_correlation_matrix, tiss_tiss_corr_pvalues, tiss_tiss_corr_lowerCI, tiss_tiss_corr_upperCI, 
     file = '/home/werner/xchrom_snp_skew/data/GTEx/snp_skews/GATK_mod/v8_pairwise_tissue_skew_correlations.Rdata')
```

```{r}
load(file = '/home/werner/xchrom_snp_skew/data/GTEx/snp_skews/GATK_mod/v8_pairwise_tissue_skew_correlations.Rdata')
```


```{r}
tiss_tiss_correlation_matrix['Colon - Sigmoid', ]

```


Pull out within and across germ layer comparisons for demonstration
within: Colon - Sigmoid and Esophagus - Gastroesophageal Junction
across: Nerve - Tibial and Adipose - Subcutaneous (ectoderm, mesoderm)


```{r}

for(i in 1:length(tissues)){
  
  tiss_1 = rownames(tiss_tiss_donors_names_matrix)[i]

  for(j in 1:length(tissues)){

    tiss_2 = colnames(tiss_tiss_donors_names_matrix)[j]
    test_donors = tiss_tiss_donors_names_matrix[[i,j]]

    if(length(test_donors) <= 1){next}
    #Make sure to reorder the skews so the donor order matches the test_donor order. Ensures that the order of skew for both tissues is the same by donor
    tiss_1_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
    tiss_1_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
    index_1 = match(test_donors, tiss_1_donors)
    tiss_1_skews = tiss_1_skews[index_1]
    
    tiss_2_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
    tiss_2_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
    index_2 = match(test_donors, tiss_2_donors)
    tiss_2_skews = tiss_2_skews[index_2]
    
    #Also get the pearson correlation
    original_cor = cor(tiss_1_skews, tiss_2_skews, method = 'pearson')
    #Save plot of tissue skews
    file_tiss_1 = gsub('-', '_', tiss_1)
    file_tiss_1 = gsub(' ', '_', file_tiss_1)
    file_tiss_2 = gsub('-', '_', tiss_2)
    file_tiss_2 = gsub(' ', '_', file_tiss_2)
    file_name = sprintf('/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/all_pairwise_tiss_skew_comparisons/%s_and_%s.pdf', file_tiss_1, file_tiss_2)

    pdf(file_name, height = 4, width = 4, useDingbats = F)
    par(pty = 's')
    plot(tiss_1_skews, tiss_2_skews, xlim = c(.5,1), ylim = c(.5,1), xlab = sprintf('%s XCI skew',tiss_1), ylab = sprintf('%s XCI skew',tiss_2), pch = 19, cex = .75, 
         main = sprintf('correlation %0.3f', original_cor))
    dev.off()
    
  }
}

```

Look at the ones that have high correlations


```{r}

for(i in 1:length(tissues)){
  
  tiss_1 = tissues[i]
  tiss_corrs = tiss_tiss_correlation_matrix[tiss_1, ]
  high_corr_index = which(tiss_corrs >= .65)
  good_tiss = tissues[high_corr_index]
  for(j in 1:length(good_tiss)){
    
    print(sprintf('%s and %s', tiss_1, good_tiss[j]))
    
  }
  
}


```


```{r}
vec = c('A','A','A','B','B','B','C','C','C')
m = matrix(vec, nrow = 3, ncol = 3)
m

c(m)

matrix(c(m), 3, 3)

```


Correct all pvalues
```{r}
unit = dim(tiss_tiss_corr_pvalues)[1]
vec_pvals = c(tiss_tiss_corr_pvalues)
vec_pvals = p.adjust(vec_pvals, method = 'BH')
matrix_pvals = matrix(vec_pvals, unit, unit)
non_sig_corrs_index = matrix_pvals > .05

```

```{r}
#Filter on the significant correlation threshold determined from the null distribution of correlations with sample size 20
index = abs(tiss_tiss_correlation_matrix) < correlation_threshold  #Nonsignificant correlations under the null of no dependence
temp_corr_matrix = tiss_tiss_correlation_matrix
temp_corr_matrix[index] = NA
#Filter out the comparisons wih only a handful of donors
index = tiss_tiss_donors_df < 20
temp_corr_matrix[index] = NA

for(i in 1:length(rownames(temp_corr_matrix))){
  
  correlations = temp_corr_matrix[i, ]
  if(sum(!is.na(correlations))<1){next}
  tissue_name = rownames(temp_corr_matrix)[i]
  hist(correlations, xlim=c(-1,1), main = sprintf('%s skew correlations to other tissues', tissue_name), breaks=16)
  
}

```

```{r}
!is.na(temp_corr_matrix) 
#colSums(!is.na(temp_corr_matrix))[colSums(!is.na(temp_corr_matrix)) < 10]

```



Resetting, not filtering on Pvalue, just data number

```{r}
temp_corr_matrix = tiss_tiss_correlation_matrix
exclude = c('Prostate', 'body_site_s', 'Testis') #Male tissues, no data
#Ignore comparisons that have less than 20 data points
index = tiss_tiss_donors_df < 20
temp_corr_matrix[index] = NA
#Exclude any tissue that has less than 10 well powered comparisons
no_power_tissues = names(colSums(!is.na(temp_corr_matrix))[colSums(!is.na(temp_corr_matrix)) < 10])
exclude = c(exclude, no_power_tissues)

temp_corr_matrix = temp_corr_matrix[!rownames(temp_corr_matrix) %in% exclude,!colnames(temp_corr_matrix) %in% exclude ]
rownames(temp_corr_matrix)
```




```{r}
#Get germ layers for the filt_tissues

germ_layers = vector(mode='character', length=length(rownames(temp_corr_matrix)))

for( i in 1:length(rownames(temp_corr_matrix))){
  index = which(skew_and_stats_df$tissue == rownames(temp_corr_matrix)[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
  
}

germ_layers

```


CLustering the correlations
```{r}
temp_corr_matrix = tiss_tiss_correlation_matrix
temp_corr_matrix = temp_corr_matrix[!rownames(temp_corr_matrix) %in% exclude,!colnames(temp_corr_matrix) %in% exclude ]

germ_layers = vector(mode='character', length=length(rownames(temp_corr_matrix)))

for( i in 1:length(rownames(temp_corr_matrix))){
  index = which(skew_and_stats_df$tissue == rownames(temp_corr_matrix)[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
  
}


germ_layer_annot = HeatmapAnnotation(germ_layer = germ_layers, col = list(germ_layer=c('Ectoderm'=ecto_color, 'Mesoderm'=meso_color, 'Endoderm' = endo_color, 'NA' = 'white')))


#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/tissue_tissue_correlations_binned_unconf_skews_11_5_20_skews_escape_filt.pdf', height = 7, width=10)


#col_fun = colorRamp2(c(-1,-correlation_threshold,correlation_threshold, 1), c("blue" ,"white",'white', "red"))
col_fun = colorRamp2(c(-1,0, 1), c("blue",'white', "red"))

Heatmap(temp_corr_matrix, name='Pearson', show_row_dend = TRUE, col = col_fun, use_raster = TRUE,
        clustering_distance_rows = "pearson", clustering_distance_columns = "pearson",
        clustering_method_rows = "ward.D",clustering_method_columns = "ward.D",
        column_names_gp = gpar(fontsize=8),row_names_gp = gpar(fontsize=8), na_col = 'black', top_annotation = germ_layer_annot )

#dev.off()

```

Plot the heatmaps for intra and inter germ layer comparisons

```{r}
germ_layer_colors

```



```{r}
#col_fun_2 = colorRamp2(c(-1,-correlation_threshold,correlation_threshold, 1), c("blue" ,"white",'white', "red"))

col_fun_2 = colorRamp2(c(-1,0, 1), c("blue" ,"white", "red"))
lgd = Legend(col_fun = col_fun_2, title = 'Pearson')
germ_layer_lgd = Legend(labels = c('Ectoderm', 'Mesoderm', 'Endoderm'), title = "germ layer", legend_gp = gpar(fill = c(ecto_color, meso_color, endo_color)))
legend_list = packLegend(lgd, germ_layer_lgd)

get_heat_map = function(correlation_matrix, row_germ_layer, col_germ_layer, color_scale, g_name, show_rows){
  row_index = rownames(correlation_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == row_germ_layer]
  col_index = colnames(correlation_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == col_germ_layer]
  
  temp_corr = correlation_matrix[row_index, col_index]
  g = Heatmap(temp_corr, name=g_name, border = TRUE, show_row_dend = FALSE, show_column_dend = FALSE, show_column_names = FALSE, show_row_names = show_rows,
              show_heatmap_legend = FALSE,
                      col = color_scale, use_raster = TRUE,
                      clustering_distance_rows = "pearson", clustering_distance_columns = "pearson",
                      clustering_method_rows = "ward.D",clustering_method_columns = "ward.D",
                      row_names_gp = gpar(fontsize=8), na_col = 'black')  
  return(g)
}

ecto_ecto = get_heat_map(temp_corr_matrix, 'Ectoderm', 'Ectoderm', col_fun_2, 'ecto_ecto', TRUE)
ComplexHeatmap::draw(ecto_ecto)
meso_meso = get_heat_map(temp_corr_matrix, 'Mesoderm', 'Mesoderm', col_fun_2, 'meso_meso', TRUE)
ComplexHeatmap::draw(meso_meso)
endo_endo = get_heat_map(temp_corr_matrix, 'Endoderm', 'Endoderm', col_fun_2, 'endo_endo', TRUE)
ComplexHeatmap::draw(endo_endo)


#Need to ensure the ordering of tissues matches that from the same-to-same clustering

#Subest for ecto and meso tissues
row_index = rownames(temp_corr_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == 'Ectoderm']
col_index = colnames(temp_corr_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == 'Mesoderm']
temp_corr = temp_corr_matrix[row_index, col_index]

#Reorder to match the clustered orders of the same-to-same gerrm layer clustering
row_index = row_order(ecto_ecto)
col_index = column_order(meso_meso)
temp_corr = temp_corr[row_index, col_index]

#Get the heatmap without clustering
ecto_meso = Heatmap(temp_corr, name='ecto_meso', border = TRUE, show_row_dend = FALSE, show_column_dend = FALSE, show_column_names = FALSE, show_row_names = TRUE, 
                    show_heatmap_legend = FALSE,
                    cluster_rows = FALSE, cluster_columns = FALSE,
                    col = col_fun_2, use_raster = TRUE,
                    row_names_gp = gpar(fontsize=8), na_col = 'black') 
ComplexHeatmap::draw(ecto_meso)


#Subest for ecto and endo tissues
row_index = rownames(temp_corr_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == 'Ectoderm']
col_index = colnames(temp_corr_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == 'Endoderm']
temp_corr = temp_corr_matrix[row_index, col_index]

#Reorder to match the clustered orders of the same-to-same gerrm layer clustering
row_index = row_order(ecto_ecto)
col_index = column_order(endo_endo)
temp_corr = temp_corr[row_index, col_index]

#Get the heatmap without clustering
ecto_endo = Heatmap(temp_corr, name='ecto_endo', border = TRUE, show_row_dend = FALSE, show_column_dend = FALSE, show_column_names = FALSE, show_row_names = TRUE, 
                    show_heatmap_legend = FALSE,
                    cluster_rows = FALSE, cluster_columns = FALSE,
                    col = col_fun_2, use_raster = TRUE,
                    row_names_gp = gpar(fontsize=8), na_col = 'black') 
ComplexHeatmap::draw(ecto_endo )

#Subest for meso and endo tissues
row_index = rownames(temp_corr_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == 'Mesoderm']
col_index = colnames(temp_corr_matrix) %in% germ_layer_df$tissues[germ_layer_df$germ_layers == 'Endoderm']
temp_corr = temp_corr_matrix[row_index, col_index]

#Reorder to match the clustered orders of the same-to-same gerrm layer clustering
row_index = row_order(meso_meso)
col_index = column_order(endo_endo)
temp_corr = temp_corr[row_index, col_index]

#Get the heatmap without clustering
meso_endo = Heatmap(temp_corr, name='ecto_endo', border = TRUE, show_row_dend = FALSE, show_column_dend = FALSE, show_column_names = FALSE, show_row_names = TRUE,
                    show_heatmap_legend = FALSE,
                    cluster_rows = FALSE, cluster_columns = FALSE,
                    col = col_fun_2, use_raster = TRUE,
                    row_names_gp = gpar(fontsize=8), na_col = 'black') 
ComplexHeatmap::draw(meso_endo)


#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/ecto_ecto_corrs.pdf', height = 3, width = 6, useDingbats = FALSE)
ComplexHeatmap::draw(ecto_ecto)
#dev.off()
#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/ecto_meso_corrs.pdf', height = 3, width = 6, useDingbats = FALSE)
ComplexHeatmap::draw(ecto_meso)
#dev.off()
#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/ecto_endo_corrs.pdf', height = 3, width = 6, useDingbats = FALSE)
ComplexHeatmap::draw(ecto_endo)
#dev.off()
#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/meso_meso_corrs.pdf', height = 3, width = 6, useDingbats = FALSE)
ComplexHeatmap::draw(meso_meso)
#dev.off()
#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/meso_endo_corrs.pdf', height = 3, width = 6, useDingbats = FALSE)
ComplexHeatmap::draw(meso_endo)
#dev.off()
#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/endo_endo_corrs.pdf', height = 3, width = 6, useDingbats = FALSE)
ComplexHeatmap::draw(endo_endo)
#dev.off()

#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/correlation_legend.pdf', height = 3, width = 6, useDingbats = FALSE)
grid.rect()
ComplexHeatmap::draw(legend_list, just = c('left', 'bottom'))
#dev.off()

```


  

Only show the significant correlations, the pairwise correlations can be used to look at individual tissues, but for this graph we really just want to say
There are significant sskew correlations across the germ layers

Plot the distribution of correlations for tissues across germ layers
```{r}
#Filtering on pvalue
#index = tiss_tiss_corr_pvalues > .05
temp_corr_matrix = tiss_tiss_correlation_matrix
temp_corr_matrix[non_sig_corrs_index] = NA
index = tiss_tiss_donors_df < 20
temp_corr_matrix[index] = NA

temp_corr_matrix = temp_corr_matrix[!rownames(temp_corr_matrix) %in% exclude,!colnames(temp_corr_matrix) %in% exclude ]



germ_layers = vector(mode='character', length=length(rownames(temp_corr_matrix)))
for( i in 1:length(rownames(temp_corr_matrix))){
  index = which(skew_and_stats_df$tissue == rownames(temp_corr_matrix)[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
}

meso_to_ecto = c(temp_corr_matrix[germ_layers == 'Mesoderm', germ_layers == 'Ectoderm'])
meso_to_endo = c(temp_corr_matrix[germ_layers == 'Mesoderm', germ_layers == 'Endoderm'])
endo_to_ecto = c(temp_corr_matrix[germ_layers == 'Endoderm', germ_layers == 'Ectoderm'])

index = upper.tri(temp_corr_matrix[germ_layers == 'Endoderm', germ_layers == 'Endoderm'], diag = FALSE)
endo_to_endo = temp_corr_matrix[germ_layers == 'Endoderm', germ_layers == 'Endoderm'][index]

index = upper.tri(temp_corr_matrix[germ_layers == 'Ectoderm', germ_layers == 'Ectoderm'], diag = FALSE)
ecto_to_ecto = temp_corr_matrix[germ_layers == 'Ectoderm', germ_layers == 'Ectoderm'][index]

index = upper.tri(temp_corr_matrix[germ_layers == 'Mesoderm', germ_layers == 'Mesoderm'], diag = FALSE)
meso_to_meso = temp_corr_matrix[germ_layers == 'Mesoderm', germ_layers == 'Mesoderm'][index]


labels = c(rep('Mesoderm--Ectoderm', length(meso_to_ecto)), rep('Mesoderm--Endoderm', length(meso_to_endo)), rep('Endoderm--Ectoderm', length(endo_to_ecto)), 
           rep('Mesoderm--Mesoderm', length(meso_to_meso)), rep('Ectoderm--Ectoderm', length(ecto_to_ecto)), rep('Endoderm--Endoderm', length(endo_to_endo)))
within_across_label = c(rep('Across germ layers', length(c(meso_to_ecto,meso_to_endo,endo_to_ecto))), rep('Within germ layer', length(c(meso_to_meso,ecto_to_ecto,endo_to_endo))) )
data =c(meso_to_ecto,meso_to_endo, endo_to_ecto ,meso_to_meso,ecto_to_ecto, endo_to_endo)

labels = factor(labels, levels=c('Mesoderm--Ectoderm','Mesoderm--Endoderm','Endoderm--Ectoderm', 'Endoderm--Endoderm', 'Mesoderm--Mesoderm','Ectoderm--Ectoderm'))

temp_df = data.frame(skew_correlations = data, labels = labels, germ_layer_comp =within_across_label  )


fill_col = c('#A4A4A4', '#A4A4A4', '#A4A4A4', endo_color, meso_color, ecto_color)

ggplot(temp_df, aes(x=labels, y=skew_correlations)) + geom_violin(color = 'black') + geom_jitter(position=position_jitter(0.1), size = .75) +
  xlab('Germ layer comparisons') + ylab('Pearson correlations') +
  ggtitle('Skew is correlated across germ layers') + geom_hline(yintercept = 0, color = 'red', linetype = 'dashed', size = 1) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=12), axis.title.x = element_text(face='bold',size=12),
        plot.title=element_text(hjust=.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        axis.text.y = element_text(size=12, face = 'bold'), axis.title.y = element_text(face='bold',size=12),
        panel.border = element_rect(colour = "black", fill=NA, size=1), panel.background = element_rect(fill = 'white'),
        plot.margin = margin(.1, .1, .1, .1, 'in')) +
  facet_wrap(~germ_layer_comp, scales = 'free_x') 


#ggsave(filename ='violin_skew_correlations_across_germlayers.pdf', device = 'pdf', 
#      path = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/', dpi = 300, height = 4, width = 6, useDingbats = FALSE )




```

```{r}

across_index = temp_df$germ_layer_comp == 'Across germ layers'
range(temp_df$skew_correlations[across_index], na.rm = T)

within_index = temp_df$germ_layer_comp == 'Within germ layer'
range(temp_df$skew_correlations[within_index], na.rm = T)
```





Try the ridge plots


Only show the significant correlations, the pairwise correlations can be used to look at individual tissues, but for this graph we really just want to say
There are significant sskew correlations across the germ layers

```{r}

ggplot(temp_df, aes(x = skew_correlations, y=labels)) + 
  geom_density_ridges_gradient(scale=1, aes(fill = labels),
    jittered_points = TRUE, rel_min_height = 0.01,
    position = position_points_jitter(width = 0, height = 0),
    point_shape = '|', point_size = 3, point_alpha = 1, alpha = 0.7) +
  xlim(0,1)+
  xlab('Tissue skew correlations')+
  scale_fill_manual(values=c('#A4A4A4', '#A4A4A4', '#A4A4A4',endo_color, meso_color, ecto_color)) +
  theme(plot.title=element_text(hjust=.5, size=14), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
       panel.border = element_rect(colour = "black", fill=NA, size=1), 
       axis.title.y = element_blank(), axis.text.y = element_text(size = 12), 
       axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 12),
       legend.position = "none", 
       panel.background = element_rect(fill = 'white')) +
  facet_wrap(~germ_layer_comp, scales = 'free_y')


ggsave(filename ='ridge_density_skew_correlations_across_germlayers.pdf', device = 'pdf', 
     path = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/', dpi = 300, height = 4, width = 6, useDingbats = FALSE )

```

Show density plots of within germ layer over across germ layer

```{r}

within_germ_layer = c(ecto_to_ecto,endo_to_endo, meso_to_meso)
across_germ_layer = c(meso_to_ecto,meso_to_endo, endo_to_ecto)

labels = c(rep('Within germ layers', length(within_germ_layer)), rep('Across germ layers', length(across_germ_layer)))

temp_df = data.frame(tissue_correlations = c(within_germ_layer, across_germ_layer), labels = labels)

ggplot(data=temp_df, aes(x=tissue_correlations, group=labels, fill=labels)) +
    geom_density(alpha=.5, trim = TRUE) + scale_fill_brewer(palette="Dark2")


```



Combine all the within or across comparisons at the skew level, get an overall 2D density plot of the skew comparisons from tissues that derive the correlation values
Just the significant comparisons

```{r}
temp_corr_matrix = tiss_tiss_correlation_matrix
temp_corr_matrix[non_sig_corrs_index] = NA
index = tiss_tiss_donors_df < 20
temp_corr_matrix[index] = NA

germ_layers = vector(mode='character', length=length(rownames(temp_corr_matrix)))
for( i in 1:length(rownames(temp_corr_matrix))){
  index = which(skew_and_stats_df$tissue == rownames(temp_corr_matrix)[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
}

r_names = rownames(temp_corr_matrix)
c_names = r_names

get_germ_comp_skews = function(germ_1_tissues, germ_2_tissues, comparison_name, within = FALSE){
  
  #go through and grab the skews for each tissue comparison
  tiss_1_name = c()
  tiss_2_name = c()
  all_tiss_1_skews = c()
  all_tiss_2_skews = c()
  
  for( i in 1:length(germ_1_tissues)){
    
    if(within){
      j_indices = i:length(germ_1_tissues)
    }else{j_indices = 1:length(germ_2_tissues)}
    
    
    for(j in j_indices){
      
      #Check that there is a significant correlation
      #r_index = r_names == germ_1_tissues[i]
      #c_index = c_names == germ_2_tissues[j]
      if(is.na(temp_corr_matrix[i, j])){next}
      
      #Grab and order the skews according to donor
      tiss_1 = germ_1_tissues[i]
      tiss_2 = germ_2_tissues[j]
      test_donors = tiss_tiss_donors_names_matrix[[tiss_1,tiss_2]]
      
      if(length(test_donors) <= 1){next}
      #Make sure to reorder the skews so the donor order matches the test_donor order. Ensures that the order of skew for both tissues is the same by donor
      tiss_1_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
      tiss_1_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
      index_1 = match(test_donors, tiss_1_donors)
      tiss_1_skews = tiss_1_skews[index_1]
      
      
      tiss_2_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
      tiss_2_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
      index_2 = match(test_donors, tiss_2_donors)
      tiss_2_skews = tiss_2_skews[index_2]   
      
      
      tiss_1_label = rep(germ_1_tissues[i], length(tiss_1_skews))
      tiss_2_label = rep(germ_2_tissues[j], length(tiss_2_skews))
      
      tiss_1_name = c(tiss_1_name,tiss_1_label)
      tiss_2_name = c(tiss_2_name,tiss_2_label)
      all_tiss_1_skews = c(all_tiss_1_skews, tiss_1_skews)
      all_tiss_2_skews = c(all_tiss_2_skews, tiss_2_skews)
    }
  }
  
  comparison_label = rep(comparison_name, length(tiss_1_name))
  comp_df = data.frame(comparison = comparison_label, tissue_1 = tiss_1_name, tissue_2 = tiss_2_name, tissue_1_skews = all_tiss_1_skews, tissue_2_skews = all_tiss_2_skews  )
  return(comp_df)
}


meso_names = r_names[germ_layers == 'Mesoderm']
ecto_names = r_names[germ_layers == 'Ectoderm']
endo_names = r_names[germ_layers == 'Endoderm']


comp_df_1 = get_germ_comp_skews(meso_names, ecto_names, 'Mesoderm-Ectoderm')
comp_df_2 = get_germ_comp_skews(meso_names, endo_names, 'Mesoderm-Endoderm')
comp_df_3 = get_germ_comp_skews(endo_names, ecto_names, 'Endoderm-Ectoderm')
#To make it symmetric
comp_df_1_1 = get_germ_comp_skews(ecto_names, meso_names, 'Ectoderm-Mesoderm')
comp_df_2_2 = get_germ_comp_skews(endo_names, meso_names, 'Endoderm-Mesoderm')
comp_df_3_3 = get_germ_comp_skews(ecto_names, endo_names, 'Ectoderm-Endoderm')


across_germ_skew_df = rbind(comp_df_1, comp_df_2, comp_df_3, comp_df_1_1, comp_df_2_2, comp_df_3_3)

comp_df_4 = get_germ_comp_skews(ecto_names, ecto_names, 'Ectoderm-Ectoderm', within = F)
comp_df_5 = get_germ_comp_skews(endo_names, endo_names, 'Endoderm-Endoderm', within = F)
comp_df_6 = get_germ_comp_skews(meso_names, meso_names, 'Mesoderm-Mesoderm', within = F)

within_germ_skew_df = rbind(comp_df_4, comp_df_5, comp_df_6)

```

```{r}
#Exclude the same tissue comparisons

same_tiss_index = within_germ_skew_df$tissue_1 == within_germ_skew_df$tissue_2
within_germ_skew_df = within_germ_skew_df[!same_tiss_index, ]


```

See what binning all unconf skews to .55 looks like
```{r}
temp_df = across_germ_skew_df

temp_df$tissue_1_skews[temp_df$tissue_1_skews < .6] = .55 
temp_df$tissue_2_skews[temp_df$tissue_2_skews < .6] = .55
```





BrBG, PiYG, PRGn, PuOr, RdBu, RdGy, RdYlBu, RdYlGn, Spectral

```{r}
# Plot
ggplot(across_germ_skew_df, aes(x=tissue_1_skews, y=tissue_2_skews) ) +
  stat_density_2d(aes(fill = ..density..), geom = "raster", contour = FALSE, n = 200) +
  scale_fill_distiller(palette= "Spectral", direction=-1) +
  ylim(.5,1) + xlim(.5,1) + coord_fixed() +
  ylab('Different germ layer skews') + xlab('Different germ layer skews') +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
       panel.border = element_rect(colour = "black", fill=NA, size=1), 
       axis.title.y = element_text(size =12), axis.text.y = element_text(size = 12), 
       axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 12),
       panel.background = element_rect(fill = 'white'))

ggsave(filename ='combined_skews_density_across_germlayers.pdf', device = 'pdf', 
       path = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/', dpi = 300, height = 4, width = 6, useDingbats = FALSE )


ggplot(within_germ_skew_df, aes(x=tissue_1_skews, y=tissue_2_skews) ) +
  stat_density_2d(aes(fill = ..density..), geom = "raster", contour = FALSE, n = 200) +
  scale_fill_distiller(palette= "Spectral", direction=-1) +
  ylim(.5,1) + xlim(.5,1) + coord_fixed() +
  ylab('Same germ layer skews') + xlab('Same germ layer skews') +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
       panel.border = element_rect(colour = "black", fill=NA, size=1), 
       axis.title.y = element_text(size =12), axis.text.y = element_text(size = 12), 
       axis.title.x = element_text(size = 12), axis.text.x = element_text(size = 12),
       panel.background = element_rect(fill = 'white'))

ggsave(filename ='combined_skews_density_within_germlayers.pdf', device = 'pdf', 
       path = '/home/werner/xchrom_snp_skew/code/graphs/tiss_tiss_correlations/updated/', dpi = 300, height = 4, width = 6, useDingbats = FALSE )
```

Individual comparisons

```{r}

plot_combined_skews_corr = function(df, title){
  g = ggplot(df, aes(x=tissue_1_skews, y=tissue_2_skews) ) + coord_fixed() +
    #stat_density_2d(aes(fill = ..density..), geom = "raster", contour = FALSE, n = 100) +
    #scale_fill_distiller(palette= "Spectral", direction=-1) +
    geom_point(size = .1)+
    ggtitle(title) + ylim(.5,1) + xlim(.5,1) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
         panel.border = element_rect(colour = "black", fill=NA, size=1), 
         panel.background = element_rect(fill = 'white'))
  return(g)
}

same_tiss_index = comp_df_4$tissue_1 == comp_df_4$tissue_2
comp_df_4 = comp_df_4[!same_tiss_index, ]
same_tiss_index = comp_df_5$tissue_1 == comp_df_5$tissue_2
comp_df_5 = comp_df_5[!same_tiss_index, ]
same_tiss_index = comp_df_6$tissue_1 == comp_df_6$tissue_2
comp_df_6 = comp_df_6[!same_tiss_index, ]


g = plot_combined_skews_corr(comp_df_1,'Mesoderm-Ectoderm' )
print(g)
g = plot_combined_skews_corr(comp_df_2,'Mesoderm-Endoderm' )
print(g)
g = plot_combined_skews_corr(comp_df_3,'Endoderm-Ectoderm' )
print(g)
g = plot_combined_skews_corr(comp_df_4,'Ectoderm-Ectoderm' )
print(g)
g = plot_combined_skews_corr(comp_df_5,'Endoderm-Endoderm' )
print(g)
g = plot_combined_skews_corr(comp_df_6,'Mesoderm-Mesoderm' )
print(g)

g = plot_combined_skews_corr(across_germ_skew_df,'Across germ layers' )
print(g)

g = plot_combined_skews_corr(within_germ_skew_df,'Within germ layers' )
print(g)

```


Why are there weird structures in the skews, like regular groupings??


```{r}
hist(skew_and_stats_df$skew,breaks=seq(.5,1,.001), ylim = c(0,75), xlim = c(.55,.75),  xaxt = 'n')
hist(mean_skews, breaks = seq(.5,1,.001), ylim = c(0,75), xlim = c(.55,.75),  xaxt = 'n', add=T, col = 'black')
axis(side = 1, at = seq(.5,1,.001))

```



#################################################################################################
Using Nesbit 1971 method for estimating lineage cell numbers

Idea is each tissue's skew variance is built from 2 binomial samples. The first is the binomial derived from the number of cells at the time of XCI. The second is the binomial derived from the number of cells committing to that tissue's lineage. 

This means the variance in skew for a particular tissue is a combination of the two distribution's variances. 
Our previous cell number estimates fitting to the total variance in skew per tissue is most certainly an overestimate of variance and an underestimate of XCI cell number

Nesbitt formulated the relation

1/N = 1/n1 + 1/n2 - 1/n1*n2

Where N is the cell number fitting to the total tissue varianc (our previous estimates), n1 is the XCI cell number and n2 is the tissue-specific number
n1 is estimated through the relationship that cov(tiss_1, tiss_2) = var(XCI skew)
The degree two tisssues covary, the variation they share, is coming from the binomial sampling they share, which is the original XCI skew.
So, using cov = corr * SD1 * SD2
we can estimate n1 using tiss_1-tiss_2 correlations and their SDs

explicitly

n1 = p * q / ( corr * SD1 * SD2 )

Can specifically look at the significant correlations too

Once I have the n1 estimates, solve for n2

n2 = (1 - n1)  / (1 - (n1/N))

For n1 estimates, I'll need
tiss_tiss_correlation_matrix
tiss_tiss_corr_pvalues
and the SDs of each tissues' skew distributions

Will end up with a pairwise matrix of n1 estimates
####################################################################################################

I think it's supposed to be calculated with the unfolded skews. But should filter out the tissue comparisons that have negative correlations. Unfolding will give a positive correlation (Simpsons paradox)
We know the vast majority of tissues have positive skew correlations, the negative ones, are probably tissue specific biology not related to initial skew variance


```{r}

xci_num_cell_estimates = matrix(nrow = length(tissues), ncol = length(tissues))
rownames(xci_num_cell_estimates) = tissues
colnames(xci_num_cell_estimates) = tissues

for(i in 1:length(tissues)){
  
  tiss_1 = rownames(tiss_tiss_donors_names_matrix)[i]

  for(j in 1:length(tissues)){

    tiss_2 = colnames(tiss_tiss_donors_names_matrix)[j]
    test_donors = tiss_tiss_donors_names_matrix[[i,j]]
    
    if(length(test_donors) <= 1){next}
    #Make sure to reorder the skews so the donor order matches the test_donor order. Ensures that the order of skew for both tissues is the same by donor
    tiss_1_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
    tiss_1_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_1 & skew_and_stats_df$donor %in% test_donors]
    index_1 = match(test_donors, tiss_1_donors)
    tiss_1_skews = tiss_1_skews[index_1]
    
    
    tiss_2_skews = skew_and_stats_df$skew[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
    tiss_2_donors = skew_and_stats_df$donor[skew_and_stats_df$tissue == tiss_2 & skew_and_stats_df$donor %in% test_donors]
    index_2 = match(test_donors, tiss_2_donors)
    tiss_2_skews = tiss_2_skews[index_2]
    
    #Grab the correlation, if the original correlation iss negative, skip that tissue comparison
    original_corr = tiss_tiss_correlation_matrix[i,j]
    if(original_corr < 0 | is.na(original_corr)){next}
    correlation = cor(unfold(tiss_1_skews), unfold(tiss_2_skews), method  ='pearson')
    #Grab the standard deviations
    tiss_1_sd = sd(unfold(tiss_1_skews))
    tiss_2_sd = sd(unfold(tiss_2_skews))
  
    #Calculate the n1 estimate, under the assumption that p = .5
    xci_num_cell_estimates[i,j] = .25 / (correlation * tiss_1_sd * tiss_2_sd)
    
  }
  
}


#Ignore the cultured tissue types
xci_num_cell_estimates[ ,colnames(xci_num_cell_estimates) %in% c('Cells - Cultured fibroblasts','Cells - EBV-transformed lymphocytes','Cells - Transformed fibroblasts')] = NA

#Also ignore the diagonal
diag(xci_num_cell_estimates) = NA

#Ignore NA
index_na = !is.na(xci_num_cell_estimates)

```




```{r}

n1_ests_increasing_data_points = c()
labels = c()

min_data_thresh = c(20,30,40,50,60,70,80,90,100)
for(i in 1:length(min_data_thresh)){
  
  #Filter out the comparisons wih only a handful of donors
  index_num_donors = tiss_tiss_donors_df >= min_data_thresh[i]
  final_index = index_num_donors & index_na
  n1_ests = xci_num_cell_estimates[final_index]
  label = rep(sprintf('Using a minimum of %i data points', min_data_thresh[i]), length(n1_ests))
  
  n1_ests_increasing_data_points = c(n1_ests_increasing_data_points, n1_ests)
  labels = c(labels, label)

}

n1_ests_increasing_data_points_df = data.frame(n1_ests =n1_ests_increasing_data_points, minimum_number_of_data_points = labels )
factor_levels = unique(n1_ests_increasing_data_points_df$minimum_number_of_data_points)
n1_ests_increasing_data_points_df$minimum_number_of_data_points = factor(n1_ests_increasing_data_points_df$minimum_number_of_data_points, levels = factor_levels)

ggplot(n1_ests_increasing_data_points_df, aes(x = minimum_number_of_data_points, y = n1_ests)) + geom_violin() + stat_summary(fun=mean, geom="point", size=2, color="red") +
  theme(axis.text.x = element_text(angle = 90)) + geom_hline(yintercept = 16) + geom_hline(yintercept = 8)


```

Why is there a bimodal shift with the tissues that have the most data points? Check it out

```{r}
temp_n1_ests = xci_num_cell_estimates

index_num_donors = tiss_tiss_donors_df >= 100
temp_n1_ests[!index_num_donors] = NA

rowMeans(temp_n1_ests, na.rm = TRUE)
```



Going with an n1 estimate of 13 at the moment

```{r}
n1_ests_increasing_data_points_df %>% group_by(minimum_number_of_data_points) %>% summarise_at(vars(n1_ests), funs(mean(., na.rm=TRUE)))

```


```{r}
n1_df_ggplot = melt(xci_num_cell_estimates)
ggplot(n1_df_ggplot, aes(x = Var2, y = value)) + geom_boxplot() +
  theme(axis.text.x = element_text(angle = 90))
```


Now using the N and n1 estimates, get an estimate for the n2, tissue-specific cell numbers
est_stem_cell_pop_df has the N estimates, use the normal_est_n
est_tissue_lineage_cell_num_df has the old n2 estimates using the z-score method, good to compare, use the est_n_error 

```{r}

n1 = 13
N = est_stem_cell_pop_df$normal_est_n


n2_estimates = ((N*(1 - n1)) / (N - n1))
est_tissue_lineage_cell_num_df$Nesbitt_method = n2_estimates

est_tissue_lineage_cell_num_df
```


```{r}

ggplot(est_tissue_lineage_cell_num_df, aes(x=est_n, y=Nesbitt_method, color=germ_layers, size=num_samples)) + geom_point()+
  theme(plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.y = element_text(face="bold", size=12), axis.text.x = element_text(face="bold", size=12)) + 
  ggtitle('Lineage specific cell number estimates') + ylab('Nesbitt method') +
  xlab('Using actual zscore sample mean') + geom_abline(intercept = 0, slope = 1, color = 'red') +
  scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color))

ggplot(est_tissue_lineage_cell_num_df, aes(x=est_n, y=Nesbitt_method, color=germ_layers, size=num_samples)) + geom_point()+ ylim(0,80) + xlim(0,80) +
  theme(plot.title=element_text(hjust=.5), plot.margin=margin(.1,.1,.01,.1,'in'),
        panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.text.y = element_text(face="bold", size=12), axis.text.x = element_text(face="bold", size=12)) + 
  ggtitle('Lineage specific cell number estimates') + ylab('Nesbitt method') +
  xlab('Using actual zscore sample mean') + geom_abline(intercept = 0, slope = 1, color = 'red') +
  scale_color_manual(breaks = c('Ectoderm','Endoderm','Mesoderm'), values=c(ecto_color, endo_color, meso_color))


```


```{r}
est_stem_cell_pop_df

```



There's issues with the Nesbitt method. Really assumes that the N estimate will come from a distribution that over-represents the skew variation created from XCI, because of the second round of sampling with the tissue. But what happens when N is a pretty close estimate to n1? This can arise when tissues are derived from large cell numbers, mostly follow the initial skew set by XCI and thus most of the skew variation is from XCI. Then N ~= n1 and you can get weird n2 estimates from the Nesbitt method, negative infinity. That's also assumming you sample each tissue with enough depth to capture the full extent of the variation. Some of these tissues are only sampled from ~20-30 donors, so mostlikely not capturing the true variation and thus are underestimates of variation, n1 < N resulting in negative n2 estimates.

We know there are some strong skew correlations across germ layer tissues, suggesting XCI occurs early and in a small number of cells. But what about the tissue specification? Most likely there's going to be a range, some derived from large cell numbers and some derived from small cell numbers. What would this look like. Small cell number derived tissues will have more skew variance, less likelyh to have strong correlations with other tissues, both within and across germ layers. Using the z-scored based method for n2 estimates, which has nothing to do with skew correlations, see if the smaller n2 tissues have smaller correlations. Do they have smaller within germ layer correlations than expected?


Get an order of tissues based on their n2 estimates, using the z-scored method. Plot their skew correlations

```{r}

n2_index = order(est_tissue_lineage_cell_num_df$est_n)
ordered_tissues_by_n2 = est_tissue_lineage_cell_num_df$tissues[n2_index]

correlations_ranked_by_n2_df = melt(tiss_tiss_correlation_matrix, value.name = 'Pearson', varnames = c('pairwise_tissue', 'tissue'))
correlations_ranked_by_n2_df$tissue = factor(correlations_ranked_by_n2_df$tissue, levels = ordered_tissues_by_n2)

ggplot(correlations_ranked_by_n2_df, aes(x = tissue, y = Pearson))  + geom_violin() + stat_summary(fun=mean, geom="point", size=1, color="black") +
  theme(axis.text.x = element_text(angle = 90),panel.grid.major = element_blank(), panel.grid.minor = element_blank() )
  
```


```{r}
pearson_means = correlations_ranked_by_n2_df %>% group_by(tissue) %>% summarise_at(vars(Pearson), funs(mean(., na.rm=TRUE)))
pearson_means = pearson_means$Pearson
n2_estimates = est_tissue_lineage_cell_num_df$est_n[n2_index]

length(pearson_means)

plot(n2_estimates, pearson_means, xlab = 'Tissue-specific primordial stem pool cell number', ylab = 'Mean skew correlation' )
abline(lm(pearson_means~n2_estimates), lwd = 3)

```


Now split the pearson means by within and across germ layers


```{r}

mean_pearson_within_germlayer = c()
mean_pearson_across_germlayer = c()


for(i in 1:length(ordered_tissues_by_n2)){
  
  #Get the germlayer for that tissue
  current_germ = germ_layer_df$germ_layers[germ_layer_df$tissues == as.character(ordered_tissues_by_n2[i])]
  
  within_germ = as.character(germ_layer_df$tissues[germ_layer_df$germ_layers == current_germ])
  across_germ = as.character(germ_layer_df$tissues[germ_layer_df$germ_layers != current_germ])
  
  within_corrs = tiss_tiss_correlation_matrix[rownames(tiss_tiss_correlation_matrix) %in% within_germ , as.character(ordered_tissues_by_n2[i]) ]
  across_corrs = tiss_tiss_correlation_matrix[rownames(tiss_tiss_correlation_matrix) %in% across_germ , as.character(ordered_tissues_by_n2[i]) ]
  
  mean_pearson_within_germlayer = c(mean_pearson_within_germlayer, mean(within_corrs, na.rm = TRUE))
  mean_pearson_across_germlayer = c(mean_pearson_across_germlayer, mean(across_corrs, na.rm = TRUE))

}


n2_est_within_across_skew_corrs_df = data.frame(tissue = as.character(ordered_tissues_by_n2), n2_ests = n2_estimates,
                                                mean_within_germlayer_corrs = mean_pearson_within_germlayer, mean_across_germlayer_corrs = mean_pearson_across_germlayer  ) 

plot(n2_estimates,mean_pearson_within_germlayer, col = 'red', pch = 19, cex=.5, xlab = 'Tissue-specific primordial stem pool cell number', ylab = 'Mean skew correlation' )
abline(lm(mean_pearson_within_germlayer ~ n2_estimates), col = 'red3', lwd = 3)
points(n2_estimates, mean_pearson_across_germlayer, col = 'blue', pch = 19, cex=.5 )
abline(lm(mean_pearson_across_germlayer ~ n2_estimates), col = 'blue3', lwd = 3)

legend("bottomright", 
       pch = c(19, 19, 19), 
       c("Within germlayers", "Across germlayers"), 
       col = c("red", "blue"))

plot(n2_estimates,mean_pearson_within_germlayer, col = 'red', pch = 19, cex=.5, xlab = 'Tissue-specific primordial stem pool cell number', ylab = 'Mean skew correlation' )
abline(lm(mean_pearson_within_germlayer ~ n2_estimates), col = 'red3', lwd = 3)
plot(n2_estimates, mean_pearson_across_germlayer, col = 'blue', pch = 19, cex=.5, xlab = 'Tissue-specific primordial stem pool cell number', ylab = 'Mean skew correlation' )
abline(lm(mean_pearson_across_germlayer ~ n2_estimates), col = 'blue3', lwd = 3)

```

```{r}
as.character(ordered_tissues_by_n2)[mean_pearson_across_germlayer < .3 & n2_estimates > 30]

print(' ')

as.character(ordered_tissues_by_n2)[mean_pearson_across_germlayer > .3 ]


```






########################################
Diving into the AUCs
########################################


The global skew vs AUC distribution for all pair-wise comparisons

```{r}
plot(all_skews, all_aucs, xlim=c(.5,1), ylim = c(0,1), main='AUC distribution for all pair-wise tissue comparisons', cex=.5)


```

```{r}
#Convert Nans to NA for plotting

for( i in 1:length(tissues)){
  
  tiss_tiss_7_aucs_matrix[i, !is.finite(tiss_tiss_7_aucs_matrix[i, ]) ] = NA
  binned_tiss_tiss_7_aucs_matrix[i, !is.finite(tiss_tiss_7_aucs_matrix[i, ]) ] = NA
  
}


```



Look at the .70 skew thresholded AUC distributions for each tissue


```{r}
for( i in 1:length(tissues)){
  
  
  rid_diagonal = seq(1, length(tissues), 1)
  
  #Dont include the diagonal AUC, will always be one
  aucs_70 = tiss_tiss_7_aucs_matrix[i, rid_diagonal[rid_diagonal != i] ]
  
  if(sum(is.na(aucs_70))==length(tissues)-1){next}
  
  
  title = sprintf('%s', tissues[i])
  x_title = "AUCs at a .7 skew +'s threshold for all tissues "
  dens = density(aucs_70, cut=0, na.rm=TRUE)
  hist(aucs_70, main = title, xlab=x_title, xlim=c(0,1), ylim = c(0,max(dens$y)+1.5), freq=FALSE)
  lines(dens, lwd=2)
  
  
  
  #Save adipose_sub and blood as examples
  if(tissues[i] %in% c('Adipose - Subcutaneous', 'Whole Blood')){
    
    file_name = 'temp'
    
    if(tissues[i] == 'Adipose - Subcutaneous'){file_name = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/adipose_sub_auc_70_predicting_other_tiss_skews.pdf'}
    if(tissues[i] == 'Whole Blood'){file_name = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/blood_auc_70_predicting_other_tiss_skews.pdf'}
    
    #pdf(file = file_name, width=5, height=4)
    
    title = sprintf('%s', tissues[i])
    dens = density(aucs_70, na.rm=TRUE)
    hist(aucs_70, main = title, xlab="AUC for predicting other tissue's skew", xlim=c(0,1), ylim = c(0,max(dens$y)+1.5), freq=FALSE, cex.lab=1.5, cex.main=2, cex.axis=1.5)
    lines(dens, lwd=4)    

    #dev.off()
    
    
  }
  
  
  
  
  
}



```

Get a summary plot for all the tissues, plot mean versus var for the AUCs at a .7 skew threshold

```{r}
auc_means = vector(mode='numeric', length=length(tissues))
auc_vars = vector(mode='numeric', length=length(tissues))

for(i in 1:length(tissues)){
  
  auc_means[i] = mean(tiss_tiss_7_aucs_matrix[i, ], na.rm=TRUE)
  auc_vars[i] = var(tiss_tiss_7_aucs_matrix[i, ], na.rm=TRUE)

}
#pdf(file = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/tiss_auc_predict_other_tiss_skews_mean_var.pdf',width=5, height=4 )
par(pty='s', mar=c(4,1,2,1))
mean_thresh = .6
var_thresh=.1

plot(auc_means, auc_vars, xlim = c(.325,1),ylim=c(0,.25), main='All tissues AUC mean and variance', xlab='AUC mean', ylab='AUC variance', pch=19, cex=.7)
index = auc_means<= mean_thresh | auc_vars >=var_thresh
text(auc_means[index], auc_vars[index], labels = tissues[index], cex=.70, pos=3)
lines(x=c(mean_thresh, mean_thresh), y=c(-.01, var_thresh), col='red')
lines(x=c(mean_thresh, 1.1), y=c(var_thresh, var_thresh), col='red')
#dev.off()

par(pty='s', mar=c(4,1,2,1))
mean_thresh = .6
var_thresh=.1

plot(auc_means, auc_vars, xlim = c(.325,1),ylim=c(0,.25), main='All tissues AUC mean and variance', xlab='AUC mean', ylab='AUC variance', pch=19, cex=.7)
index = auc_means<= mean_thresh | auc_vars >=var_thresh
text(auc_means[index], auc_vars[index], labels = tissues[index], cex=.70, pos=3)
lines(x=c(mean_thresh, mean_thresh), y=c(-.01, var_thresh), col='red')
lines(x=c(mean_thresh, 1.1), y=c(var_thresh, var_thresh), col='red')
```

```{r}
#Get germ layers for the filt_tissues

germ_layers = vector(mode='character', length=length(tissues))

for( i in 1:length(tissues)){
  index = which(skew_and_stats_df$tissue == tissues[i])
  germ_layers[i]= as.character(skew_and_stats_df$germ_layer[index[1]])
  
}

germ_layers


```



Or instead just do a series of boxplots

```{r}
diag(tiss_tiss_7_aucs_matrix) = NA
temp_aucs_matrix =tiss_tiss_7_aucs_matrix[!rownames(tiss_tiss_7_aucs_matrix) %in% exclude, !colnames(tiss_tiss_7_aucs_matrix)%in% exclude]
auc_means = rowMeans(temp_aucs_matrix , na.rm=TRUE)
auc_means = auc_means[order(auc_means)]



#Make a data frame, AUCs in one column, tissue in another, reorder tissues to match auc_means order

auc_vector = c()
tiss_vector = c()
germ_layer_vec = c()

for(i in 1:length(rownames(temp_aucs_matrix))){
  aucs = temp_aucs_matrix [i, ]
  aucs = aucs[!is.na(aucs)]
  if(length(aucs)==0){next}
  auc_vector = c(auc_vector, aucs)
  
  tiss_names = rep(rownames(temp_aucs_matrix)[i], length(aucs))
  tiss_vector = c(tiss_vector, tiss_names)
  germ = rep(germ_layers[i], length(aucs))
  germ_layer_vec = c(germ_layer_vec, germ)          
}
  

auc_70_plotting_df = data.frame(tissue = tiss_vector, aucs = auc_vector, germ_layer = germ_layer_vec)

#diag(binned_tiss_tiss_7_aucs_matrix) = NA
#Make a data frame, AUCs in one column, tissue in another, reorder tissues to match auc_means order

#auc_vector = c()
#tiss_vector = c()
#germ_layer_vec = c()

#for(i in 1:length(tissues)){
#  aucs = binned_tiss_tiss_7_aucs_matrix[i, ]
##  aucs = aucs[!is.na(aucs)]
#  if(length(aucs)==0){next}
#  auc_vector = c(auc_vector, aucs)
  
#  tiss_names = rep(tissues[i], length(aucs))
#  tiss_vector = c(tiss_vector, tiss_names)
# germ = rep(germ_layers[i], length(aucs))
#  germ_layer_vec = c(germ_layer_vec, germ)          
#}
  

#binned_auc_70_plotting_df = data.frame(tissue = tiss_vector, aucs = auc_vector, germ_layer = germ_layer_vec)

```

```{r}
auc_70_plotting_df
```



```{r}

auc_70_plotting_df$tissue = with(auc_70_plotting_df, reorder(tissue, aucs, median))

ggplot(auc_70_plotting_df,  aes(x=tissue, y=aucs, fill=germ_layers)) + 
  geom_boxplot(fill='#A4A4A4', color="black") +
  ylab("AUC") + ggtitle('Tissue AUC distributions for predicting other tissue skews') + xlab('')+
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,face="bold", size=8), 
        plot.title=element_text(hjust=.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        axis.text.y = element_text(face="bold", size=12), 
        plot.margin = margin(.1, .1, .1, .1, 'in')) +
  geom_hline(yintercept = .75, col='red', linetype='dashed', size=.5) +scale_fill_brewer(palette="Dark2")

#ggsave(filename='auc_distributions_per_tiss_predicting_other_tiss_skews.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/', units='in', height=5, width=7, dpi=300)


#binned_auc_70_plotting_df$tissue = with(binned_auc_70_plotting_df, reorder(tissue, aucs, median))

#ggplot(binned_auc_70_plotting_df,  aes(x=tissue, y=aucs, fill=germ_layers)) + 
#  geom_boxplot(fill='#A4A4A4', color="black") +
#  ylab("AUC") + ggtitle('Tissue AUC distributions for predicting other tissue skews, binned version') + xlab('')+
#  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,face="bold", size=8), 
#        plot.title=element_text(hjust=.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
#        axis.text.y = element_text(face="bold", size=12), 
#        plot.margin = margin(.1, .1, .1, .1, 'in')) +
#  geom_hline(yintercept = .75, col='red', linetype='dashed', size=.5) +scale_fill_brewer(palette="Dark2")



```


Get example plots demonstrating the process


```{r}

predicting_tiss = 'Esophagus - Muscularis'
good_thresh_tiss = 'Adipose - Visceral (Omentum)'
bad_thresh_tiss = 'Stomach'



good_donors = tiss_tiss_donors_names_matrix[[predicting_tiss, good_thresh_tiss]]
bad_donors = tiss_tiss_donors_names_matrix[[predicting_tiss, bad_thresh_tiss]]



good_auc_1_skews = skew_and_stats_df$skew[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == predicting_tiss]
good_auc_1_donors = skew_and_stats_df$donor[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == predicting_tiss]
index = match(good_donors,good_auc_1_donors )
good_auc_1_skews = good_auc_1_skews[index]

good_auc_2_skews = skew_and_stats_df$skew[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == good_thresh_tiss]
good_auc_2_donors = skew_and_stats_df$donor[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == good_thresh_tiss]
index = match(good_donors,good_auc_2_donors )
good_auc_2_skews = good_auc_2_skews[index]

par(pty='s')
auc = tiss_tiss_7_aucs_matrix[predicting_tiss,good_thresh_tiss]
title = sprintf('AUC : %f', auc)
index = good_auc_2_skews >= .7
#pdf('/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/good_auc_example.pdf', height=4, width=4)
plot(good_auc_1_skews[index], good_auc_2_skews[index], pch=19, col='red', xlim=c(.5,1), ylim=c(.5,1), 
     xlab=sprintf('%s skews', predicting_tiss), ylab=sprintf('%s skews', good_thresh_tiss), main=title, cex.main=1.5, cex.lab=1.25, cex.axis=1.25)
abline(h = .70, col='red', lty='dashed')
points(good_auc_1_skews[!index], good_auc_2_skews[!index], pch=19)
#dev.off()





bad_auc_1_skews = skew_and_stats_df$skew[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == predicting_tiss]
bad_auc_1_donors = skew_and_stats_df$donor[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == predicting_tiss]
index = match(bad_donors,bad_auc_1_donors )
bad_auc_1_skews = bad_auc_1_skews[index]

bad_auc_2_skews = skew_and_stats_df$skew[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == bad_thresh_tiss]
bad_auc_2_donors = skew_and_stats_df$donor[skew_and_stats_df$donor %in% adipose_muscle_donors & skew_and_stats_df$tissue == bad_thresh_tiss]
index = match(bad_donors,bad_auc_2_donors )
bad_auc_2_skews = bad_auc_2_skews[index]

par(pty='s')
auc = tiss_tiss_7_aucs_matrix[predicting_tiss,bad_thresh_tiss]
title = sprintf('AUC : %f', auc)
index = bad_auc_2_skews >= .7
#pdf('/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/bad_auc_example.pdf', height=4, width=4)
plot(bad_auc_1_skews[index], bad_auc_2_skews[index], pch=19, col='red', xlim=c(.5,1), ylim=c(.5,1),
     xlab=sprintf('%s skews', predicting_tiss), ylab=sprintf('%s skews', bad_thresh_tiss), main=title, cex.main=1.5, cex.lab=1.25, cex.axis=1.25)
abline(h = .70, col='red', lty='dashed')
points(bad_auc_1_skews[!index], bad_auc_2_skews[!index], pch=19)
#dev.off()




```




Looking at the paired pvalue distributions for each tissue


Correct way to handle multiple testing correction:
This is all pairwise tissue comparisons, so the matrix is symmetric. that means the test in [2,3] is the same test as in [3,2].
So correcting pvalues in each row or the entire matrix doesn't make sense, as there are multiple pvalues representing the same test
So, will need to grab the upper triangular matrix of pvalues and correct that, then put it back in matrix form for easy exploration



```{r}


temp_paired_wilcoxen_pvals_matrix = tiss_tiss_paired_wilcox_pvals_matrix[!rownames(tiss_tiss_paired_wilcox_pvals_matrix) %in% exclude, !colnames(tiss_tiss_paired_wilcox_pvals_matrix)%in% exclude]

corrected_paired_wilcox_pvals = matrix(nrow = dim(temp_paired_wilcoxen_pvals_matrix)[1], ncol = dim(temp_paired_wilcoxen_pvals_matrix)[1] )

upper_index = upper.tri(temp_paired_wilcoxen_pvals_matrix)
pvals = temp_paired_wilcoxen_pvals_matrix[upper_index] #Upper matrix as a vector
#Do correction here
pvals = p.adjust(pvals, method='BH')

#Put it back into a matrix
corrected_paired_wilcox_pvals[upper.tri(corrected_paired_wilcox_pvals, diag=FALSE)] = pvals
#And make it symmetric again
corrected_paired_wilcox_pvals[lower.tri(corrected_paired_wilcox_pvals)] = t(corrected_paired_wilcox_pvals)[lower.tri(t(corrected_paired_wilcox_pvals))]

rownames(corrected_paired_wilcox_pvals) = rownames(temp_paired_wilcoxen_pvals_matrix)
colnames(corrected_paired_wilcox_pvals) = colnames(temp_paired_wilcoxen_pvals_matrix)
head(corrected_paired_wilcox_pvals)
```

uncorrected pvalue distributions


```{r}

for(i in 1:length(rownames(temp_paired_wilcoxen_pvals_matrix))){
  
  title = sprintf('%s', rownames(temp_paired_wilcoxen_pvals_matrix)[i])
  pvals = temp_paired_wilcoxen_pvals_matrix[i, ]
  
  if(sum(is.na(pvals)) == length(rownames(temp_paired_wilcoxen_pvals_matrix))){next}
  
  hist(pvals, main=title, xlab='paired wilcoxen p.values for all other tissues', freq=FALSE, xlim=c(0,1))
  lines(density(pvals, cut=0, na.rm = TRUE), lwd=2)
}
```



corrected pvalue distributions

```{r}

for(i in 1:length(rownames(temp_paired_wilcoxen_pvals_matrix))){
  
  title = sprintf('%s', rownames(temp_paired_wilcoxen_pvals_matrix)[i])
  pvals = corrected_paired_wilcox_pvals[i, ]
  
  if(sum(is.na(pvals)) == length(rownames(temp_paired_wilcoxen_pvals_matrix))){next}
  
  hist(pvals, main=title, xlab='paired wilcoxen p.values for all other tissues', freq=FALSE, xlim=c(0,1))
  lines(density(pvals, cut=0, na.rm = TRUE), lwd=2)
}
```


```{r}

for( i in 1:length(tissues)){
  
  for(j in 1:length(tissues)){
    
    if(is.na(corrected_paired_wilcox_pvals[i,j])){next}
    
    if(corrected_paired_wilcox_pvals[i,j] <= .05){
      
      tiss_1 = rownames(corrected_paired_wilcox_pvals)[i]
      tiss_2 = colnames(corrected_paired_wilcox_pvals)[j]
      print(sprintf('%s and %s are different', tiss_1, tiss_2))
      
    }
  }
}




```


c('Adipose - Subcutaneous','Artery - Aorta','Artery - Coronary','Cells - EBV-transformed lymphocytes','Liver','Nerve - Tibial','Whole Blood')


```{r}
corrected_paired_wilcox_pvals_df = as.data.frame(temp_paired_wilcoxen_pvals_matrix)
avg_p_val = rowMeans(corrected_paired_wilcox_pvals_df, na.rm=TRUE)
corrected_paired_wilcox_pvals_df =  corrected_paired_wilcox_pvals_df[order(avg_p_val),order(avg_p_val)]
corrected_paired_wilcox_pvals_df = melt(t(corrected_paired_wilcox_pvals_df))

ggplot(corrected_paired_wilcox_pvals_df, aes(x = log10(value), y=Var2, fill=stat(x))) + geom_density_ridges_gradient(scale=8) +
  scale_fill_viridis_c(name='log10(p.value)', direction=-1, option='C') + ggtitle('Wilcoxen signed-rank test p values') + xlab('log10( p.value )') + ylab("") +
  theme(plot.title=element_text(hjust=.5, size=14), axis.text.x = element_text(size = 14, face='bold'), axis.text.y = element_text(size = 10) )  + theme_bw()

ggsave(filename='wilcoxen_paired_pvalues_per_tissue_ridge_plot.pdf', device='pdf', path = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/', units='in', height=6, width=7, dpi=300)

```




```{r}

corrected_paired_wilcox_pvals_df = as.data.frame(corrected_paired_wilcox_pvals)
avg_p_val = rowMeans(corrected_paired_wilcox_pvals_df, na.rm=TRUE)
corrected_paired_wilcox_pvals_df =  corrected_paired_wilcox_pvals_df[order(avg_p_val),order(avg_p_val)]
corrected_paired_wilcox_pvals_df = melt(t(corrected_paired_wilcox_pvals_df))

ggplot(corrected_paired_wilcox_pvals_df, aes(x = log10(value), y=Var2, fill=stat(x))) + geom_density_ridges_gradient(scale=8) +
  scale_fill_viridis_c(name='log10(p.value)', direction=-1, option='C') + ggtitle('Wilcoxen signed-rank test p values') + xlab('log10( FDR adjusted p.value )') + ylab("") +
  theme(plot.title=element_text(hjust=.5, size=14), axis.text.x = element_text(size = 14, face='bold'), axis.text.y = element_text(size = 10) )  + theme_bw()

ggsave(filename='wilcoxen_paired_corrected_pvalues_per_tissue_ridge_plot.pdf', device='pdf', 
       path = '/home/werner/xchrom_snp_skew/code/graphs/skew_variance_within_donors/', units='in', height=6, width=7, dpi=300)

```





#####################################################################
Modeling individual tissue stem cell population sizes.
The population cell size estimate we have is at the time of XCI
Can we extend that to estimate the stem cell size pop at the time the tissue's lineage was determined?


This is another binomial model. The skew of the tissue will follow a binomial where the n is the number of cells committing to that tissue
and p is the skew of individual. Variance about the skew of the individual is linked to the number of cells for that tissue

an individual has a skew of .65, but has a tissue skew of .5 or .8. That's pretty far from the skew of the individual, suggesting a small stem cell pop.
Larger cell number would lead to more closly following the skew of the individual.

Starting off with something simple to test the idea
1.) pull out the donors that have a mean skew between .6 and .65, roughly the same.
2.) From that donor pool, coompile the tissue specific skew distributions. Theoretically these should be binomial distributions of n = N, p = .625
3.) In the folded space, use the same folded normal fitting method.
4.) Use that to solve for N = p*q / var


######################################################################

Create a donor X tissue matrix with the skews as the data

```{r}
all_tissues = names(table(skew_and_stats_df$tissue)) 
all_donors = names(table(skew_and_stats_df$donor))

skew_donor_tissue_mat = matrix(nrow = length(all_donors), ncol = length(all_tissues))
rownames(skew_donor_tissue_mat) = all_donors
colnames(skew_donor_tissue_mat) = all_tissues

for(i in 1:length(all_donors)){
  donor_subset = skew_and_stats_df[skew_and_stats_df$donor == all_donors[i], ]
  donor_index = rownames(skew_donor_tissue_mat) == all_donors[i]
  for(j in 1:nrow(donor_subset)){
    
    tissue_index = colnames(skew_donor_tissue_mat) == as.character(donor_subset$tissue[j])
    skew_donor_tissue_mat[donor_index, tissue_index] = donor_subset$skew[j]
  }
}

```









###################################################
Now we have a sense of which tissues do not share skews with other tissues
-Deviance from donor mean
-AUC analysis, those that do not predict other tissues well, but this approach probably suffers from sample size and tissue sampling bias
-pairwise WSRT, showcasing most tissues do come from the same distribution, select few that dont
-Skillings-Mack test, demonstrating no significant differences in the whole ensemble of skews when excluding those tissues that fail the WSRT

SOOOOOOO

Since it seems that the majority of tissues follow the skew of the individual, next is to lump skews across tissues and get the general binomial distirbution for human XCI
-First just use the average across all tissues, estimate cell number
-Then start excluding the tissues that we think are bad, maybe just do the subset that fail all the above tests?
    -That would be whole blood and the two cell lines as starters, than pancreas, liver, and maybe skin?
-Then do exclusion of tissues by random sampling, excluding 1,2,3,4,5, ect random tissues from each person 

Will showcase the robustness of the cell pop size estimate

Plan is to extend to all of AWS, which may not have tissue annotations, so it'll be great if we can show tissue selection doesn't really matter that much
Sure, there are a few tissues that deviate from the rest, but across hundreds of people, it shouldnt matter


####################################################  


Turn the binomial percentile estimation into a function

null_check = sapply(1:length(ratios.max.genes), function(i) is.null(ratios.max.genes[[i]]))


```{r}


sample_binom = function(num_trials, N, p, empir_quants, quant_filt){
  sample = rbinom(n=num_trials, size=N, prob=p)
  sample_skews = sample/N
  sample_quants = quantile(sample_skews, probs = seq(0, 1, 0.01), type=1)
  sum((empir_quants[quant_filt] - sample_quants[quant_filt])^2)
  
}


cell_pop_est_binom_percentiles = function(skews = NULL, p=.5, n=2:64, num_samples=200, to_plot=TRUE){
  
  #Get the empirical skews for that tissue
  empir_skews = unfold(skews)
  num_skews = length(empir_skews)
  #Get the quantiles for these skews
  empir_quants = quantile(empir_skews, probs = seq(0, 1, 0.01), type=1, na.rm=TRUE)
  #Just minimize over the quantiles <= .4 or >= .6
  quant_filt = empir_quants <= .4 | empir_quants >= .6

  error_matrix = sapply(1:length(n), function(j) sapply(1:num_samples, function(i) sample_binom(num_trials = num_skews, N=n[j], p=.5, empir_quants=empir_quants, quant_filt=quant_filt)))

  estimated_num = n[which.min(colMeans(error_matrix))]
  model_error =  colMeans(error_matrix)[which.min(colMeans(error_matrix))]
  if(to_plot){
    x = rep(n[1], num_samples)
    plot(x, error_matrix[ ,1], ylab = 'Sum sq_error confident empirical - sample percentiles', xlab = 'N', main = 'Error for Bin(n=N) to empirical skews', 
         xlim = c(0,n[length(n)]), ylim = c(0, 8))
    abline(v = estimated_num, col='red')
    for(i in 2:length(n)){
      x =  rep(n[i], num_samples)
      points(x,error_matrix[ ,i] )
      
    }
  }
  
  return(c(estimated_num, model_error))
}


```


```{r}
exclude

```




Excluding low sample tissues and those that seem to deviate from donor skew

```{r}

n = 2:64

bin_error_vector = vector(mode = 'numeric', length = length(n))
norm_error_vector = vector(mode = 'numeric', length = length(n))

exclude_more = c('Cells - EBV-transformed lymphocytes', 'Ovary')
temp_skew_matrix = skew_matrix[!rownames(skew_matrix) %in% exclude_more, !colnames(skew_matrix) %in% exclude_more ]
rownames(temp_skew_matrix)

skews = colMeans(temp_skew_matrix, na.rm=TRUE)
skews = skews[!is.na(skews)]
num_skews = length(skews)
empir_quants = quantile(skews, probs = percentiles, type=1)
quant_filt = empir_quants <= .4 | empir_quants >= .6

  
for(i in 1:length(n)){
  #Get the normal and binomial percentiles for the n parameter
  norm_sd = sqrt( (p*(1-p)) / n[i] )
  norm_percentiles = qnorm(percentiles, mean=p, sd = norm_sd)
  bin_percentiles = qbinom(percentiles, size=n[i], prob=p) / n[i]
  #And the error between the empirical and theoretical percentiles
  bin_error_vector[i] = sum((empir_quants[quant_filt] - bin_percentiles[quant_filt])^2)
  norm_error_vector[i] = sum((empir_quants[quant_filt] - norm_percentiles[quant_filt])^2)
}

#Visualize the error over the different n parameters
plot(n, bin_error_vector, ylab = 'Sum sq_error empirical - binomial percentiles', xlab = 'N', 
     main = sprintf('Num donors= %g', num_skews), xlim = c(0,n[length(n)]))
plot(n, norm_error_vector, ylab = 'Sum sq_error empirical - normal percentiles', xlab = 'N', 
   main = sprintf('Num donors= %g', num_skews), xlim = c(0,n[length(n)]))

#Save the best fitting n
binomial_fitted_lumped = n[which.min(bin_error_vector)]
normal_fitted_lumped = n[which.min(norm_error_vector)]

#Save the variance explained by the best model
norm_sd = sqrt( (p*(1-p)) / normal_fitted_lumped )
norm_percentiles = qnorm(percentiles, mean=p, sd = norm_sd)
#Assess variance explained across whole distribution, rather than the confident skews we fitted to
error = empir_quants - norm_percentiles 
total_sum_squares = sum( (empir_quants  - mean(empir_quants))^2 )
sum_squares_error = sum(error^2)
normal_variance_explained = 1 - (sum_squares_error / total_sum_squares)

#And for the binomial fits
bin_percentiles = qbinom(percentiles, size=binomial_fitted_lumped, prob=p) / binomial_fitted_lumped
error = empir_quants - bin_percentiles 
total_sum_squares = sum( (empir_quants  - mean(empir_quants))^2 )
sum_squares_error = sum(error^2)
binomial_variance_explained = 1 - (sum_squares_error / total_sum_squares) 



dnorm_x = 0:1000/1000
#Fitted normal
fitted_normal = dnorm(dnorm_x, mean = p, sd=norm_sd)

theoretical_norms = list()
n = c(4,8,16,32)
for(i in 1:length(n)){ 
  sigma = sqrt(p*(1-p)/n[i])
  theoretical_norms[[i]] = dnorm(dnorm_x, mean=p, sd = sigma)
  }

hist(unfold(skews), main = 'Average XCI skew per donor', xlim = c(0,1), breaks = 64, freq = FALSE, xlab = 'Average XCI across tissues', ylim = c(0,7))
#hist(skews, main = sprintf('XCI skews: %s', tissue), xlim = c(0,1), breaks = 16, prob=TRUE)
lines(dnorm_x,fitted_normal, lwd = 5, col = 'black')

line_colors = c('tan1','steelblue4','turquoise3','slateblue3')
for(i in 1:length(n)){ 
  lines(dnorm_x,theoretical_norms[[i]], col = line_colors[i], lwd=2)
  }

legend(x = 'topright', legend=c('4 cells','8 cells', '16 cells', '32 cells', sprintf('Estimated %i cells', normal_fitted_lumped)), 
       col = c(line_colors, 'black'), lwd = c(2,2,2,2,5))




```








Excluding the 11 tissues that are consistently bad

```{r}


tiss_exclude = c('Whole Blood','Cells - EBV-transformed lymphocytes','Cells - Transformed fibroblasts', 'Pancreas', 'Liver', 'Adipose - Subcutaneous',
                 'Colon - Sigmoid', 'Esophagus - Muscularis', 'Nerve - Tibial', 'Skin - Sun Exposed (Lower leg)', 'Skin - Not Sun Exposed (Suprapubic)')
temp_skew_matrix = skew_matrix[!rownames(skew_matrix) %in% tiss_exclude, ]

skews = colMeans(temp_skew_matrix, na.rm=TRUE)
hist(unfold(skews), xlim=c(0,1), main='Average donor skew excluding 11 divergent tissues', breaks=64)
#lines(density(unfold(skews)))
cell_pop_est_binom_percentiles(skews)


```




```{r}
tissues_to_samp = tissues[!tissues %in% c('Prostate','Testis','body_site_s')]

num_tiss_to_exclude = 10
num_trials = 100

stem_cell_pop_est_matrix = matrix(nrow = num_trials, ncol = num_tiss_to_exclude)

for( i in 1:num_tiss_to_exclude){
  print(sprintf('Excluding %i tissues... Start: %s', i, Sys.time()))
  for(j in 1:num_trials){

    excluded_tiss = sample(tissues_to_samp, size = i, replace = FALSE)
    tissues_to_keep = tissues_to_samp[!tissues_to_samp %in% excluded_tiss]
    
    temp_skew_matrix = skew_matrix[rownames(skew_matrix) %in% tissues_to_keep, ]
    skews = colMeans(temp_skew_matrix, na.rm=TRUE)
    stem_cell_pop_est_matrix[j,i] = cell_pop_est_binom_percentiles(skews, to_plot=FALSE)
    
  }
  print(sprintf('Excluding %i tissues... End: %s', i, Sys.time()))
}

```

```{r}
to_plot_matrix = melt(stem_cell_pop_est_matrix)
to_plot_matrix$Var2 = as.character(to_plot_matrix$Var2) 
to_plot_matrix$Var2[grepl('10', to_plot_matrix$Var2)] = 'ten'

ggplot(to_plot_matrix,  aes(x=Var2, y=value)) + 
  geom_boxplot(fill='#A4A4A4', color="black") + 
  ylim(0,32) +
  geom_jitter(shape=16, position=position_jitter(0.1)) +
  ylab("Estimated stem cell number") + ggtitle('Excluding random tissues and estimating stem cell number') + xlab('Number of random tissues excluded')+ geom_hline(yintercept = 16, color='red') +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,face="bold", size=10), 
        plot.title=element_text(hjust=.5), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), 
        axis.text.y = element_text(face="bold", size=12), 
        plot.margin = margin(.1, .1, .1, .1, 'in'))

```


```{r}
group_by(to_plot_matrix, Var2)

```

























































